<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Reprophylo by HullUni-bioinformatics</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>Reprophylo</h1>
        <h2>Reproducible Phylogenetics</h2>
        <a href="https://github.com/HullUni-bioinformatics/ReproPhylo" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h3>
<a id="reprophylo" class="anchor" href="#reprophylo" aria-hidden="true"><span class="octicon octicon-link"></span></a>ReproPhylo</h3>

<p>ReproPhylo is a reproducible phylogenetics pipeline written in python and making use of <a href="http://www.biopython.org/">BioPython</a> and other open tools. It is open source software (CC0 public domain) for you to use, modify and distribute as you see fit. This is community software, we would welcome your contributions.</p>

<h3>
<a id="reproducible-phylogenetics" class="anchor" href="#reproducible-phylogenetics" aria-hidden="true"><span class="octicon octicon-link"></span></a>Reproducible Phylogenetics</h3>

<p>We are still largely missing the benefits of reproducibility in phylogenetics. This makes our lives unnecessarily difficult and makes us particularly poorly prepared to confront modern data-rich phylogenomics.</p>

<p>Our ideas are outlined in a series of reproducible phylogenetics blog posts: <a href="http://www.davelunt.net/evophylo/2014/10/reproducible-phylogenetics-part-1-why/">part 1</a>, <a href="http://www.davelunt.net/evophylo/2014/11/reproducible-phylogenetics-part-2a-what/">part 2a</a>, <a href="http://www.davelunt.net/evophylo/2014/11/reproducible-phylogenetics-part-2b-what/">part 2b</a>. ReproPhylo allows you to carry out a phylogenetic analysis using pre-written or self-written commands. This programmatic approach ensures that all stages of the analysis are explicitly recorded, and can be exactly reproduced if required. Input sequence data, and generated intermediate data files (e.g. alignments, metadata), are tracked and held in version control- meaning that there can be no doubt which version of which file was used for any analysis. Reprophylo will write a human-readable detailed graphical report for each experiment. ReproPhylo will create an experimental .zip archive to upload to FigShare or equivalent repository containing the entire experiment. ReproPhylo is best deployed as a Docker container, which includes not just the experimental components but also the phylogenetics programs and any dependencies, ensuring that it can be run exactly as it was on the previous phylogeneticists machine.</p>

<h3>
<a id="authors-and-contributors" class="anchor" href="#authors-and-contributors" aria-hidden="true"><span class="octicon octicon-link"></span></a>Authors and Contributors</h3>

<p>ReproPhylo was developed by Amir Szitenberg (<a href="https://github.com/szitenberg" class="user-mention">@szitenberg</a>) and Dave Lunt (<a href="https://github.com/davelunt" class="user-mention">@davelunt</a>) with contributions and suggestions from a number of other people. We really welcome your bug reports, comments, and contributions to the code, design or documentation.</p>

<h3>
<a id="documentation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation</h3>

<p>Documentation can be found at <a href="http://goo.gl/yW6J1J">http://goo.gl/yW6J1J</a></p>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/HullUni-bioinformatics/ReproPhylo/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/HullUni-bioinformatics/ReproPhylo/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/HullUni-bioinformatics/ReproPhylo"></a> is maintained by <a href="https://github.com/HullUni-bioinformatics">HullUni-bioinformatics</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>
