<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Reprophylo : Reproducible Phylogenomics">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Reprophylo</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/HullUni-bioinformatics/ReproPhylo">View on GitHub</a>

          <h1 id="project_title">ReproPhylo</h1>
          <h2 id="project_tagline">Reproducible Phylogenomics</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/HullUni-bioinformatics/ReproPhylo/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/HullUni-bioinformatics/ReproPhylo/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="reprophylo" class="anchor" href="#reprophylo" aria-hidden="true"><span class="octicon octicon-link"></span></a>ReproPhylo</h3>

<p>ReproPhylo is a reproducible phylogenomics pipeline written in python and making use of <a href="http://www.biopython.org/">Biopython</a> and other open tools. It is open source software (CC0 public domain) for you to use, modify and distribute as you see fit. This is community software, we would welcome your contributions.<br><br>
ReproPhylo is described in this <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004447">publication</a>.</p>
<p xmlns:dct="http://purl.org/dc/terms/">
  <a rel="license"
     href="http://creativecommons.org/publicdomain/zero/1.0/">
    <img src="http://i.creativecommons.org/p/zero/1.0/88x31.png" style="border-style: none;" alt="CC0" />
  </a>
  <br />
</p>
<h3>
<a id="reproducible-phylogenomics" class="anchor" href="#reproducible-phylogenomics" aria-hidden="true"><span class="octicon octicon-link"></span></a>Reproducible Phylogenomics</h3>

<p>We are still largely missing the benefits of reproducibility in phylogenomics. This makes our lives unnecessarily difficult and makes us particularly poorly prepared to confront modern data-rich phylogenomics.</p>

<p>Our ideas are outlined in a series of reproducible phylogenetics blog posts: <a href="http://www.davelunt.net/evophylo/2014/10/reproducible-phylogenetics-part-1-why/">part 1</a>, <a href="http://www.davelunt.net/evophylo/2014/11/reproducible-phylogenetics-part-2a-what/">part 2a</a>, <a href="http://www.davelunt.net/evophylo/2014/11/reproducible-phylogenetics-part-2b-what/">part 2b</a>. ReproPhylo allows you to carry out a phylogenomic analysis using pre-written or self-written commands. This programmatic approach ensures that all stages of the analysis are explicitly recorded, and can be exactly reproduced if required. Input sequence data, and generated intermediate data files (e.g. alignments, metadata), are tracked and held in version control- meaning that there can be no doubt which version of which file was used for any analysis. Reprophylo will write a human-readable detailed graphical report for each experiment. ReproPhylo will create an experimental .zip archive to upload to <a href="http://figshare.com/">FigShare</a> or equivalent repository containing the entire experiment. ReproPhylo is best deployed as a Docker container, which includes not just the experimental components but also the phylogenetics programs and any dependencies, ensuring that it can be run exactly as it was on the previous phylogeneticists machine.</p>

<h3>
<a id="authors-and-contributors" class="anchor" href="#authors-and-contributors" aria-hidden="true"><span class="octicon octicon-link"></span></a>Authors and Contributors</h3>

<p>ReproPhylo was developed by Amir Szitenberg (<a href="https://github.com/szitenberg" class="user-mention">@szitenberg</a>) and Dave Lunt (<a href="https://github.com/davelunt" class="user-mention">@davelunt</a>) with contributions and suggestions from a number of other people. We really welcome your bug reports, comments, and contributions to the code, design or documentation. Follow <a href="https://twitter.com/ReproPhylo" class="user-mention">@ReproPhylo</a> on Twitter for updates and usage hints.</p>

<h3>
<a id="documentation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation</h3>

<p>Documentation can be found at <a href="http://goo.gl/yW6J1J">http://goo.gl/yW6J1J</a></p>

<h3>
<a id="Installation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h3>

<p>Instructions for Docker, Vagrant, WinPython or local installation <a href="https://docs.google.com/document/d/1Q-8B0cvkZw2zMkuP0Af4zZ7FiAvBQPDdGbrLLMgtx_4/edit#heading=h.ocbhdun2ates">here</a></p>

<h3>
<a id="additional-sources" class="anchor" href="#additional-sources" aria-hidden="true"><span class="octicon octicon-link"></span></a>Additional Sources</h3>

<ol><li>
<a href="https://groups.google.com/forum/?hl=en-GB#!forum/reprophylo">Google groups</a> - Support for users and developers.</li>
<li>
<a href="https://registry.hub.docker.com/u/szitenberg/reprophylo/">The DockerHub reporitory</a> - A <a href="https://www.docker.com/">Docker</a> image of the python module in an <a href="http://ipython.org/notebook.html">Jupyter notebook</a> environment.</li>
<li>
<a href="https://github.com/HullUni-bioinformatics/ReproPhyloDockerfile">The Dockerfile reporitory</a> - For a local Docker image build or a local installation on Linux</li>
<li>
<a href="https://github.com/szitenberg/ReproPhyloVagrant">The ReproPhyloVagrant reporitory</a> - A <a href="https://www.vagrantup.com/">Vagrant</a> VM builder for any OS.</li>
<li>
<a href="https://github.com/HullUni-bioinformatics/ReproPhyloGalaxy">The Galaxy distribution repository</a> - A Galaxy distriution including the Galaxy ReproPhylo tools, their dependencies and an INSTALL file.</li>
</ol>
            <a class="twitter-timeline"  href="https://twitter.com/ReproPhylo" data-widget-id="590120698280022016">Tweets by @ReproPhylo</a>
            <script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+"://platform.twitter.com/widgets.js";fjs.parentNode.insertBefore(js,fjs);}}(document,"script","twitter-wjs");</script>
                    
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Reprophylo maintained by <a href="https://github.com/HullUni-bioinformatics">HullUni-bioinformatics</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>             
  </body>
</html>

