{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "\n",
      "## Build guides to describe the phylogenetic partitions"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# like this\n",
      "\n",
      "ssu = Guide('dna','rRNA','18S', ['18s','18S','SSU rRNA','18S ribosomal RNA'])\n",
      "\n",
      "# or like this\n",
      "\n",
      "coi = Guide()\n",
      "coi.name = 'coi'\n",
      "coi.char_type = 'dna'\n",
      "coi.feature_type = 'CDS'\n",
      "coi.aliases = ('cox1','COX1','coi','COI','CoI')\n",
      "\n",
      "# parsing this info from csv file to come.\n",
      "\n",
      "guide_list = GuideList()\n",
      "\n",
      "for guide in (ssu, coi):\n",
      "    guide_list.add_guide_to_list(guide)\n",
      "    \n",
      "# guide_list is a class and not just a list because it only accepts Guide objects\n",
      "\n",
      "print str(guide_list)\n",
      "\n",
      "for guide in guide_list.content:\n",
      "    print str(guide)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "GuideList(18S, coi)\n",
        "Guide(char_type=dna, feature_type=rRNA, name=18S, aliases=18s; 18S; SSU rRNA; 18S ribosomal RNA)\n",
        "Guide(char_type=dna, feature_type=CDS, name=coi, aliases=cox1; COX1; coi; COI; CoI)\n"
       ]
      }
     ],
     "prompt_number": 87
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "    \n",
      "## Build sequence + metadata database"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "db = Database(guide_list)\n",
      "\n",
      "# Read genbank files into the db\n",
      "\n",
      "data_folder = '/home/amir/Dropbox/ReproPhylo/data/'\n",
      "\n",
      "input_filenames_list = (data_folder + 'primates.gb',\n",
      "                        data_folder + 'primates1.gb')\n",
      "\n",
      "db.read_embl_genbank(input_filenames_list) # also removes all the features that are not requested by guides\n",
      "\n",
      "# Add denovo sequences\n",
      "\n",
      "db.read_denovo('data/denovo.fas', 'CDS', 'dna')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Accepted: /home/amir/Dropbox/ReproPhylo/data/primates.gb\n",
        "Accepted:"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " /home/amir/Dropbox/ReproPhylo/data/primates1.gb\n"
       ]
      }
     ],
     "prompt_number": 88
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Edit the metadata"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Add spam qualifier to features by id\n",
      "\n",
      "db.add_qualifier(['KF830702.1_f0','AB572419.1_f0'], 'spam', 'totally')\n",
      "\n",
      "# Add a gene qualifier to the 18S entries based on their product qualifier\n",
      "\n",
      "db.if_this_then_that('18S ribosomal RNA', 'product', 'SSU rRNA', 'gene')\n",
      "\n",
      "# To each of the features, add the species from the source feature\n",
      "\n",
      "db.add_qualifier_from_source('organism')\n",
      "\n",
      "# For the denovo sequences, add the organism from the description\n",
      "\n",
      "db.copy_paste_qual_value('denovo_desc','organism')\n",
      "\n",
      "# add gene qualifier to the denovo sequence so that they can be read by the guide\n",
      "\n",
      "db.add_qualifier(['denovo0_f0','denovo1_f0'], 'gene', 'coi')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 89
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Write records to view (cvs) and retain the db edits for future use (gb)"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "db.write('test.csv', format = 'csv')  # test.csv\n",
      "db.write('test.gb', format = 'genbank') # test.gb"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 90
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Write records per partition "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Put a dictionary with per partition records in the db\n",
      "db.extract_by_guide() # will be done by 'write_by_guide()' automatically if not performed\n",
      "\n",
      "db.write_by_guide() # written to files named guide_name.chosen_format\n",
      "                    # coi.fasta,  18S.fasta\n",
      "                    # writes fasta by default but can be given any supported format as value\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 91
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#Classes\n",
      "\n",
      "# Guide(), GuideList() and Database()"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from Bio import SeqIO\n",
      "import os, csv\n",
      "from Bio.Seq import Seq\n",
      "from Bio.Alphabet import IUPAC\n",
      "from Bio.SeqRecord import SeqRecord\n",
      "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
      "\n",
      "class Guide:\n",
      "       \n",
      "    char_type = 'NotSet'\n",
      "    feature_type = 'NotSet'\n",
      "    name = 'NotSet'\n",
      "    aliases = []\n",
      "    \n",
      "    def __init__(self, char_type=char_type, feature_type=feature_type,\n",
      "                 name=name, aliases=aliases):\n",
      "        self.char_type = char_type\n",
      "        self.feature_type = feature_type\n",
      "        self.name = name\n",
      "        self.aliases = aliases\n",
      "\n",
      "    \n",
      "    def __str__(self):\n",
      "        aliases_str = ('; ').join(self.aliases)\n",
      "        return ('Guide(char_type='+self.char_type+', feature_type='+self.feature_type+\n",
      "                ', name='+self.name+', aliases='+aliases_str+')')\n",
      "                \n",
      "        \n",
      "class GuideList():\n",
      "    \n",
      "    def __init__(self):\n",
      "        self.content = []\n",
      "\n",
      "\n",
      "        \n",
      "    def add_guide_to_list(self, guide):\n",
      "        names = []\n",
      "        for g in self.content:\n",
      "            names.append(g.name)\n",
      "        if isinstance(guide, Guide):\n",
      "            if not guide.name in names:\n",
      "                self.content.append(guide)\n",
      "            else: \n",
      "                raise NameError('Guide ' + guide.name + ' is already included')\n",
      "        else:\n",
      "            raise TypeError(\"Expecting Guide() object\") \n",
      "    \n",
      "    def __str__(self):\n",
      "        guide_list_str = 'GuideList('\n",
      "        for guide in self.content:\n",
      "            guide_list_str += guide.name+', '\n",
      "        return guide_list_str[:-2]+')' \n",
      "\n",
      "# tools to prepare the db\n",
      "\n",
      "def keep_feature(feature, guide_list):\n",
      "    keep = 0\n",
      "    for g in guide_list.content:\n",
      "        if not g.name in g.aliases:\n",
      "            g.aliases.append(g.name)\n",
      "        if feature.type == 'source':\n",
      "            keep = 1\n",
      "        elif feature.type == g.feature_type:\n",
      "            qaul = False\n",
      "            if 'gene' in feature.qualifiers.keys():\n",
      "                qual = 'gene'\n",
      "            elif 'product' in feature.qualifiers.keys():\n",
      "                qual = 'product'\n",
      "            if not qual == False and feature.qualifiers[qual][0] in g.aliases:\n",
      "                keep = 1\n",
      "    if keep == 1:\n",
      "        return True\n",
      "    else:\n",
      "        return False\n",
      "    \n",
      "def dwindle_record(record, guide_list):\n",
      "    dwindled_features = []\n",
      "    feature_count = 0\n",
      "    for feature in record.features:\n",
      "        if keep_feature(feature, guide_list)== True:\n",
      "            if feature.type == 'source' and not 'feature_id' in feature.qualifiers.keys():\n",
      "                feature.qualifiers['feature_id'] = [record.id + '_source']\n",
      "            elif not 'feature_id' in feature.qualifiers.keys():\n",
      "                feature.qualifiers['feature_id'] = [record.id + '_f' + str(feature_count)]\n",
      "                feature_count += 1                                        \n",
      "            dwindled_features.append(feature)\n",
      "    record.features = dwindled_features\n",
      "    return record\n",
      "            \n",
      "            \n",
      "def guess_format(input_filename):\n",
      "\n",
      "    if os.path.exists(input_filename):\n",
      "        os.system('perl guesser.pl ' + input_filename)\n",
      "        guess = open(input_filename + '.format','r').read();\n",
      "        os.remove(input_filename + '.format')\n",
      "        return guess\n",
      "    else:\n",
      "        sys.exit('Cannot guess ' + input_filename + '. Does it exist?')\n",
      "        \n",
      "def is_embl_or_gb(input_filename):\n",
      "    input_format = guess_format(input_filename)\n",
      "    if input_format == 'embl' or input_format == 'genbank':\n",
      "        return True\n",
      "    else:\n",
      "        return False\n",
      "\n",
      "def parse_input(input_filename):\n",
      "    return SeqIO.parse(input_filename, guess_format(input_filename))\n",
      "\n",
      "\n",
      "def list_to_string(List):\n",
      "    string = ''\n",
      "    for i in List:\n",
      "        if type(i) is str and '\\n' in i:\n",
      "            string += lines_to_line(i).rstrip()+';'\n",
      "        else:\n",
      "            string += str(i)+';'\n",
      "    return string[:-1]\n",
      "\n",
      "def lines_to_line(lines):\n",
      "    lines = lines.split('\\n')\n",
      "    return (' ').join(lines)\n",
      "\n",
      "def type_to_single_line_str(var):\n",
      "    if type(var) is str and '\\n' in var:\n",
      "        return lines_to_line(var)\n",
      "    elif type(var) is str or type(var) is int or type(var) is float:\n",
      "        return str(var)\n",
      "    elif type(var) is list and len(var) == 1:\n",
      "        return str(var[0])\n",
      "    elif type(var) is list and len(var) > 0:\n",
      "        return list_to_string(var)\n",
      "    else:\n",
      "        return var\n",
      "\n",
      "class Database:\n",
      "    \n",
      "    \n",
      "    def __init__(self, guide_list):\n",
      "        self.records = []\n",
      "        self.guide_list = guide_list\n",
      "        self.records_by_guide = {}\n",
      "    \n",
      "    def read_embl_genbank(self, input_filenames_list):\n",
      "        generators = []\n",
      "        for input_filename in input_filenames_list:\n",
      "            if is_embl_or_gb(input_filename):\n",
      "                generators.append(parse_input(input_filename))\n",
      "                print 'Accepted:', input_filename\n",
      "            else:\n",
      "                print ('Rejected: ' + input_filename + ' is ' + \n",
      "                       guess_format(input_filename) + ' . Should be genbank or embl.')\n",
      "            for generator in generators:\n",
      "                for record in generator:\n",
      "                    dwindled_record = dwindle_record(record, guide_list)\n",
      "                    if len(record.features) > 1:\n",
      "                        self.records.append(dwindled_record)\n",
      "                    elif len(record.features) == 1 and not record.features[0].type == 'source':\n",
      "                        self.records.append(dwindled_record)\n",
      "\n",
      "    def read_denovo(self, input_filename, feature_type, char_type):\n",
      "        count = 0\n",
      "        denovo = SeqIO.parse(input_filename, guess_format(input_filename))\n",
      "        for record in denovo:\n",
      "            feature = SeqFeature(FeatureLocation(0, len(record.seq)), type=feature_type, strand=1)\n",
      "            feature.qualifiers['denovo_id'] = [record.id]\n",
      "            feature.qualifiers['denovo_desc'] = [(' ').join(record.description.split()[1:])]\n",
      "            record.id = 'denovo'+str(count)\n",
      "            count += 1\n",
      "            feature.qualifiers['feature_id'] = [record.id+'_f0']\n",
      "            record.features = [feature]\n",
      "            if char_type == 'prot':\n",
      "                record.seq.alphabet = IUPAC.protein\n",
      "            elif char_type == 'dna':\n",
      "                record.seq.alphabet = IUPAC.ambiguous_dna\n",
      "            self.records.append(record)\n",
      "\n",
      "    def write(self, filename, format = 'genbank'):\n",
      "        if format == 'genbank' or format == 'embl':\n",
      "            SeqIO.write(self.records, filename, format)\n",
      "        elif format == 'csv':\n",
      "            # get titles for source and othe features\n",
      "            source_qualifiers = []\n",
      "            feature_qualifiers = []\n",
      "            for record in self.records:\n",
      "                for feature in record.features:\n",
      "                    for key in feature.qualifiers.keys():\n",
      "                        if feature.type == 'source' and not key in source_qualifiers:\n",
      "                            source_qualifiers.append(key)\n",
      "                        elif not feature.type == 'source' and not key in feature_qualifiers:\n",
      "                            feature_qualifiers.append(key)\n",
      "            \n",
      "            with open(filename, 'wb') as csvfile:\n",
      "                linewriter = csv.writer(csvfile, delimiter='\\t',\n",
      "                                        quotechar='|',\n",
      "                                        quoting=csv.QUOTE_MINIMAL)\n",
      "                linewriter.writerow(['record_id','seq']+source_qualifiers+['taxonomy']+feature_qualifiers)\n",
      "                for record in self.records:\n",
      "                    seq = ''\n",
      "                    if len(record.seq) <= 10:\n",
      "                        seq = str(record.seq)[0:11]\n",
      "                    else:\n",
      "                        seq = str(record.seq)[0:6] + '...' + str(record.seq)[-5:]\n",
      "                    \n",
      "                    line_start = [record.id, seq]\n",
      "                    source = None\n",
      "                    for feature in record.features:\n",
      "                        if feature.type == 'source':\n",
      "                            source = feature\n",
      "                    if not source == None:\n",
      "                        for qual in source_qualifiers:\n",
      "                            if qual in source.qualifiers.keys():\n",
      "                                line_start.append(type_to_single_line_str(source.qualifiers[qual]))\n",
      "                            else:\n",
      "                                line_start.append('null')\n",
      "                    elif source == None:\n",
      "                        for qual in source_qualifiers:\n",
      "                            line_start.append('null')\n",
      "                    if 'taxonomy' in record.annotations.keys():\n",
      "                        line_start.append(type_to_single_line_str(record.annotations['taxonomy']))\n",
      "                    else:\n",
      "                        line_start.append(['null'])\n",
      "                    for feature in record.features:\n",
      "                        if not feature.type == 'source':\n",
      "                            line = line_start\n",
      "                            for qual in feature_qualifiers:\n",
      "                                if qual in feature.qualifiers.keys() and qual == 'translation':\n",
      "                                    trans = feature.qualifiers[qual][0]\n",
      "                                    if len(trans)<=10:\n",
      "                                        line.append(trans)\n",
      "                                    else:\n",
      "                                        line.append(trans[:6] + '...' + trans[-5:])\n",
      "                                elif qual in feature.qualifiers.keys():\n",
      "                                    line.append(type_to_single_line_str(feature.qualifiers[qual]))\n",
      "                                else:\n",
      "                                    line.append('null')\n",
      "                            linewriter.writerow(line)\n",
      "                            \n",
      "\n",
      "                    \n",
      "    def if_this_then_that(self, IF_THIS, IN_THIS, THEN_THAT, IN_THAT, mode = 'whole'):\n",
      "        for record in self.records:\n",
      "            for feature in record.features:\n",
      "                if not feature.type == 'source':\n",
      "                    if IN_THIS in feature.qualifiers.keys():\n",
      "                        if not type(feature.qualifiers[IN_THIS]) is list:\n",
      "                            feature.qualifiers[IN_THIS] = [feature.qualifiers[IN_THIS]]\n",
      "                        for i in feature.qualifiers[IN_THIS]:\n",
      "                            if mode == 'whole':\n",
      "                                if i == IF_THIS:\n",
      "                                    feature.qualifiers[IN_THAT] = [THEN_THAT]\n",
      "                            elif mode == 'part':\n",
      "                                if IF_THIS in i:\n",
      "                                    feature.qualifiers[IN_THAT] = [THEN_THAT]\n",
      "    \n",
      "    \n",
      "    def add_qualifier(self, feature_ids, name, value):\n",
      "        if not type(value) is list:\n",
      "                    value = [value]\n",
      "        for record in self.records:\n",
      "            for feature in record.features:\n",
      "                if feature.qualifiers['feature_id'][0] in feature_ids:\n",
      "                    feature.qualifiers[name] = value\n",
      "\n",
      "    def add_qualifier_from_source(self, qualifier):\n",
      "        for record in self.records:\n",
      "            source = None\n",
      "            for feature in record.features:\n",
      "                if feature.type == 'source':\n",
      "                    source = feature\n",
      "            value = None\n",
      "            if not source == None:\n",
      "              if qualifier in source.qualifiers.keys():\n",
      "                    value = source.qualifiers[qualifier]       \n",
      "            if not value == None:\n",
      "                if not type(value) is list:\n",
      "                    value = [value]\n",
      "                for feature in record.features:\n",
      "                    if not feature.type == 'source':\n",
      "                        feature.qualifiers[qualifier] = value\n",
      "\n",
      "    def copy_paste_qual_value(self, from_qualifier, to_qualifier):\n",
      "        for record in self.records:\n",
      "            for feature in record.features:\n",
      "                if not feature.type == 'source':\n",
      "                    if from_qualifier in feature.qualifiers.keys():\n",
      "                        feature.qualifiers[to_qualifier] = feature.qualifiers[from_qualifier]\n",
      "\n",
      "\n",
      "    def extract_by_guide(self):\n",
      "        data_by_guide = {}\n",
      "        for guide in self.guide_list.content:\n",
      "            if not guide.name in guide.aliases:\n",
      "                guide.aliases.append(guide.name)\n",
      "            records = []\n",
      "            for record in self.records:\n",
      "                for feature in record.features:\n",
      "                    if (feature.type == guide.feature_type and \n",
      "                        (feature.qualifiers['gene'][0] in guide.aliases or\n",
      "                         feature.qaulifiers['product'][0] in guide.aliases)):\n",
      "                        if guide.char_type == 'dna':\n",
      "                            S = feature.extract(record.seq)\n",
      "                        elif guide.char_type == 'prot':\n",
      "                            S = Seq(feature.qualifiers['translation'][0], IUPAC.protein)\n",
      "                        feature_record = SeqRecord(seq = S, id = feature.qualifiers['feature_id'][0],\n",
      "                                                   description = '')\n",
      "                        records.append(feature_record)\n",
      "            data_by_guide[guide.name] = records\n",
      "        self.records_by_guide = data_by_guide\n",
      "\n",
      "\n",
      "    def write_by_guide(self, format = 'fasta'):\n",
      "        if self.records_by_guide == {}:\n",
      "            self.extract_by_guide\n",
      "        for key in self.records_by_guide.keys():\n",
      "            SeqIO.write(self.records_by_guide[key], key+'.'+format, format)\n",
      "            "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 83
    }
   ],
   "metadata": {}
  }
 ]
}