{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Build guides to describe the phylogenetic partitions"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# like this\n",
      "\n",
      "ssu = Guide('dna','rRNA','18S', ['18s','18S','SSU rRNA','18S ribosomal RNA'])\n",
      "\n",
      "# or like this\n",
      "\n",
      "coi = Guide()\n",
      "coi.name = 'coi'\n",
      "coi.char_type = 'dna'\n",
      "coi.feature_type = 'CDS'\n",
      "coi.aliases = ('cox1','COX1','coi','COI','CoI')\n",
      "\n",
      "# parsing this info from csv file to come.\n",
      "\n",
      "guide_list = GuideList()\n",
      "\n",
      "for guide in (ssu, coi):\n",
      "    guide_list.add_guide_to_list(guide)\n",
      "    \n",
      "# guide_list is a class and not just a list because it only accepts Guide objects\n",
      "\n",
      "print str(guide_list)\n",
      "\n",
      "for guide in guide_list.content:\n",
      "    print str(guide)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "GuideList(18S, coi)\n",
        "Guide(char_type=dna, feature_type=rRNA, name=18S, aliases=18s; 18S; SSU rRNA; 18S ribosomal RNA)\n",
        "Guide(char_type=dna, feature_type=CDS, name=coi, aliases=cox1; COX1; coi; COI; CoI)\n"
       ]
      }
     ],
     "prompt_number": 40
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "    \n",
      "## Build sequence + metadata database"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "db = Database(guide_list)\n",
      "\n",
      "# Read genbank files into the db\n",
      "\n",
      "data_folder = '/home/amir/Dropbox/ReproPhylo/data/'\n",
      "\n",
      "input_filenames_list = (data_folder + 'primates.gb',\n",
      "                        data_folder + 'primates1.gb')\n",
      "\n",
      "db.read_embl_genbank(input_filenames_list) # also removes all the features that are not requested by guides\n",
      "\n",
      "# Add denovo sequences\n",
      "\n",
      "db.read_denovo('data/denovo.fas', 'CDS', 'dna')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Accepted: /home/amir/Dropbox/ReproPhylo/data/primates.gb\n",
        "Accepted:"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " /home/amir/Dropbox/ReproPhylo/data/primates1.gb\n"
       ]
      }
     ],
     "prompt_number": 41
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Edit the metadata"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Add spam qualifier to features by id\n",
      "\n",
      "db.add_qualifier(['KF830702.1_f0','AB572419.1_f0'], 'spam', 'totally')\n",
      "\n",
      "# Add a gene qualifier to the 18S entries based on their product qualifier\n",
      "\n",
      "db.if_this_then_that('18S ribosomal RNA', 'product', 'SSU rRNA', 'gene')\n",
      "\n",
      "# To each of the features, add the species from the source feature\n",
      "\n",
      "db.add_qualifier_from_source('organism')\n",
      "\n",
      "# For the denovo sequences, add the organism from the description\n",
      "\n",
      "db.copy_paste_from_features_to_source('original_desc','organism')\n",
      "\n",
      "# add gene qualifier to the denovo sequence so that they can be read by the guide\n",
      "\n",
      "db.add_qualifier(['denovo0_f0','denovo1_f0'], 'gene', 'coi')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 42
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Write records to view (cvs) and retain the db edits for future use (gb)"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "db.write('test.csv', format = 'csv')  # test.csv\n",
      "db.write('test.gb', format = 'genbank') # test.gb"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 43
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Write records per partition "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Put a dictionary with per partition records in the db\n",
      "db.extract_by_guide() # will be done by 'write_by_guide()' automatically if not performed\n",
      "\n",
      "db.write_by_guide() # written to files named guide_name.chosen_format\n",
      "                    # coi.fasta,  18S.fasta\n",
      "                    # writes fasta by default but can be given any supported format as value\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 44
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Tetillidae Data\n",
      "---------------"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ssu = Guide('dna','rRNA','18S', ['18s','18S','SSU rRNA','18S ribosomal RNA'])\n",
      "lsu = Guide('dna','rRNA','28S', ['28s','28S','SSU rRNA','28S ribosomal RNA'])\n",
      "coi = Guide('dna','CDS','coi', ['cox1','COX1','coi','COI','CoI'])\n",
      "\n",
      "guide_list = GuideList()\n",
      "\n",
      "for guide in (ssu, lsu, coi):\n",
      "    guide_list.add_guide_to_list(guide)\n",
      "\n",
      "db = Database(guide_list)\n",
      "\n",
      "data_folder = '/media/amir/DATA/work/Dropbox/ReproPhylo/data/'\n",
      "\n",
      "input_filenames_list = [data_folder + 'Tetillidae.gb']\n",
      "\n",
      "db.read_embl_genbank(input_filenames_list)\n",
      "\n",
      "db.write('Tet_Repro.csv', format = 'csv')\n",
      "db.write('Tet_Repro.gb', format = 'genbank')\n",
      "\n",
      "db.species_vs_guides('Tetillidae_genes_per_species.csv')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Accepted: /media/amir/DATA/work/Dropbox/ReproPhylo/data/Tetillidae.gb\n"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "#Classes\n",
      "\n",
      "# Guide(), GuideList() and Database()"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%file reprophylo.py\n",
      "\n",
      "from Bio import SeqIO\n",
      "import os, csv, sys\n",
      "from Bio.Seq import Seq\n",
      "from Bio.Alphabet import IUPAC\n",
      "from Bio.SeqRecord import SeqRecord\n",
      "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
      "\n",
      "class Guide:\n",
      "       \n",
      "    char_type = 'NotSet'\n",
      "    feature_type = 'NotSet'\n",
      "    name = 'NotSet'\n",
      "    aliases = []\n",
      "    \n",
      "    def __init__(self, char_type=char_type, feature_type=feature_type,\n",
      "                 name=name, aliases=aliases):\n",
      "        \"\"\" Use to tell ReproPhylo which partition to build from a mixed (or not) data.\n",
      "        \n",
      "        >>> guide = Guide('dna', 'CDS', 'coi', ['cox1','COX1','coi','COI','CoI'])\n",
      "        >>> print guide\n",
      "        Guide(char_type=dna, feature_type=CDS, name=coi, aliases=cox1; COX1; coi; COI; CoI)\"\"\"\n",
      "\n",
      "        \n",
      "        self.char_type = char_type\n",
      "        self.feature_type = feature_type\n",
      "        self.name = name\n",
      "        self.aliases = aliases\n",
      "        \n",
      "        valid = ['dna','prot']\n",
      "        if not self.char_type in valid:\n",
      "            raise ValueError('self.char_type should be \\'dna\\' or \\'prot\\'')\n",
      "        if not type(self.feature_type) is str:\n",
      "            raise ValueError('self.feature_type should be a string')\n",
      "        if not type(self.name) is str:\n",
      "            raise ValueError('self.name should be a string')\n",
      "        if not type(self.aliases) is list:\n",
      "            raise ValueError('self.aliases should be a list')\n",
      "        else:\n",
      "            for a in self.aliases:\n",
      "                if not type(a) is str:\n",
      "                    raise ValueError('aliases in self.aliases have to be strings')\n",
      "            \n",
      "\n",
      "    \n",
      "    def __str__(self):\n",
      "        aliases_str = ('; ').join(self.aliases)\n",
      "        return ('Guide(char_type='+self.char_type+', feature_type='+self.feature_type+\n",
      "                ', name='+self.name+', aliases='+aliases_str+')')\n",
      "            \n",
      "                \n",
      "\n",
      "if __name__ == \"__main__\":\n",
      "    import doctest\n",
      "    doctest.testmod() \n",
      "        \n",
      "class GuideList():\n",
      "    \n",
      "    def __init__(self):\n",
      "        self.content = []\n",
      "        \n",
      "        \"\"\" A list which accepts only Guide() objects.\n",
      "        \n",
      "        >>> guide = Guide('dna', 'CDS', 'coi', ['cox1','COX1','coi','COI','CoI'])\n",
      "        >>> guide_list = GuideList()\n",
      "        >>> guide_list.add_guide_to_list(guide)\n",
      "        >>> print guide_list.content[0].aliases[2]\n",
      "        coi\"\"\"\n",
      "        \n",
      "\n",
      "        \n",
      "    def add_guide_to_list(self, guide):\n",
      "        \n",
      "        \"\"\" adds a guide to a GuideList only if it is a Guide\n",
      "        \n",
      "        >>> guide_list = GuideList()\n",
      "        >>> guide_list.add_guide_to_list('not a guide')\n",
      "        Traceback (most recent call last):\n",
      "        ...\n",
      "        TypeError: Expecting Guide() object\"\"\"\n",
      "    \n",
      "        names = []\n",
      "        for g in self.content:\n",
      "            names.append(g.name)\n",
      "        if isinstance(guide, Guide):\n",
      "            if not guide.name in names:\n",
      "                self.content.append(guide)\n",
      "            else: \n",
      "                raise NameError('Guide ' + guide.name + ' is already included')\n",
      "        else:\n",
      "            raise TypeError(\"Expecting Guide() object\") \n",
      "    \n",
      "    def __str__(self):\n",
      "        guide_list_str = 'GuideList('\n",
      "        for guide in self.content:\n",
      "            guide_list_str += guide.name+', '\n",
      "        return guide_list_str[:-2]+')' \n",
      "            \n",
      "if __name__ == \"__main__\":\n",
      "    import doctest\n",
      "    doctest.testmod() \n",
      "\n",
      "# tools to prepare the db\n",
      "    \n",
      "def keep_feature(feature, guide_list):\n",
      "    \n",
      "    \"\"\" Returns true if a feature's type is in one of the guides and if the gene\n",
      "    or product qualifiers is in the aliases of one of the guides\n",
      "    \n",
      "    >>> coi = Guide('dna','CDS','coi', ['cox1','COX1','coi','COI','CoI'])\n",
      "    >>> guide_list = GuideList()\n",
      "    >>> guide_list.add_guide_to_list(coi)\n",
      "    >>> location = FeatureLocation(1,100)\n",
      "    >>> feature = SeqFeature()\n",
      "    >>> feature.location = location\n",
      "    >>> feature.type = 'CDS'\n",
      "    >>> feature.qualifiers['gene'] = ['CoI']\n",
      "    >>> a = keep_feature(feature, guide_list)\n",
      "    >>> print a\n",
      "    True\"\"\"\n",
      "    \n",
      "    keep = 0\n",
      "    for g in guide_list.content:\n",
      "        if not g.name in g.aliases:\n",
      "            g.aliases.append(g.name)\n",
      "        if feature.type == 'source':\n",
      "            keep = 1\n",
      "        elif feature.type == g.feature_type:\n",
      "            qaul = False\n",
      "            if 'gene' in feature.qualifiers.keys():\n",
      "                qual = 'gene'\n",
      "            elif 'product' in feature.qualifiers.keys():\n",
      "                qual = 'product'\n",
      "            if not qual == False and feature.qualifiers[qual][0] in g.aliases:\n",
      "                keep = 1\n",
      "    if keep == 1:\n",
      "        return True\n",
      "    else:\n",
      "        return False\n",
      "    \n",
      "def dwindle_record(record, guide_list):\n",
      "    dwindled_features = []\n",
      "    feature_count = 0\n",
      "    for feature in record.features:\n",
      "        if keep_feature(feature, guide_list)== True:\n",
      "            if feature.type == 'source' and not 'feature_id' in feature.qualifiers.keys():\n",
      "                feature.qualifiers['feature_id'] = [record.id + '_source']\n",
      "            elif not 'feature_id' in feature.qualifiers.keys():\n",
      "                feature.qualifiers['feature_id'] = [record.id + '_f' + str(feature_count)]\n",
      "                feature_count += 1                                        \n",
      "            dwindled_features.append(feature)\n",
      "    record.features = dwindled_features\n",
      "    return record\n",
      "            \n",
      "            \n",
      "def guess_format(input_filename):\n",
      "    \n",
      "    \"\"\" Runs a perl script that returns the input's format\n",
      "    \n",
      "    >>> filename = 'data/22_rotifer_COI_seqs.fas'\n",
      "    >>> print guess_format(filename)\n",
      "    fasta\"\"\"\n",
      "    \n",
      "    if os.path.exists(input_filename):\n",
      "        os.system('perl guesser.pl ' + input_filename)\n",
      "        guess = open(input_filename + '.format','r').read();\n",
      "        os.remove(input_filename + '.format')\n",
      "        return guess\n",
      "    else:\n",
      "        sys.exit('Cannot guess ' + input_filename + '. Does it exist?')\n",
      "        \n",
      "def is_embl_or_gb(input_filename):\n",
      "    input_format = guess_format(input_filename)\n",
      "    if input_format == 'embl' or input_format == 'genbank':\n",
      "        return True\n",
      "    else:\n",
      "        return False\n",
      "\n",
      "def parse_input(input_filename):\n",
      "    return SeqIO.parse(input_filename, guess_format(input_filename))\n",
      "\n",
      "\n",
      "def list_to_string(List):\n",
      "    string = ''\n",
      "    for i in List:\n",
      "        if type(i) is str and '\\n' in i:\n",
      "            string += lines_to_line(i).rstrip()+';'\n",
      "        else:\n",
      "            string += str(i)+';'\n",
      "    return string[:-1]\n",
      "\n",
      "def lines_to_line(lines):\n",
      "    lines = lines.split('\\n')\n",
      "    return (' ').join(lines)\n",
      "\n",
      "def type_to_single_line_str(var):\n",
      "    if type(var) is str and '\\n' in var:\n",
      "        return lines_to_line(var)\n",
      "    elif type(var) is str or type(var) is int or type(var) is float:\n",
      "        return str(var)\n",
      "    elif type(var) is list and len(var) == 1:\n",
      "        return str(var[0])\n",
      "    elif type(var) is list and len(var) > 0:\n",
      "        return list_to_string(var)\n",
      "    else:\n",
      "        return var\n",
      "\n",
      "\n",
      "\n",
      "class Database:\n",
      "    \n",
      "    \n",
      "    def __init__(self, guide_list):\n",
      "        self.records = []\n",
      "        self.guide_list = guide_list\n",
      "        self.records_by_guide = {}\n",
      "        self.sets = []\n",
      "        self.alignments = {}\n",
      "        self.trees = {}\n",
      "    \n",
      "    def read_embl_genbank(self, input_filenames_list):\n",
      "        generators = []\n",
      "        for input_filename in input_filenames_list:\n",
      "            if is_embl_or_gb(input_filename):\n",
      "                generators.append(parse_input(input_filename))\n",
      "                print 'Accepted:', input_filename\n",
      "            else:\n",
      "                print ('Rejected: ' + input_filename + ' is ' + \n",
      "                       guess_format(input_filename) + ' . Should be genbank or embl.')\n",
      "            for generator in generators:\n",
      "                for record in generator:\n",
      "                    dwindled_record = dwindle_record(record, guide_list)\n",
      "                    if len(record.features) > 1:\n",
      "                        self.records.append(dwindled_record)\n",
      "                    elif len(record.features) == 1 and not record.features[0].type == 'source':\n",
      "                        self.records.append(dwindled_record)\n",
      "\n",
      "    def read_denovo(self, input_filename, feature_type, char_type, source_qualifiers = {}):\n",
      "        count = 0\n",
      "        denovo = SeqIO.parse(input_filename, guess_format(input_filename))\n",
      "        for record in denovo:\n",
      "            feature = SeqFeature(FeatureLocation(0, len(record.seq)), type=feature_type, strand=1)\n",
      "            source = SeqFeature(FeatureLocation(0, len(record.seq)), type='source', strand=1)\n",
      "            if len(source_qualifiers.keys())>0:\n",
      "                for key in source_qualifiers.keys():\n",
      "                    source.qualifiers[key] = source_qualifiers[key]\n",
      "            feature.qualifiers['original_id'] = [record.id]\n",
      "            feature.qualifiers['original_desc'] = [(' ').join(record.description.split()[1:])]\n",
      "            record.id = 'denovo'+str(count)\n",
      "            count += 1\n",
      "            feature.qualifiers['feature_id'] = [record.id+'_f0']\n",
      "            source.qualifiers['feature_id'] = [record.id+'_source']\n",
      "            record.features = [source, feature]\n",
      "            if char_type == 'prot':\n",
      "                record.seq.alphabet = IUPAC.protein\n",
      "            elif char_type == 'dna':\n",
      "                record.seq.alphabet = IUPAC.ambiguous_dna\n",
      "            self.records.append(record)\n",
      "\n",
      "    def write(self, filename, format = 'genbank'):\n",
      "        if format == 'genbank' or format == 'embl':\n",
      "            SeqIO.write(self.records, filename, format)\n",
      "        elif format == 'csv':\n",
      "            # get titles for source and othe features\n",
      "            source_qualifiers = []\n",
      "            feature_qualifiers = []\n",
      "            for record in self.records:\n",
      "                for feature in record.features:\n",
      "                    for key in feature.qualifiers.keys():\n",
      "                        if feature.type == 'source' and not key in source_qualifiers:\n",
      "                            source_qualifiers.append(key)\n",
      "                        elif not feature.type == 'source' and not key in feature_qualifiers:\n",
      "                            feature_qualifiers.append(key)\n",
      "            \n",
      "            with open(filename, 'wb') as csvfile:\n",
      "                linewriter = csv.writer(csvfile, delimiter='\\t',\n",
      "                                        quotechar='|',\n",
      "                                        quoting=csv.QUOTE_MINIMAL)\n",
      "                linewriter.writerow(['record_id','seq']+source_qualifiers+['taxonomy']+feature_qualifiers)\n",
      "                for record in self.records:\n",
      "                    seq = ''\n",
      "                    if len(record.seq) <= 10:\n",
      "                        seq = str(record.seq)[0:11]\n",
      "                    else:\n",
      "                        seq = str(record.seq)[0:6] + '...' + str(record.seq)[-5:]\n",
      "                    \n",
      "                    line_start = [record.id, seq]\n",
      "                    source = None\n",
      "                    for feature in record.features:\n",
      "                        if feature.type == 'source':\n",
      "                            source = feature\n",
      "                    if not source == None:\n",
      "                        for qual in source_qualifiers:\n",
      "                            if qual in source.qualifiers.keys():\n",
      "                                line_start.append(type_to_single_line_str(source.qualifiers[qual]))\n",
      "                            else:\n",
      "                                line_start.append('null')\n",
      "                    elif source == None:\n",
      "                        for qual in source_qualifiers:\n",
      "                            line_start.append('null')\n",
      "                    if 'taxonomy' in record.annotations.keys():\n",
      "                        line_start.append(type_to_single_line_str(record.annotations['taxonomy']))\n",
      "                    else:\n",
      "                        line_start.append(['null'])\n",
      "                    for feature in record.features:\n",
      "                        if not feature.type == 'source':\n",
      "                            line = line_start\n",
      "                            for qual in feature_qualifiers:\n",
      "                                if qual in feature.qualifiers.keys() and qual == 'translation':\n",
      "                                    trans = feature.qualifiers[qual][0]\n",
      "                                    if len(trans)<=10:\n",
      "                                        line.append(trans)\n",
      "                                    else:\n",
      "                                        line.append(trans[:6] + '...' + trans[-5:])\n",
      "                                elif qual in feature.qualifiers.keys():\n",
      "                                    line.append(type_to_single_line_str(feature.qualifiers[qual]))\n",
      "                                else:\n",
      "                                    line.append('null')\n",
      "                            linewriter.writerow(line)\n",
      "                            \n",
      "\n",
      "                    \n",
      "    def if_this_then_that(self, IF_THIS, IN_THIS, THEN_THAT, IN_THAT, mode = 'whole'):\n",
      "        for record in self.records:\n",
      "            for feature in record.features:\n",
      "                if not feature.type == 'source':\n",
      "                    if IN_THIS in feature.qualifiers.keys():\n",
      "                        if not type(feature.qualifiers[IN_THIS]) is list:\n",
      "                            feature.qualifiers[IN_THIS] = [feature.qualifiers[IN_THIS]]\n",
      "                        for i in feature.qualifiers[IN_THIS]:\n",
      "                            if mode == 'whole':\n",
      "                                if i == IF_THIS:\n",
      "                                    feature.qualifiers[IN_THAT] = [THEN_THAT]\n",
      "                            elif mode == 'part':\n",
      "                                if IF_THIS in i:\n",
      "                                    feature.qualifiers[IN_THAT] = [THEN_THAT]\n",
      "    \n",
      "    \n",
      "    def add_qualifier(self, feature_ids, name, value):\n",
      "        if not type(value) is list:\n",
      "                    value = [value]\n",
      "        for record in self.records:\n",
      "            for feature in record.features:\n",
      "                if feature.qualifiers['feature_id'][0] in feature_ids:\n",
      "                    feature.qualifiers[name] = value\n",
      "\n",
      "    def add_qualifier_from_source(self, qualifier):\n",
      "        for record in self.records:\n",
      "            source = None\n",
      "            for feature in record.features:\n",
      "                if feature.type == 'source':\n",
      "                    source = feature\n",
      "            value = None\n",
      "            if not source == None:\n",
      "              if qualifier in source.qualifiers.keys():\n",
      "                    value = source.qualifiers[qualifier]       \n",
      "            if not value == None:\n",
      "                if not type(value) is list:\n",
      "                    value = [value]\n",
      "                for feature in record.features:\n",
      "                    if not feature.type == 'source':\n",
      "                        feature.qualifiers[qualifier] = value\n",
      "    \n",
      "\n",
      "    def copy_paste_within_feature(self, from_qualifier, to_qualifier):\n",
      "        for record in self.records:\n",
      "            for feature in record.features:\n",
      "                if not feature.type == 'source':\n",
      "                    if from_qualifier in feature.qualifiers.keys():\n",
      "                        feature.qualifiers[to_qualifier] = feature.qualifiers[from_qualifier]\n",
      "                        \n",
      "    def copy_paste_from_features_to_source(self, from_feature_qual, to_source_qual):\n",
      "        for record in self.records:\n",
      "            source = None\n",
      "            values_from_features = []\n",
      "            for feature in record.features:\n",
      "                if not feature.type == 'source':\n",
      "                    if from_feature_qual in feature.qualifiers.keys():\n",
      "                        if not feature.qualifiers[from_feature_qual] in values_from_features:\n",
      "                            values_from_features += feature.qualifiers[from_feature_qual]\n",
      "                else:\n",
      "                    source = feature\n",
      "            if source == None:\n",
      "                source = SeqFeature(FeatureLocation(0, len(record.seq)), type='source', strand=1)\n",
      "                source.qualifiers['feature_id'] = record.id + '_source'\n",
      "                record.features = [source] + record.features\n",
      "            if not to_source_qual in source.qualifiers.keys():\n",
      "                source.qualifiers[to_source_qual] = values_from_features\n",
      "                \n",
      "    def species_vs_guides(self, outfile_name):\n",
      "        species_vs_guides = {}\n",
      "        for record in self.records:\n",
      "            organism = 'undef'\n",
      "            for feature in record.features:\n",
      "                if feature.type == 'source':\n",
      "                    if 'organism' in feature.qualifiers.keys():\n",
      "                        organism = feature.qualifiers['organism'][0]\n",
      "            if not organism in species_vs_guides.keys():\n",
      "                species_vs_guides[organism] = {}    \n",
      "            for feature in record.features:\n",
      "                if not feature.type == 'source':\n",
      "                    for guide in self.guide_list.content:\n",
      "                        if not guide.name in guide.aliases:\n",
      "                            guide.aliases.append(guide.name)\n",
      "                        if 'gene' in feature.qualifiers.keys():\n",
      "                            if feature.qualifiers['gene'][0] in guide.aliases:\n",
      "                                if not guide.name in species_vs_guides[organism].keys():\n",
      "                                    species_vs_guides[organism][guide.name] = 1\n",
      "                                else:\n",
      "                                    species_vs_guides[organism][guide.name] += 1\n",
      "                        elif 'product' in feature.qualifiers.keys():\n",
      "                            if feature.qualifiers['product'][0] in guide.aliases:\n",
      "                                if not guide.name in species_vs_guides[organism].keys():\n",
      "                                    species_vs_guides[organism][guide.name] = 1\n",
      "                                else:\n",
      "                                    species_vs_guides[organism][guide.name] += 1\n",
      "        with open(outfile_name, 'wb') as csvfile:\n",
      "            linewriter = csv.writer(csvfile, delimiter='\\t',\n",
      "                                    quotechar='|',\n",
      "                                    quoting=csv.QUOTE_MINIMAL)\n",
      "            guide_names = []\n",
      "            for g in self.guide_list.content:\n",
      "                guide_names.append(g.name)\n",
      "            linewriter.writerow(['species']+guide_names)\n",
      "            for organism in species_vs_guides.keys():\n",
      "                line = [organism]\n",
      "                for name in guide_names:\n",
      "                    if name in species_vs_guides[organism].keys():\n",
      "                        line.append(str(species_vs_guides[organism][name]))\n",
      "                    else:\n",
      "                        line.append('0')\n",
      "                linewriter.writerow(line)\n",
      "                    \n",
      "            \n",
      "            \n",
      "    def extract_by_guide(self):\n",
      "        data_by_guide = {}\n",
      "        for guide in self.guide_list.content:\n",
      "            if not guide.name in guide.aliases:\n",
      "                guide.aliases.append(guide.name)\n",
      "            records = []\n",
      "            for record in self.records:\n",
      "                for feature in record.features:\n",
      "                    if (feature.type == guide.feature_type and \n",
      "                        (feature.qualifiers['gene'][0] in guide.aliases or\n",
      "                         feature.qaulifiers['product'][0] in guide.aliases)):\n",
      "                        if guide.char_type == 'dna':\n",
      "                            S = feature.extract(record.seq)\n",
      "                        elif guide.char_type == 'prot':\n",
      "                            S = Seq(feature.qualifiers['translation'][0], IUPAC.protein)\n",
      "                        feature_record = SeqRecord(seq = S, id = feature.qualifiers['feature_id'][0],\n",
      "                                                   description = '')\n",
      "                        records.append(feature_record)\n",
      "            data_by_guide[guide.name] = records\n",
      "        self.records_by_guide = data_by_guide\n",
      "\n",
      "\n",
      "    def write_by_guide(self, format = 'fasta'):\n",
      "        if self.records_by_guide == {}:\n",
      "            self.extract_by_guide\n",
      "        for key in self.records_by_guide.keys():\n",
      "            SeqIO.write(self.records_by_guide[key], key+'.'+format, format)\n",
      "            \n",
      "if __name__ == \"__main__\":\n",
      "    import doctest\n",
      "    doctest.testmod()   \n",
      "          "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Overwriting reprophylo.py\n"
       ]
      }
     ],
     "prompt_number": 24
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "!python -m doctest -v  reprophylo.py"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Trying:\r\n",
        "    guide = Guide('dna', 'CDS', 'coi', ['cox1','COX1','coi','COI','CoI'])\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    print guide\r\n",
        "Expecting:\r\n",
        "    Guide(char_type=dna, feature_type=CDS, name=coi, aliases=cox1; COX1; coi; COI; CoI)\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    guide_list = GuideList()\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    guide_list.add_guide_to_list('not a guide')\r\n",
        "Expecting:\r\n",
        "    Traceback (most recent call last):\r\n",
        "    ...\r\n",
        "    TypeError: Expecting Guide() object\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    filename = 'data/22_rotifer_COI_seqs.fas'\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    print guess_format(filename)\r\n",
        "Expecting:\r\n",
        "    fasta\r\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "ok\r\n",
        "Trying:\r\n",
        "    coi = Guide('dna','CDS','coi', ['cox1','COX1','coi','COI','CoI'])\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    guide_list = GuideList()\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    guide_list.add_guide_to_list(coi)\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    location = FeatureLocation(1,100)\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    feature = SeqFeature()\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    feature.location = location\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    feature.type = 'CDS'\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    feature.qualifiers['gene'] = ['CoI']\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    a = keep_feature(feature, guide_list)\r\n",
        "Expecting nothing\r\n",
        "ok\r\n",
        "Trying:\r\n",
        "    print a\r\n",
        "Expecting:\r\n",
        "    True\r\n",
        "ok\r\n",
        "25 items had no tests:\r\n",
        "    reprophylo\r\n",
        "    reprophylo.Database\r\n",
        "    reprophylo.Database.__init__\r\n",
        "    reprophylo.Database.add_qualifier\r\n",
        "    reprophylo.Database.add_qualifier_from_source\r\n",
        "    reprophylo.Database.copy_paste_from_features_to_source\r\n",
        "    reprophylo.Database.copy_paste_within_feature\r\n",
        "    reprophylo.Database.extract_by_guide\r\n",
        "    reprophylo.Database.if_this_then_that\r\n",
        "    reprophylo.Database.read_denovo\r\n",
        "    reprophylo.Database.read_embl_genbank\r\n",
        "    reprophylo.Database.species_vs_guides\r\n",
        "    reprophylo.Database.write\r\n",
        "    reprophylo.Database.write_by_guide\r\n",
        "    reprophylo.Guide\r\n",
        "    reprophylo.Guide.__str__\r\n",
        "    reprophylo.GuideList\r\n",
        "    reprophylo.GuideList.__init__\r\n",
        "    reprophylo.GuideList.__str__\r\n",
        "    reprophylo.dwindle_record\r\n",
        "    reprophylo.is_embl_or_gb\r\n",
        "    reprophylo.lines_to_line\r\n",
        "    reprophylo.list_to_string\r\n",
        "    reprophylo.parse_input\r\n",
        "    reprophylo.type_to_single_line_str\r\n",
        "4 items passed all tests:\r\n",
        "   2 tests in reprophylo.Guide.__init__\r\n",
        "   2 tests in reprophylo.GuideList.add_guide_to_list\r\n",
        "   2 tests in reprophylo.guess_format\r\n",
        "  10 tests in reprophylo.keep_feature\r\n",
        "16 tests in 29 items.\r\n",
        "16 passed and 0 failed.\r\n",
        "Test passed.\r\n"
       ]
      }
     ],
     "prompt_number": 25
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/python\n",
      "# create html table from csv\n",
      "# Author(s): Chris Trombley <ctroms@gmail.com>\n",
      "# Version 2 - added css class to all columns except header\n",
      "\n",
      "#!/usr/bin/python\n",
      "# create html table from csv\n",
      "\n",
      "import sys\n",
      "import csv\n",
      "\n",
      "\n",
      "# Open the CSV file for reading\n",
      "reader = csv.reader(open('Tetillidae_genes_per_species.csv'), delimiter='\\t')\n",
      "\n",
      "# Create the HTML file for output\n",
      "htmlfile = open('Tetillidae_genes_per_species.csv.html',\"w\")\n",
      "\n",
      "# initialize rownum variable\n",
      "rownum = 0\n",
      "\n",
      "# write <table> tag\n",
      "htmlfile.write('<table>')\n",
      "\n",
      "# generate table contents\n",
      "for row in reader: # Read a single row from the CSV file\n",
      "\n",
      "\t# write header row. assumes first row in csv contains header\n",
      "\tif rownum == 0:\n",
      "\t\thtmlfile.write('<tr>') # write <tr> tag\n",
      "  \t\tfor column in row:\n",
      "  \t\t\thtmlfile.write('<th>' + column + '</th>')\n",
      "  \t\thtmlfile.write('</tr>')\n",
      "\n",
      "  \t#write all other rows\t\n",
      "  \telse:\n",
      "  \t\thtmlfile.write('<tr>')\t\n",
      "  \t\tfor column in row:\n",
      "  \t\t\thtmlfile.write('<td>' + column + '</td>')\n",
      "  \t\thtmlfile.write('</tr>')\n",
      "\t\n",
      "\t#increment row count\t\n",
      "\trownum += 1\n",
      "\n",
      "# write </table> tag\n",
      "htmlfile.write('</table>')\n",
      "\n",
      "# print results to shell\n",
      "print \"Created \" + str(rownum) + \" row table.\"\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Created 41 row table.\n"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}