


% Header, overrides base

    % Make sure that the sphinx doc style knows who it inherits from.
    \def\sphinxdocclass{article}

    % Declare the document class
    \documentclass[letterpaper,10pt,english]{/usr/share/sphinx/texinputs/sphinxhowto}

    % Imports
    \usepackage[utf8]{inputenc}
    \DeclareUnicodeCharacter{00A0}{\\nobreakspace}
    \usepackage[T1]{fontenc}
    \usepackage{babel}
    \usepackage{times}
    \usepackage{import}
    \usepackage[Bjarne]{/usr/share/sphinx/texinputs/fncychap}
    \usepackage{longtable}
    \usepackage{/usr/share/sphinx/texinputs/sphinx}
    \usepackage{multirow}

    \usepackage{amsmath}
    \usepackage{amssymb}
    \usepackage{ucs}
    \usepackage{enumerate}

    % Used to make the Input/Output rules follow around the contents.
    \usepackage{needspace}

    % Pygments requirements
    \usepackage{fancyvrb}
    \usepackage{color}
    % ansi colors additions
    \definecolor{darkgreen}{rgb}{.12,.54,.11}
    \definecolor{lightgray}{gray}{.95}
    \definecolor{brown}{rgb}{0.54,0.27,0.07}
    \definecolor{purple}{rgb}{0.5,0.0,0.5}
    \definecolor{darkgray}{gray}{0.25}
    \definecolor{lightred}{rgb}{1.0,0.39,0.28}
    \definecolor{lightgreen}{rgb}{0.48,0.99,0.0}
    \definecolor{lightblue}{rgb}{0.53,0.81,0.92}
    \definecolor{lightpurple}{rgb}{0.87,0.63,0.87}
    \definecolor{lightcyan}{rgb}{0.5,1.0,0.83}

    % Needed to box output/input
    \usepackage{tikz}
        \usetikzlibrary{calc,arrows,shadows}
    \usepackage[framemethod=tikz]{mdframed}

    \usepackage{alltt}

    % Used to load and display graphics
    \usepackage{graphicx}
    \graphicspath{ {figs/} }
    \usepackage[Export]{adjustbox} % To resize

    % used so that images for notebooks which have spaces in the name can still be included
    \usepackage{grffile}


    % For formatting output while also word wrapping.
    \usepackage{listings}
    \lstset{breaklines=true}
    \lstset{basicstyle=\small\ttfamily}
    \def\smaller{\fontsize{9.5pt}{9.5pt}\selectfont}

    %Pygments definitions
    
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\expandafter\def\csname PY@tok@gd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@gu\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PY@tok@gs\endcsname{\let\PY@bf=\textbf}
\expandafter\def\csname PY@tok@gr\endcsname{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@cm\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@vg\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@m\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@go\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\expandafter\def\csname PY@tok@ge\endcsname{\let\PY@it=\textit}
\expandafter\def\csname PY@tok@vc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@il\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@cs\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\expandafter\def\csname PY@tok@gi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@gh\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@ni\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\expandafter\def\csname PY@tok@nl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@nn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@no\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@na\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\expandafter\def\csname PY@tok@nb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@ne\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\expandafter\def\csname PY@tok@nf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@si\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@s2\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@vi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@nt\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nv\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@s1\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sx\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@bp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@c1\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@kc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@c\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@err\endcsname{\def\PY@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PY@tok@kd\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@ss\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sr\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@mo\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@kn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@gp\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@kr\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@s\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@kp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@w\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PY@tok@kt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@sb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@k\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@se\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\expandafter\def\csname PY@tok@sd\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    %Set pygments styles if needed...
    
        \definecolor{nbframe-border}{rgb}{0.867,0.867,0.867}
        \definecolor{nbframe-bg}{rgb}{0.969,0.969,0.969}
        \definecolor{nbframe-in-prompt}{rgb}{0.0,0.0,0.502}
        \definecolor{nbframe-out-prompt}{rgb}{0.545,0.0,0.0}

        \newenvironment{ColorVerbatim}
        {\begin{mdframed}[%
            roundcorner=1.0pt, %
            backgroundcolor=nbframe-bg, %
            userdefinedwidth=1\linewidth, %
            leftmargin=0.1\linewidth, %
            innerleftmargin=0pt, %
            innerrightmargin=0pt, %
            linecolor=nbframe-border, %
            linewidth=1pt, %
            usetwoside=false, %
            everyline=true, %
            innerlinewidth=3pt, %
            innerlinecolor=nbframe-bg, %
            middlelinewidth=1pt, %
            middlelinecolor=nbframe-bg, %
            outerlinewidth=0.5pt, %
            outerlinecolor=nbframe-border, %
            needspace=0pt
        ]}
        {\end{mdframed}}
        
        \newenvironment{InvisibleVerbatim}
        {\begin{mdframed}[leftmargin=0.1\linewidth,innerleftmargin=3pt,innerrightmargin=3pt, userdefinedwidth=1\linewidth, linewidth=0pt, linecolor=white, usetwoside=false]}
        {\end{mdframed}}

        \renewenvironment{Verbatim}[1][\unskip]
        {\begin{alltt}\smaller}
        {\end{alltt}}
    

    % Help prevent overflowing lines due to urls and other hard-to-break 
    % entities.  This doesn't catch everything...
    \sloppy

    % Document level variables
    \title{Tutorial}
    \date{December 4, 2014}
    \release{}
    \author{Unknown Author}
    \renewcommand{\releasename}{}

    % TODO: Add option for the user to specify a logo for his/her export.
    \newcommand{\sphinxlogo}{}

    % Make the index page of the document.
    \makeindex

    % Import sphinx document type specifics.
     


% Body

    % Start of the document
    \begin{document}

        
            \maketitle
        

        


        
        \section{ReproPhylo in an IPython notebook Docker
container}\label{reprophylo-in-an-ipython-notebook-docker-container}

\subsection{Working with ReproPhylo in IPython
notebook}\label{working-with-reprophylo-in-ipython-notebook}

This is an IPython notebook. It consists of text (markdown) cells like
this one, which contain comments and explanations, but do not affect the
program. Actual script is written in code cells, which have a shaded
background. The code in the code cells can be executed (run) by placing
the curser anywhere inside a code cell and clicking
\texttt{shift+enter}.

\textbf{Run the first code cell bellow.} It will upload ReproPhylo and
its dependencies. There is no output to expect, except for a number that
will appear, or change, in the square brackets on the left hand side of
the code cell.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}1{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{k+kn}{from} \PY{n+nn}{reprophylo} \PY{k+kn}{import} \PY{o}{*}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
\subsection{Version control in IPython
notebook}\label{version-control-in-ipython-notebook}

A version control program called git is incorperated in the ReproPhylo
code. This ensures that you will always be able to roll back to older
versions of files which might have been overwritten.\\\textbf{To get git
to start working at the background, run the next cell.}

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}2{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{start\PYZus{}git}\PY{p}{(}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}

[master (root-commit) 6f74b88] 2 script file(s) from Sun Nov 30
22:24:15 2014
 2 files changed, 4351 insertions(+)
 create mode 100644 .ipynb\_checkpoints/Tutorial-checkpoint.ipynb
 create mode 100644 Tutorial.ipynb

\end{alltt}

            \end{InvisibleVerbatim}
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}/home/amir/Dropbox/python\_modules/rpgit.py:68: UserWarning: Thanks to
Stack-Overflow users Shane Geiger and Billy Jin for the git wrappers
code
  warnings.warn('Thanks to Stack-Overflow users Shane Geiger and Billy
Jin for the git wrappers code')
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
The message received above was generated by git and it is letting us
know that a repositary has been created and that this IPython notebook
has been recorded in it, and will be tracked for any changes. Input data
files and python scripts, should you write any, will be controlled as
well.\section{Overview of this tutorial}\label{overview-of-this-tutorial}

In this tutorial we are going to do the following:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\itemsep1pt\parskip0pt\parsep0pt
\item
  Start the environment (you have already done this)
\item
  Examine lots of data from a taxonomic group in a single genbank file
  using ReproPhylo

  \begin{itemize}
  \itemsep1pt\parskip0pt\parsep0pt
  \item
    Identify which genes are found, and how often
  \item
    Identify synonyms in the gene names
  \item
    Select the genes to use in our analysis
  \end{itemize}
\item
  Start a ReproPhylo project to contain the data
\item
  Import the genbank loci you have selected above

  \begin{itemize}
  \itemsep1pt\parskip0pt\parsep0pt
  \item
    Explore which taxonomic groups are present
  \end{itemize}
\item
  Add any additional sequences

  \begin{itemize}
  \itemsep1pt\parskip0pt\parsep0pt
  \item
    You may have your own sequences to add to those in genbank
  \end{itemize}
\item
  Edit the metadata, either manually or programatically\\
\item
  Examine the sequence data statistics

  \begin{itemize}
  \itemsep1pt\parskip0pt\parsep0pt
  \item
    GC content
  \item
    sequence length distributions
  \item
    frequencies of ambiguous nucleotides
  \end{itemize}
\item
  Configure and run and trim a sequence alignment

  \begin{itemize}
  \itemsep1pt\parskip0pt\parsep0pt
  \item
    configure what and how to align
  \item
    align
  \item
    trim algnment columns based on criteria supplied to trimAL
  \end{itemize}
\item
  Configure and build a tree with RAxML
\item
  Annotate the tree using ETE
\item
  Use the output to design and run aconcatenated analysis
\item
  Archive the phylogenetic experiment

  \begin{itemize}
  \itemsep1pt\parskip0pt\parsep0pt
  \item
    archive data
  \item
    write experimental report to html file
  \end{itemize}
\end{enumerate}\subsection{Working with a GenBank
file}\label{working-with-a-genbank-file}

ReproPhylo is designed to easily handle genbank files. The first step
would be to make an online NCBI search in a the Nucleotide database.
Remember to save a record of your search term for reproducibility! Next
save the search results as a genbank file by using menu on the right
hand side of the Nucleotide webpage (see figure below).

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}3{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{c}{\PYZsh{}execute if figure is small}
\PY{k+kn}{from} \PY{n+nn}{IPython.display} \PY{k+kn}{import} \PY{n}{Image}
\PY{n}{Image}\PY{p}{(}\PY{n}{filename}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{Selection\PYZus{}003.png}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-out-prompt}Out\hspace{4pt}{[}3{]}:\hspace{4pt}}\\*
                \vspace{-2.55\baselineskip}\begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
    \begin{center}
    \includegraphics[max size={\textwidth}{\textheight}]{Tutorial_files/Tutorial_7_0.png}
    \par
    \end{center}
    
            \end{InvisibleVerbatim}
            
        
    
In most cases, entries in the NCBI Protein database are included in
their counterpart Nucleotide database entry under the
\texttt{translation} qualifier. ReproPhylo allows you to switch easily
between the datatypes, as long as you provide this sort of a
\textbf{Nucleotide} genbank file. For the ocassions in which a protein
family is not well represented in the Nucleotide database, it is also
possible to work with protein (or nucleotide) fasta files. For the
purpose of this tutorial, we have included a genbank file,
\texttt{Tetillidae.gb}. Below is the
\href{http://en.wikipedia.org/wiki/Tetillidae}{Tetillidae wikipedia
page} if you are interested what sort of beautiful creatures they are
(sponges).

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}4{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{k+kn}{from} \PY{n+nn}{IPython.display} \PY{k+kn}{import} \PY{n}{HTML}
\PY{n}{HTML}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{\PYZlt{}iframe src=http://en.wikipedia.org/wiki/Tetillidae?useformat=mobile width=900 height=500\PYZgt{}\PYZlt{}/iframe\PYZgt{}}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-out-prompt}Out\hspace{4pt}{[}4{]}:\hspace{4pt}}\\*
                \vspace{-2.55\baselineskip}\begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}<IPython.core.display.HTML at 0x7fe1f7f86190>\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
\#\# Exploring the locus content of a genbank file

The first step in ReproPhylo would be to list the genes present in a
genbank file and to quantify how many times each gene occurs. This is
done with the \texttt{list\_loci\_in\_genbnk()} function. It requires
you to specify (as `arguments') the genbank file name and an output
filename. The output file will be a comma separated values (CSV) file,
which can be read by ReproPhylo (details below). CSV files are text
representations of tables and we use the name loosly to also include tab
delimited text.\\The \texttt{list\_loci\_in\_genbnk()} function will
print the counts and names of the loci in alphabetical order and then by
descending counts to see which loci are most frequent. It will allow you
to choose which genes to carry forward in the analysis and also to check
if there are synonyms for any gene. This function is used as follows:\\

\textbf{Write all the loci from \texttt{Tetillidae.gb} to this notebook,
as well as a CSV output by running the next cell}

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}5{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
 \PY{n}{list\PYZus{}loci\PYZus{}in\PYZus{}genbank}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{Tetillidae.gb}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{Tetillidae\PYZus{}loci.csv}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}
There are 57 gene names (or gene product names) detected
----------------------------------
Gene and count sorted by gene name
----------------------------------
1 instances of CDS,ALD
2 instances of CDS,alg11
1 instances of CDS,ATP synthase beta subunit
3 instances of CDS,atp6
1 instances of CDS,ATP6
3 instances of CDS,atp8
1 instances of CDS,ATP8
1 instances of CDS,ATP9
2 instances of CDS,atp9
1 instances of CDS,catalase
1 instances of CDS,CchGa
1 instances of CDS,CchGb
1 instances of CDS,CchGc
1 instances of CDS,CchGd
1 instances of CDS,CchGe
1 instances of CDS,CchGf
2 instances of CDS,cob
1 instances of CDS,coi
11 instances of CDS,COI
78 instances of CDS,cox1
1 instances of CDS,COX1
3 instances of CDS,cox2
1 instances of CDS,COX2
3 instances of CDS,cox3
1 instances of CDS,COX3
3 instances of CDS,coxI
1 instances of CDS,CYTB
1 instances of CDS,elongation factor 1 alpha
1 instances of CDS,Hsp70
1 instances of CDS,MAT
2 instances of CDS,nad1
2 instances of CDS,nad2
2 instances of CDS,nad3
2 instances of CDS,nad4
2 instances of CDS,nad4L
2 instances of CDS,nad5
2 instances of CDS,nad6
1 instances of CDS,ND1
1 instances of CDS,ND2
1 instances of CDS,ND3
2 instances of CDS,ND4
1 instances of CDS,ND4L
1 instances of CDS,ND5
1 instances of CDS,ND6
7 instances of CDS,putative LAGLIDADG protein
1 instances of CDS,TPI
41 instances of rRNA,18S ribosomal RNA
2 instances of rRNA,18S rRNA
6 instances of rRNA,28S large subunit ribosomal RNA
42 instances of rRNA,28S ribosomal RNA
1 instances of rRNA,5.8S ribosomal RNA
1 instances of rRNA,5.8S rRNA
1 instances of rRNA,5S rRNA
3 instances of rRNA,rnl
3 instances of rRNA,rns
1 instances of rRNA,rrnL
9 instances of rRNA,small subunit 18S ribosomal RNA
-------------------------------
Gene and count sorted by counts
-------------------------------
78 instances of CDS,cox1
42 instances of rRNA,28S ribosomal RNA
41 instances of rRNA,18S ribosomal RNA
11 instances of CDS,COI
9 instances of rRNA,small subunit 18S ribosomal RNA
7 instances of CDS,putative LAGLIDADG protein
6 instances of rRNA,28S large subunit ribosomal RNA
3 instances of CDS,atp6
3 instances of CDS,atp8
3 instances of CDS,cox2
3 instances of CDS,cox3
3 instances of CDS,coxI
3 instances of rRNA,rnl
3 instances of rRNA,rns
2 instances of CDS,alg11
2 instances of CDS,atp9
2 instances of CDS,cob
2 instances of CDS,nad1
2 instances of CDS,nad2
2 instances of CDS,nad3
2 instances of CDS,nad4
2 instances of CDS,nad4L
2 instances of CDS,nad5
2 instances of CDS,nad6
2 instances of CDS,ND4
2 instances of rRNA,18S rRNA
1 instances of CDS,ALD
1 instances of CDS,ATP synthase beta subunit
1 instances of CDS,ATP6
1 instances of CDS,ATP8
1 instances of CDS,ATP9
1 instances of CDS,catalase
1 instances of CDS,CchGa
1 instances of CDS,CchGb
1 instances of CDS,CchGc
1 instances of CDS,CchGd
1 instances of CDS,CchGe
1 instances of CDS,CchGf
1 instances of CDS,coi
1 instances of CDS,COX1
1 instances of CDS,COX2
1 instances of CDS,COX3
1 instances of CDS,CYTB
1 instances of CDS,elongation factor 1 alpha
1 instances of CDS,Hsp70
1 instances of CDS,MAT
1 instances of CDS,ND1
1 instances of CDS,ND2
1 instances of CDS,ND3
1 instances of CDS,ND4L
1 instances of CDS,ND5
1 instances of CDS,ND6
1 instances of CDS,TPI
1 instances of rRNA,5.8S ribosomal RNA
1 instances of rRNA,5.8S rRNA
1 instances of rRNA,5S rRNA
1 instances of rRNA,rrnL
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
\subsection{The loci CSV file}\label{the-loci-csv-file}

The additional output file can be veiwed by executing the next cell,
using the linux \texttt{cat} command. Note that terminal commands can be
executed from within this notebook by prefixing them with an exclamation
mark \texttt{!} .\\\textbf{Run the next cell to write
Tetillidae\_loci.csv in this notebook}\\This output is the loci CSV file
which can be used to instruct ReproPhylo which loci to include in the
analysis and which data type (DNA or protein) to analyse.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}6{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{o}{!}cat Tetillidae\PYZus{}loci.csv
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}dna,rRNA,18s,18S ribosomal RNA,18S rRNA,small subunit 18S ribosomal
RNA
dna,rRNA,28s,28S large subunit ribosomal RNA,28S ribosomal RNA
dna,rRNA,5.8S\_rRNA,5.8S rRNA
dna,rRNA,5.8S\_ribosomal\_RNA,5.8S ribosomal RNA
dna,rRNA,5S\_rRNA,5S rRNA
dna,CDS,ALD,ALD
dna,CDS,ATP9,ATP9,atp9
dna,CDS,ATP\_synthase\_beta\_subunit,ATP synthase beta subunit
dna,CDS,CchGa,CchGa
dna,CDS,CchGb,CchGb
dna,CDS,CchGc,CchGc
dna,CDS,CchGd,CchGd
dna,CDS,CchGe,CchGe
dna,CDS,CchGf,CchGf
dna,CDS,Hsp70,Hsp70
dna,CDS,MAT,MAT
dna,CDS,MT-ATP6,atp6,ATP6
dna,CDS,MT-ATP8,atp8,ATP8
dna,CDS,MT-CO1,coi,COI,cox1,COX1,coxI
dna,CDS,MT-CO2,cox2,COX2
dna,CDS,MT-CO3,cox3,COX3
dna,CDS,MT-CYB,CYTB
dna,CDS,MT-ND1,nad1,ND1
dna,CDS,MT-ND2,nad2,ND2
dna,CDS,MT-ND3,nad3,ND3
dna,CDS,MT-ND4,nad4,ND4
dna,CDS,MT-ND4L,ND4L
dna,CDS,MT-ND5,nad5,ND5
dna,CDS,MT-ND6,nad6,ND6
dna,CDS,TPI,TPI
dna,CDS,alg11,alg11
dna,CDS,catalase,catalase
dna,CDS,cob,cob
dna,CDS,ef1a,elongation factor 1 alpha
dna,CDS,nad4L,nad4L
dna,CDS,putative\_LAGLIDADG\_protein,putative LAGLIDADG protein
dna,rRNA,rnl,rnl
dna,rRNA,rns,rns
dna,rRNA,rrnL,rrnL
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
This file has one line for each locus, where each line contains at least
four comma-separated values. The first value is the character type
(either \texttt{dna} or \texttt{prot}).

The second is the type of locus (ie `feature type' in genbank
terminology); it can be \texttt{CDS}, \texttt{rRNA}, \texttt{tRNA} or
anything else. It has to match to at least one feature type in your
genebank file in order for it to have any effect (more below). The third
value is the locus name (e.g.~COI) that will be used in ReproPhylo.

All following values are synonyms of the locus name (e.g.~coi,cox1)
which might be found as a gene name or product name in the genbank
records.

The name cannot have any white space in it, while the synonyms need to
appear as they do in the genbank file. Note that in some cases, such as
in the first line (the 18S gene) synonyms have been pooled together into
one line. In other cases, such as in the last line, (the rrnL gene,
which also apears as rnl two lines above) synonyms were not recognised
by the \texttt{list\_loci\_in\_genbank()} function and have remained in
seperate lines. In order to tell ReproPhylo that these two lines are the
same gene, a shared integer can be added to both lines in this manner:

I have used the number five to show that any number can be used, as long
as the lines to join share that number. Also, it's important to remember
the comma before the integer (this is CSV file) otherwise the number
will be taken as a part of the last alias.

Another possible way to influence the analysis by editing the loci CSV
is to change the first value from \texttt{dna} to \texttt{prot}, or to
delete lines containing loci which are not interesting. Note that loci
matching less than four unique sequences in the genbank file will be
dropped automatically in subsequent stages of the analysis as they will
not produce a meaningful tree.

It is possible to edit the loci CSV within this notebook, by starting a
code cell with the line \texttt{\%\%file filename}. When this type of
cell is executed with \texttt{shift+enter}, its content, excluding the
first line, will be written to the file \texttt{filename}. In the code
cell below, I have copied and pasted the content of
\texttt{Tetillidae\_loci.csv}, which we have printed above, using
\texttt{!cat}. I deleted most of the lines so that it now contains only
the 18s, 28s and cox1 genes, which will carry through the
analysis.\\\textbf{Run the next cell to write the edited loci CSV file
to \texttt{Tetillidae\_loci\_edited.csv}.}

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}7{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{o}{\PYZpc{}\PYZpc{}}\PY{k}{file} \PY{n}{Tetillidae\PYZus{}loci\PYZus{}edited}\PY{o}{.}\PY{n}{csv}
\PY{n}{dna}\PY{p}{,}\PY{n}{rRNA}\PY{p}{,}\PY{l+m+mi}{18}\PY{n}{s}\PY{p}{,}\PY{l+m+mi}{18}\PY{n}{S} \PY{n}{ribosomal} \PY{n}{RNA}\PY{p}{,}\PY{l+m+mi}{18}\PY{n}{S} \PY{n}{rRNA}\PY{p}{,}\PY{n}{small} \PY{n}{subunit} \PY{l+m+mi}{18}\PY{n}{S} \PY{n}{ribosomal} \PY{n}{RNA}
\PY{n}{dna}\PY{p}{,}\PY{n}{rRNA}\PY{p}{,}\PY{l+m+mi}{28}\PY{n}{s}\PY{p}{,}\PY{l+m+mi}{28}\PY{n}{S} \PY{n}{large} \PY{n}{subunit} \PY{n}{ribosomal} \PY{n}{RNA}\PY{p}{,}\PY{l+m+mi}{28}\PY{n}{S} \PY{n}{ribosomal} \PY{n}{RNA}
\PY{n}{dna}\PY{p}{,}\PY{n}{CDS}\PY{p}{,}\PY{n}{MT}\PY{o}{\PYZhy{}}\PY{n}{CO1}\PY{p}{,}\PY{n}{coi}\PY{p}{,}\PY{n}{COI}\PY{p}{,}\PY{n}{cox1}\PY{p}{,}\PY{n}{COX1}\PY{p}{,}\PY{n}{coxI}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}Overwriting Tetillidae\_loci\_edited.csv
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
\subsection{Starting a ReproPhylo
Project}\label{starting-a-reprophylo-project}

Now that we have a grasp of the loci content of our data and have
prepared a loci file, we can use it to start a ReproPhylo
\texttt{Project}. The \texttt{Project} is a ReproPhylo object which will
contain all the input and output datasets in our pipeline and will allso
record important methods and platform information, which will allow us
to report, reproduce and extend the analysis. We can start a
\texttt{Project} by saving a \texttt{Project()} instance into a variable
and providing it with a loci file
name:\\\texttt{pj = Project("loci\_file\_name")}

\textbf{Run the next cell to start a \texttt{Project} instance, which
will include the locuse descriptions specified in
\texttt{Tetillidae\_loci\_edited.csv}.}

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}8{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj} \PY{o}{=} \PY{n}{Project}\PY{p}{(}\PY{l+s}{\PYZdq{}}\PY{l+s}{Tetillidae\PYZus{}loci\PYZus{}edited.csv}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
\subsection{Reading data from a genbank file into the \texttt{Project}
instance}\label{reading-data-from-a-genbank-file-into-the-project-instance}

The \texttt{Project} object has a bunch of methods that allow it to read
data, manage it, configure analyses and run them. In Python, methods are
called using a dot notation:
\texttt{object.method(the, method, variables)}. Variables can be
positional, ie their meaning is determined by their position. These are
required. In addition there are sometimes optional variables, in which
case they will be assigned using the format \texttt{keyword=value}.
Reading a genbank file, or several genbank files, is done with the
method \texttt{read\_embl\_genbank}. It receives one positional
argument, which is a list of one or more genbank filenames, as follows:

or

As the files are read, only loci that correspond to loci that were
described in the loci CSV file will be read, and others will be dropped.

\textbf{To read the file \texttt{Tetillidae.gb} into the
\texttt{Project} using the method \texttt{read\_embl\_genbank()}, run
the next cell. }

As its name conveys, embl formatted files are just as welcome.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}9{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{read\PYZus{}embl\PYZus{}genbank}\PY{p}{(}\PY{p}{[}\PY{l+s}{\PYZdq{}}\PY{l+s}{Tetillidae.gb}\PY{l+s}{\PYZdq{}}\PY{p}{]}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}



[master aa07fe5] 1 genbank/embl data file(s) from Sun Nov 30 22:24:25
2014
 1 file changed, 15099 insertions(+)
 create mode 100644 Tetillidae.gb

\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
Since we are running git at the background, reading the file invoked a
message from git saying the file was registered in the repository. In
git terminology, adding a file to the repository, or recording canges to
a file is called a \textbf{commit}. In other words, we have commited the
canges to the repository. \#\# Exploring the OTU content of the data Now
that we have specified the loci to analyse, and also provided sequence
data and its associated metadata, we can write a table summarising the
species repesentation for each locus. This is done with the
\texttt{species\_vs\_loci} method, which is used as follows:\\

\textbf{By running the next cell, write a summary table to the file
``species.csv'' using the \texttt{species\_vs\_loci} method.}

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}10{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{species\PYZus{}vs\PYZus{}loci}\PY{p}{(}\PY{l+s}{\PYZdq{}}\PY{l+s}{species.csv}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
A utility function called \texttt{view\_csv\_as\_table} can be used to
display the table in this notebook. It is not a \texttt{Project} method
and is thus simply used as follows:\\

where `separator' is the character used to separate the values to
columns. In this case tha separator is \texttt{"\textbackslash{}t"}
(tab). The table will be sorted by the alphabetical order of the species
names.

\textbf{Use the \texttt{view\_csv\_as\_table} function to print
``species.csv'' in this notebook by runnig the next cell}

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}11{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{view\PYZus{}csv\PYZus{}as\PYZus{}table}\PY{p}{(}\PY{l+s}{\PYZdq{}}\PY{l+s}{species.csv}\PY{l+s}{\PYZdq{}}\PY{p}{,}\PY{l+s}{\PYZdq{}}\PY{l+s+se}{\PYZbs{}t}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}species                                       18s   28s   MT-CO1
Acanthotetilla celebensis                     0     1     1
Acanthotetilla cf. seychellensis 0CDN8107-V   1     0     0
Acanthotetilla walteri                        0     1     3
Amphitethya cf. microsigma AS-2012            0     1     1
Calthropella geodioides                       0     1     1
Cinachyra antarctica                          0     1     3
Cinachyra barbata                             1     1     1
Cinachyrella alloclada                        3     1     2
Cinachyrella anomala                          0     2     5
Cinachyrella apion                            1     1     4
Cinachyrella australiensis                    3     4     16
Cinachyrella cf. paterifera 0M9H2022-P        1     0     0
Cinachyrella cf. schulzei                     0     1     0
Cinachyrella kuekenthali                      3     0     5
Cinachyrella levantinensis                    4     4     6
Cinachyrella paterifera                       0     1     2
Cinachyrella schulzei                         2     1     2
Cinachyrella sp. 1A KRC-2012                  0     0     1
Cinachyrella sp. 1B KRC-2012                  0     0     1
Cinachyrella sp. 22-XI-02-1-13 N25            1     0     0
Cinachyrella sp. 24-XI-02-3-2 N28             1     0     0
Cinachyrella sp. 3-1                          1     0     0
Cinachyrella sp. 3-10                         1     0     0
Cinachyrella sp. 3-9                          1     0     0
Cinachyrella sp. 3473                         1     1     1
Cinachyrella sp. AS-2010                      0     0     2
Cinachyrella sp. MI-2012                      2     0     0
Cinachyrella sp. USNM 1204826                 1     0     0
Cinachyrella sp. USNM 1204829                 1     0     0
Cinachyrella sp. sp11                         1     0     0
Craniella cf. leptoderma AS-2012              4     7     6
Craniella cranium                             1     1     1
Craniella neocaledoniae                       1     0     0
Craniella sagitta                             1     1     3
Craniella sp. 0CDN5142-X                      1     0     0
Craniella sp. 2-7                             1     0     0
Craniella sp. 3318                            1     0     1
Craniella sp. 3878                            2     1     2
Craniella sp. AS-2010                         0     0     1
Craniella sp. PC-2011                         0     1     1
Craniella zetlandica                          1     1     2
Fangophilina sp. MK-2012                      3     3     2
Geodia barretti                               0     0     1
Geodia cydonium                               1     0     0
Geodia neptuni                                1     0     1
Pachymatisma johnstonia                       0     1     1
Paratetilla bacca                             2     3     6
Paratetilla sp. 2209                          0     1     0
Paratetilla sp. 2656                          0     1     1
Paratetilla sp. KRC-2012                      0     0     1
Tetilla japonica                              2     1     1
Tetilla leptoderma                            0     0     1
Tetilla muricyi                               0     1     1
Tetilla pentatriaena                          0     0     1
Tetilla radiata                               0     1     1
Thenea levis                                  0     1     1
Theonella swinhoei                            0     1     1
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
\subsection{Reading de novo sequence
data}\label{reading-de-novo-sequence-data}

Often, when reconstructing a phylogenetic tree, we would like to include
our own new sequences. The \texttt{Project} method \texttt{read\_denovo}
is designed to do that. It has two positional parameters: The first is a
list of file names to read. The files have to be in the same format
(e.g.~fasta) and to have the same sequence type (dna/prot). The second
parameter is a string representing the sequence type, and could be
either \texttt{dna} or \texttt{prot}. An optional parameter is the
\texttt{format} parameter. By default this is set to
\texttt{format="fasta"}, however it can be any format recognised by
Biopython, including sequence alignments. If aligned sequences are read
however the gaps will be removed. Here is a usage example:\\

\textbf{Task: Read a novel DNA sequence from the file
\texttt{Tetillidae\_denovo\_sequence.fasta} using the \texttt{Project}
method \texttt{read\_denovo()}}

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}12{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{read\PYZus{}denovo}\PY{p}{(}\PY{p}{[}\PY{l+s}{\PYZdq{}}\PY{l+s}{Tetillidae\PYZus{}denovo\PYZus{}sequence.fasta}\PY{l+s}{\PYZdq{}}\PY{p}{]}\PY{p}{,} \PY{l+s}{\PYZdq{}}\PY{l+s}{dna}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}



[master 47ef494] 1 denovo data file(s) from Sun Nov 30 22:24:29 2014
 1 file changed, 2 insertions(+)
 create mode 100644 Tetillidae\_denovo\_sequence.fasta

\end{alltt}

            \end{InvisibleVerbatim}
            
                \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-out-prompt}Out\hspace{4pt}{[}12{]}:\hspace{4pt}}\\*
                \vspace{-2.55\baselineskip}\begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}1\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
Once again we are informed that git has recorded the new data file. The
sequence is now in the \texttt{Project}. However, it discloses no
information that will allow us to utilize it. In order for ReproPhylo to
make sense of it we need to provide information about the gene it
contains. If it contains a coding sequence, we might want to specify an
open reading frame so that it will be translated and provide us with a
protein sequence. Even if we want to analyse DNA sequences, if a protein
sequence is available we then have the option of codon-aligning our
sequences. To get a protein sequence automatically, we must provide the
reading frame, the location of the exons, the strand and the
\href{http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi}{translation
table}, as feature qualifiers.

The next code cell will add a feature to our denovo sequence and will
include all the relevant information to allow a translation to be crated
for us.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}13{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{qualifiers}\PY{o}{=}\PY{p}{\PYZob{}}\PY{l+s}{\PYZsq{}}\PY{l+s}{gene}\PY{l+s}{\PYZsq{}}\PY{p}{:} \PY{l+s}{\PYZsq{}}\PY{l+s}{cox1}\PY{l+s}{\PYZsq{}}\PY{p}{,}
            \PY{l+s}{\PYZsq{}}\PY{l+s}{transl\PYZus{}table}\PY{l+s}{\PYZsq{}}\PY{p}{:} \PY{l+m+mi}{4}\PY{p}{,}
            \PY{l+s}{\PYZsq{}}\PY{l+s}{codon\PYZus{}start}\PY{l+s}{\PYZsq{}}\PY{p}{:} \PY{l+m+mi}{1}\PY{p}{,}
            \PY{l+s}{\PYZsq{}}\PY{l+s}{organism}\PY{l+s}{\PYZsq{}}\PY{p}{:} \PY{l+s}{\PYZsq{}}\PY{l+s}{Craniella microsigma}\PY{l+s}{\PYZsq{}}\PY{p}{\PYZcb{}}
\PY{k}{for} \PY{n}{record} \PY{o+ow}{in} \PY{n}{pj}\PY{o}{.}\PY{n}{records}\PY{p}{:}
    \PY{k}{if} \PY{l+s}{\PYZsq{}}\PY{l+s}{denovo}\PY{l+s}{\PYZsq{}} \PY{o+ow}{in} \PY{n}{record}\PY{o}{.}\PY{n}{id}\PY{p}{:} \PY{c}{\PYZsh{} New sequences are assigned with IDs starting}
                              \PY{c}{\PYZsh{} with \PYZsq{}denovo\PYZsq{}. See below.}
        \PY{n}{pj}\PY{o}{.}\PY{n}{add\PYZus{}feature\PYZus{}to\PYZus{}record}\PY{p}{(}\PY{n}{record}\PY{o}{.}\PY{n}{id}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{CDS}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                                 \PY{c}{\PYZsh{}The location is specified as a list}
                                 \PY{c}{\PYZsh{} of lists. Every sub\PYZhy{}list is an exon}
                                 \PY{c}{\PYZsh{} and has the start, the end and the strand.}
                                 \PY{c}{\PYZsh{} The numbers are \PYZdq{}real\PYZdq{} positions and not}
                                 \PY{c}{\PYZsh{} machine. ie, counting starts from 1.}
                                 \PY{n}{location}\PY{o}{=}\PY{p}{[}\PY{p}{[}\PY{l+m+mi}{1}\PY{p}{,}\PY{l+m+mi}{786}\PY{p}{,}\PY{l+m+mi}{1}\PY{p}{]}\PY{p}{,}\PY{p}{[}\PY{l+m+mi}{1742}\PY{p}{,}\PY{l+m+mi}{2092}\PY{p}{,}\PY{l+m+mi}{1}\PY{p}{]}\PY{p}{]}\PY{p}{,}
                                 \PY{n}{qualifiers}\PY{o}{=}\PY{n}{qualifiers}\PY{p}{)} 
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
\subsection{Editing the metadata
programatically}\label{editing-the-metadata-programatically}

The genbank file comes with a lot of \textbf{metadata}; descriptions and
extra information about the sequence such as taxonomy, publication etc.
However, this data needs to be organized a bit and also extended for it
to be useful. Only feature qualifiers can be manipulated (ie, not source
qualifiers). The following cell shows how to copy a qualifier from the
source to the features, where it can be manipulated. For this we'll use
the \texttt{add\_qualifier\_from\_source} method.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}14{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{add\PYZus{}qualifier\PYZus{}from\PYZus{}source}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{organism}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
Here we are going to use the info in the \texttt{organism} qualifier to
create a new qualifier containing only the genus. This will allow us to
examine in a useful way the genera represented in our sequence data.

The code below loops over the tetillid genera and puts the genus name in
the new \texttt{genus} qualifier, if this name is found in the
\texttt{organism} qualifier. This is done by the
\texttt{if\_this\_than\_that} method. We will use the \texttt{part} mode
to indicate that while searching for the genus name in the species name
we do not require the whole \texttt{organism} value to fit, just part of
it.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}15{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{k}{for} \PY{n}{genus} \PY{o+ow}{in} \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{Cinachyrella}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Cinachyra}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Craniella}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Tetilla}\PY{l+s}{\PYZsq{}}\PY{p}{,}
              \PY{l+s}{\PYZsq{}}\PY{l+s}{Acanthotetilla}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Amphitethya}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Fangophilina}\PY{l+s}{\PYZsq{}}\PY{p}{,}
              \PY{l+s}{\PYZsq{}}\PY{l+s}{Paratetilla}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{:}
    \PY{n}{pj}\PY{o}{.}\PY{n}{if\PYZus{}this\PYZus{}then\PYZus{}that}\PY{p}{(}\PY{n}{genus}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{organism}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{n}{genus}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{genus}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                         \PY{n}{mode}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{part}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
Several additional methods for editing the metadata programatically can
be used. In the next cell, place the curser right after the dot and
click the tab key. A list of all the \texttt{Project} methods will
appear. You may review it to see all the methods related to metadata
manipulations as well as others.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}16{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                
            \begin{alltt}

          File "<ipython-input-16-426655ce4982>", line 1
        pj.
           \^{}
    SyntaxError: invalid syntax


\end{alltt}
        
            
        
    
\subsection{Editing the metadata
manually}\label{editing-the-metadata-manually}Sometimes, if the dataset is small, there are errors in our metadata
that we might wish to correct manually. This is not ideal for a
reproducible experiment, but ReproPhylo will keep track of all changes
via git, documenting the changes that you have made.

We are going to save all the metadata in the \texttt{Project} in a CSV
format suitable for a spreadsheet program:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}17{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{write}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{Tetillidae\PYZus{}metadata.csv}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{n}{format}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{csv}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}


[master c5d4e2f] Records csv text file from Sun Nov 30 22:24:37 2014
 1 file changed, 196 insertions(+)
 create mode 100644 Tetillidae\_metadata.csv

\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
The metadata can be altered by editing the metadata spreadsheet and
reading it back to into the \texttt{Project}. This can be done in Excel
or Libreoffice. However, while using any of those programs, be very
aware to the
\href{http://nsaunders.wordpress.com/2012/10/22/gene-name-errors-and-excel-lessons-not-learned/}{errors}
that might be introduced by the autocorrect, autocomplete and autoformat
functions running in there.

So, now that we know we need to be careful, open the file
`Tetillidae\_metadata.csv'. Try to spot the new \textbf{genus} qualifier
you have created above. In order to describe the content of this file
we'll be using Biopython and GenBank terminology, as illustrated in this
figure:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}18{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{c}{\PYZsh{} execute if figure is small}
\PY{k+kn}{from} \PY{n+nn}{IPython.display} \PY{k+kn}{import} \PY{n}{Image}
\PY{n}{Image}\PY{p}{(}\PY{n}{filename}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{genbank\PYZus{}terminology.jpg}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-out-prompt}Out\hspace{4pt}{[}18{]}:\hspace{4pt}}\\*
                \vspace{-2.55\baselineskip}\begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
    \begin{center}
    \includegraphics[max size={\textwidth}{\textheight}]{Tutorial_files/Tutorial_38_0.jpeg}
    \par
    \end{center}
    
            \end{InvisibleVerbatim}
            
        
    
Following the GenBank record structure in the figure above, our metadata
spreadsheet includes the following columns. - Column \textbf{A} is the
\textbf{Record ID}. - Column \textbf{B} is a short representation of the
\textbf{Seq}, showing it's 5' and 3' portions. This can help validate
the record's identity if a doubt arises. - Columns \textbf{C} through
\textbf{W} are the source feature qualifiers. - Column \textbf{X} is the
taxonomy line from the \textbf{ORGANISM} field in the
\textbf{Annotations} section of the GenBank record.

A genbank file may have several sequence features (ie, loci) described.
However, it will always have only a single source feature. The
information in the source feature is relevant to all the features in the
record. For this reason, each and every line in the metadata
spreadsheet, starts with the srouce feature qualifiers. Say a GenBank
record includes a source feature and five additional features: This will
be represented in the spreadsheet as five lines, one for each non-source
feature. Across the five lines, the first few columns will be identical
as they are derived from the shared source feature. The five lines
(representing the five features) will be assigned feature IDs, which
consist of the record ID and the suffixes \_f0 to \_f4 (eg,
\textbf{NC\_010198.1\_f3}).

Now scroll down to the very end of the file, to see your denovo
sequence. You'll notice that it has been assigned the record ID
\textbf{denovo0}, a source feature ID \textbf{denovo0\_source} and a
feature ID \textbf{denovo0\_f0}. The first word in the fasta header is
recorded as \textbf{original\_id} and the remaining of the header as
\textbf{original\_desc}, short for \textbf{Description}. Find the
\textbf{translation} column to make sure that a protein sequence has
been created. You may move or copy these values to other columns,
programmatically as we did above, or manually by editing the spreadsheet
as we will do next. At the moment, source qualifiers cannot be edited
programatically. They can be copied to the feature qualifiers and edited
there. It is, however, possible to edit them in the spreadsheet.In order to proceed with the analysis there are a few additional things
we need to tidy up in our metadata.

First, we should include some morphological information so that we can
annotate our trees with it (it will also be available to anyone wanting
to carry the trees out of reprophylo to do some character evolution
tests). To do this, first add the three columns \textbf{porocalyx},
\textbf{cortex} and \textbf{calthrops} at the right hand side of the
table. These are tetillid morphological features. Now sort the
spreadsheet according to the \textbf{genus} column. Add the values
bellow to each feature, based of it's genus. You can copy the values to
the first row in each genus and then drag them down all the way to the
last line of this genus. Can you think of a way to this in a script
using the \texttt{if\_this\_then\_that} method?

Next, we want to indicate our outgroup species. In fact, any line that
has not been assigned with the genus (ie it says `null') is in the
outgroup order `Astrophorida'. In the genus column, replace `null' with
`Astrophorida'. The word `null' represents a qualifier that does not
exist in the feature. Therefore, we cannot use the word `null' as the
outgroup value in the genus column. The genus qualifier will not exist
for features with the value `null', unless we change it to something
else.

In addition, we want to make sure we will be able to label our new
sequence with it's species name. So copy the species name from the
\textbf{original\_desc} column to the \textbf{source:\_organism} column,
ommiting the gene name, cox1.

Last, there is a coding sequence, \textbf{AM076987.1\_f1}, encoded from
within a cox1 mitochondrial intron, which does not belong to the cox1
CDS. Yet, it has the value `cox1' in its gene qualifier. To make sure
this sequence does not end up in our cox1 dataset, change its gene
qualifier value from cox1 to laglidadg.

Now, save the file and make sure you retain the tab delimited CSV
format.\\The method \texttt{correct\_metadata\_from\_file} will modify
the data feature qualifiers according to the canges made to the CSV
file. It is used as follows:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}19{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{correct\PYZus{}metadata\PYZus{}from\PYZus{}file}\PY{p}{(}\PY{l+s}{\PYZdq{}}\PY{l+s}{Tetillidae\PYZus{}metadata\PYZus{}edited.csv}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}


[master 7102ee8] Corrected metadata CSV file from Sun Nov 30 22:24:40
2014
 1 file changed, 196 insertions(+)
 create mode 100644 Tetillidae\_metadata\_edited.csv

\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
\subsection{Reporting sequence
statistics}\label{reporting-sequence-statistics}After making sure the sequences are assigned with the correct gene or
product name, we can split them to their repsective datasets by using
this method:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}20{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{extract\PYZus{}by\PYZus{}locus}\PY{p}{(}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
Once this is done, the first thing ReproPhylo can do with the data is to
report some sequence statistics:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}21{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{o}{\PYZpc{}}\PY{k}{matplotlib} \PY{n}{inline}
\PY{n}{pj}\PY{o}{.}\PY{n}{report\PYZus{}seq\PYZus{}stats}\PY{p}{(}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}Distribution Of Sequence Lengths
Distribution Of Sequence Statistic "Gc\_Content"
Distribution Of Sequence Statistic "Nuc\_Degen\_Prop"
Distribution Of Sequence Statistic "Prot\_Degen\_Prop"
\end{alltt}

            \end{InvisibleVerbatim}
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}/usr/lib/pymodules/python2.7/matplotlib/figure.py:371: UserWarning:
matplotlib is currently using a non-GUI backend, so cannot show the
figure
  "matplotlib is currently using a non-GUI backend, "
\end{alltt}

            \end{InvisibleVerbatim}
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
    \begin{center}
    \includegraphics[max size={\textwidth}{\textheight}]{Tutorial_files/Tutorial_46_2.png}
    \par
    \end{center}
    
            \end{InvisibleVerbatim}
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
    \begin{center}
    \includegraphics[max size={\textwidth}{\textheight}]{Tutorial_files/Tutorial_46_3.png}
    \par
    \end{center}
    
            \end{InvisibleVerbatim}
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
    \begin{center}
    \includegraphics[max size={\textwidth}{\textheight}]{Tutorial_files/Tutorial_46_4.png}
    \par
    \end{center}
    
            \end{InvisibleVerbatim}
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
    \begin{center}
    \includegraphics[max size={\textwidth}{\textheight}]{Tutorial_files/Tutorial_46_5.png}
    \par
    \end{center}
    
            \end{InvisibleVerbatim}
            
        
    
\subsection{Configuring and running the sequence
alignment}\label{configuring-and-running-the-sequence-alignment}

Analyses in ReproPhylo are configured using \texttt{Conf} objects and
are then run by passing the \texttt{Conf} object to the respective
\texttt{Project} method. For example, a sequence alignment run will be
configured by making an \texttt{AlnConf} object and then passing it to
the \texttt{align} method. The next two code cells make two
\texttt{AlnConf} objects. The first one configures the program
\texttt{MAFFT} to run the \texttt{L-ins-i} algorithm to align the MT-CO1
sequences. \texttt{MAFFT} is the default alignment program so it is not
explicitly specified.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}22{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{c}{\PYZsh{} Mafft linsi algorithm}
\PY{n}{mafftLinsi} \PY{o}{=} \PY{n}{AlnConf}\PY{p}{(}\PY{n}{pj}\PY{p}{,}
                     \PY{n}{method\PYZus{}name}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{mafftLinsi}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                     \PY{n}{loci}\PY{o}{=}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{MT\PYZhy{}CO1}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                     \PY{n}{cline\PYZus{}args}\PY{o}{=}\PY{n+nb}{dict}\PY{p}{(}\PY{n}{localpair}\PY{o}{=}\PY{n+nb+bp}{True}\PY{p}{,} \PY{n}{maxiterate}\PY{o}{=}\PY{l+m+mi}{1000}\PY{p}{)}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}mafft --localpair --maxiterate 1000 137601417386286
.73\_CDS\_proteins\_MT-CO1.fasta
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
The \texttt{AlnConf} object reports the command lines that will be run.
The next cell configures the program \texttt{MUSCLE} to align the rRNA
loci 18s and 28s.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}23{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{rRNA} \PY{o}{=} \PY{p}{[}\PY{l+s}{\PYZdq{}}\PY{l+s}{18s}\PY{l+s}{\PYZdq{}}\PY{p}{,}\PY{l+s}{\PYZdq{}}\PY{l+s}{28s}\PY{l+s}{\PYZdq{}}\PY{p}{]}

\PY{n}{muscle} \PY{o}{=} \PY{n}{AlnConf}\PY{p}{(}\PY{n}{pj}\PY{p}{,} \PY{n}{loci}\PY{o}{=}\PY{n}{rRNA}\PY{p}{,}
                 \PY{n}{method\PYZus{}name}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{MuscleDefaults}\PY{l+s}{\PYZsq{}}\PY{p}{,} 
                 \PY{n}{program\PYZus{}name}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{muscle}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}muscle -in 198921417386289.82\_18s.fasta
muscle -in 198921417386289.82\_28s.fasta
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
The complete set of parameters in the \texttt{AlnConf} object are
described here:We can run both the \texttt{AlnConf} objects we have configured in one
go, using the \texttt{align} method, which takes a single parameter, an
\texttt{AlnConf} object(s) list.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}24{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{align}\PY{p}{(}\PY{p}{[}\PY{n}{mafftLinsi}\PY{p}{,}
          \PY{n}{muscle}\PY{p}{]}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
\subsection{Checkpointing the
analysis}\label{checkpointing-the-analysis}Once the sequence alignment is done, the results only exist in the
\texttt{Project} object in memory. It's a good idea to checkpoint the
analysis by saving it to a file. \texttt{Project} objects are saved to
pickle files (a standard preservation method for both vegetables and
data) using the \texttt{pickle\_pj} function as follows:

The \texttt{Project} can be revived from a pickle file with the
\texttt{unpickle\_db} function:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}25{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pickle\PYZus{}pj}\PY{p}{(}\PY{n}{pj}\PY{p}{,} \PY{l+s}{\PYZdq{}}\PY{l+s}{tetillidae\PYZus{}tutorial}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}DEBUG:Cloud:Log file (/home/amir/.picloud/cloud.log) opened
\end{alltt}

            \end{InvisibleVerbatim}
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}
[master 222fb38] A pickled Project from Sun Nov 30 22:25:30 2014
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 tetillidae\_tutorial

\end{alltt}

            \end{InvisibleVerbatim}
            
                \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-out-prompt}Out\hspace{4pt}{[}25{]}:\hspace{4pt}}\\*
                \vspace{-2.55\baselineskip}\begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}'tetillidae\_tutorial'\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
Since we are using version control, we can overwrite the pickle file
each time we checkpoint. If we need to roll back to any overwritten
version, git will allow us to do that.\subsection{Configuring and running the alignment
trimming}\label{configuring-and-running-the-alignment-trimming}Alignments often have regions that are messy, with many gaps, which you
would like to explude from the tree building as they may introduce more
noise than signal. Alignment trimming allows such regions to be excluded
based on reproducible alignment quality rules. Currently
\href{http://trimal.cgenomics.org/}{trimAl} is the only program used for
alignment trimming in ReproPhylo. Trimming is carried out in a similar
way to the sequence alignment stage. These are the default settings of
the \texttt{TrimalConf} object.

We'll trim the alignments using the default settings which is the
\texttt{gappyout} option in trimAl:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}26{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{gappyout} \PY{o}{=} \PY{n}{TrimalConf}\PY{p}{(}\PY{n}{pj}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}trimal -in 791341417386336.02\_28s@MuscleDefaults.fasta -gappyout
trimal -in 791341417386336.02\_18s@MuscleDefaults.fasta -gappyout
trimal -in 791341417386336.02\_MT-CO1@mafftLinsi.fasta -gappyout
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}27{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{trim}\PY{p}{(}\PY{p}{[}\PY{n}{gappyout}\PY{p}{]}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
\subsection{Configuring and running the tree
reconstruction}\label{configuring-and-running-the-tree-reconstruction}\href{http://sco.h-its.org/exelixis/web/software/raxml/index.html}{RAxML}
is currently the only phylogenetic reconstruction program. The algorithm
if set using the \texttt{preset=} keyword and takes either the value
\texttt{fa} for a single ML search with rapid bootstrap, \texttt{fD\_fb}
for ML search(es) with relBootstrap or \texttt{fd\_b\_fb} for ML
search(es) with thorough bootstrap. The model is passed with
\texttt{model=} keyword and a protein substitution matrix with
\texttt{matrix=}. The number of threads to run is set with
\texttt{threads=} any additioanl argument can be passed in a dictionary
to \texttt{cline\_args=}. The number of ML searches would be passed as
\texttt{-N: 100} in this dictionary, and the number of bootstrap
replicates as \texttt{-\#: 100}. The full default settings are:

Since this is a first pass of the data we can choose \texttt{fD\_fb} as
the preset as \texttt{relBootstrap} is extremely fast.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}28{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{raxml} \PY{o}{=} \PY{n}{RaxmlConf}\PY{p}{(}\PY{n}{pj}\PY{p}{,} \PY{n}{method\PYZus{}name}\PY{o}{=}\PY{l+s}{\PYZdq{}}\PY{l+s}{fD\PYZus{}fb}\PY{l+s}{\PYZdq{}}\PY{p}{,} \PY{n}{preset}\PY{o}{=}\PY{l+s}{\PYZdq{}}\PY{l+s}{fD\PYZus{}fb}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}raxmlHPC-PTHREADS-SSE3 -f D -m GTRGAMMA -n
254401417386341.34\_28s@MuscleDefaults@gappyout0 -p 691 -s
254401417386341.34\_28s@MuscleDefaults@gappyout.fasta -T 4 -N 1
raxmlHPC-PTHREADS-SSE3 -f b -m GTRGAMMA -n
254401417386341.34\_28s@MuscleDefaults@gappyout1 -p 738 -s
254401417386341.34\_28s@MuscleDefaults@gappyout.fasta -t
RAxML\_bestTree.254401417386341.34\_28s@MuscleDefaults@gappyout0 -T 4 -z
RAxML\_rellBootstrap.254401417386341.34\_28s@MuscleDefaults@gappyout0
raxmlHPC-PTHREADS-SSE3 -f D -m GTRGAMMA -n
254401417386341.34\_18s@MuscleDefaults@gappyout0 -p 739 -s
254401417386341.34\_18s@MuscleDefaults@gappyout.fasta -T 4 -N 1
raxmlHPC-PTHREADS-SSE3 -f b -m GTRGAMMA -n
254401417386341.34\_18s@MuscleDefaults@gappyout1 -p 285 -s
254401417386341.34\_18s@MuscleDefaults@gappyout.fasta -t
RAxML\_bestTree.254401417386341.34\_18s@MuscleDefaults@gappyout0 -T 4 -z
RAxML\_rellBootstrap.254401417386341.34\_18s@MuscleDefaults@gappyout0
raxmlHPC-PTHREADS-SSE3 -f D -m GTRGAMMA -n 254401417386341.34\_MT-
CO1@mafftLinsi@gappyout0 -p 764 -s 254401417386341.34\_MT-
CO1@mafftLinsi@gappyout.fasta -T 4 -N 1
raxmlHPC-PTHREADS-SSE3 -f b -m GTRGAMMA -n 254401417386341.34\_MT-
CO1@mafftLinsi@gappyout1 -p 731 -s 254401417386341.34\_MT-
CO1@mafftLinsi@gappyout.fasta -t RAxML\_bestTree.254401417386341.34\_MT-
CO1@mafftLinsi@gappyout0 -T 4 -z RAxML\_rellBootstrap.254401417386341
.34\_MT-CO1@mafftLinsi@gappyout0
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}29{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{tree}\PY{p}{(}\PY{p}{[}\PY{n}{raxml}\PY{p}{]}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
\subsection{Printing tree figures}\label{printing-tree-figures}A very useful feature of reproducible pipelines is the ability to
auto-annotate tree figures. The \texttt{Project} method
\texttt{annotate} annotates all the trees uniformly and makes use of
\href{http://etetoolkit.org/}{ETE2}. It has the following options:

We will just make a simple annotation at this stage.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}30{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{supports} \PY{o}{=} \PY{p}{\PYZob{}}\PY{l+s}{\PYZsq{}}\PY{l+s}{black}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+m+mi}{100}\PY{p}{,}\PY{l+m+mi}{99}\PY{p}{]}\PY{p}{,}
            \PY{l+s}{\PYZsq{}}\PY{l+s}{dimgray}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+m+mi}{99}\PY{p}{,}\PY{l+m+mi}{75}\PY{p}{]}\PY{p}{,}
            \PY{l+s}{\PYZsq{}}\PY{l+s}{silver}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+m+mi}{75}\PY{p}{,}\PY{l+m+mi}{50}\PY{p}{]}\PY{p}{\PYZcb{}}

\PY{n}{pj}\PY{o}{.}\PY{n}{annotate}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{.}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{genus}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Astrophorida}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{organism}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
            \PY{n}{node\PYZus{}support\PYZus{}dict}\PY{o}{=}\PY{n}{supports}\PY{p}{,}
            \PY{n}{html}\PY{o}{=}\PY{l+s}{\PYZdq{}}\PY{l+s}{figures}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}31{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pickle\PYZus{}pj}\PY{p}{(}\PY{n}{pj}\PY{p}{,} \PY{l+s}{\PYZdq{}}\PY{l+s}{tetillidae\PYZus{}tutorial}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}
[master 6f5916f] A pickled Project from Sun Nov 30 22:26:07 2014
 1 file changed, 0 insertions(+), 0 deletions(-)

\end{alltt}

            \end{InvisibleVerbatim}
            
                \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-out-prompt}Out\hspace{4pt}{[}31{]}:\hspace{4pt}}\\*
                \vspace{-2.55\baselineskip}\begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}'tetillidae\_tutorial'\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
\subsection{Making a concatenated
matrix}\label{making-a-concatenated-matrix}

The final product in a phylogenetic analysis would usually be a
phylogenetic tree based on a supermatrix. ReproPhylo can handle the
concatenation of sequence alignments for this purpose.

To do that, the user needs to provide information that indicates which
sequences should be concatenated into a single line in the supermatrix.
In this tutorial we will use the \texttt{specimen\_voucher} to tie
together sequences of different genes that belong to the same sample or
OTU. However, some of the genbank records miss this qualifiers. In othe
cases, the voucher numbers are spelled slightly differently in different
records. All of this could be fixed manually in the metadata CSV, or
programatically.

First, we add the \texttt{specimen\_voucer} to records that miss it:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}32{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{add\PYZus{}qualifier\PYZus{}from\PYZus{}source}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{)}

\PY{c}{\PYZsh{} Add missing/ correct wrong specimen vouchers according to feature id}

\PY{n}{add\PYZus{}specimen\PYZus{}voucher} \PY{o}{=} \PY{p}{[}\PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{JX177968.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG\PYZus{}321405}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                        \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{JX177913.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                          \PY{l+s}{\PYZsq{}}\PY{l+s}{JX177935.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                          \PY{l+s}{\PYZsq{}}\PY{l+s}{JX177965.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{TAU\PYZus{}25617}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                        \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{JX177903.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                          \PY{l+s}{\PYZsq{}}\PY{l+s}{JX177938.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{TAU\PYZus{}25618}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                        \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{HM032740.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                          \PY{l+s}{\PYZsq{}}\PY{l+s}{JX177964.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{TAU\PYZus{}25621}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                        \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{HM032739.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                          \PY{l+s}{\PYZsq{}}\PY{l+s}{JX177962.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{TAU\PYZus{}25622}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                        \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{JX177968.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG\PYZus{}321405}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                        \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{JX177891.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{RMNH\PYZus{}POR\PYZus{}3100}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                        \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{JX177900.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                          \PY{l+s}{\PYZsq{}}\PY{l+s}{JX177926.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{TAU\PYZus{}25620}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                        \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{JX177901.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                          \PY{l+s}{\PYZsq{}}\PY{l+s}{JX177961.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                          \PY{l+s}{\PYZsq{}}\PY{l+s}{JX177956.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{TAU\PYZus{}25619}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                        \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{HM032742.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                          \PY{l+s}{\PYZsq{}}\PY{l+s}{JX177957.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{MNRJ\PYZus{}576}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{]}
\PY{k}{for} \PY{n}{add} \PY{o+ow}{in} \PY{n}{add\PYZus{}specimen\PYZus{}voucher}\PY{p}{:}
    \PY{n}{pj}\PY{o}{.}\PY{n}{add\PYZus{}qualifier}\PY{p}{(}\PY{n}{add}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]}\PY{p}{,}\PY{n}{add}\PY{p}{[}\PY{l+m+mi}{1}\PY{p}{]}\PY{p}{,}\PY{n}{add}\PY{p}{[}\PY{l+m+mi}{2}\PY{p}{]}\PY{p}{)}

\PY{c}{\PYZsh{} Reformat specimen voucher according to the specimen voucher}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
Then, we correct spelling differences:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}33{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{correct\PYZus{}specimen\PYZus{}voucher} \PY{o}{=} \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG321405}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG\PYZus{}321405}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{MHNM 16194}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{MHNM\PYZus{}16194}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{TAU 25456}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{TAU\PYZus{}25456}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG320636}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG\PYZus{}320636}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG320270}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG\PYZus{}320270}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN:85239}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN\PYZus{}85239}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG318785}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG\PYZus{}318785}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG316342}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG\PYZus{}316342}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG314224}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{QMG\PYZus{}314224}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{VM14754}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{VM\PYZus{}14754}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN:85240}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN\PYZus{}85240}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN:81789}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN\PYZus{}81789}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN:81787}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN\PYZus{}81787}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                            \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN:81785}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{ZMBN\PYZus{}81785}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{]}
\PY{k}{for} \PY{n}{correction} \PY{o+ow}{in} \PY{n}{correct\PYZus{}specimen\PYZus{}voucher}\PY{p}{:}
    \PY{n}{pj}\PY{o}{.}\PY{n}{if\PYZus{}this\PYZus{}then\PYZus{}that}\PY{p}{(}\PY{n}{correction}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]}\PY{p}{,}\PY{n}{correction}\PY{p}{[}\PY{l+m+mi}{1}\PY{p}{]}\PY{p}{,}\PY{n}{correction}\PY{p}{[}\PY{l+m+mi}{2}\PY{p}{]}\PY{p}{,}\PY{n}{correction}\PY{p}{[}\PY{l+m+mi}{3}\PY{p}{]}\PY{p}{)}

\PY{c}{\PYZsh{} Make a qualifier that will be used to concatenate OTU sequences}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
Now, for the outgroup, we have no prior knowlage of their voucher
numbers nor does it exists in the genbank file. We still want to tie
together sequences of the same species. We therefore add a new
qualifier, \texttt{OTU\_dict} that will include the
\texttt{specimen\_voucher} info where it exists, and species names for
outgroup speceis:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}34{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{copy\PYZus{}paste\PYZus{}within\PYZus{}feature}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{specimen\PYZus{}voucher}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{OTU\PYZus{}dict}\PY{l+s}{\PYZsq{}}\PY{p}{)}

\PY{c}{\PYZsh{} Add missing values to the OTU dictionary}

\PY{n}{add\PYZus{}to\PYZus{}concatenation\PYZus{}dict}\PY{o}{=}\PY{p}{[}\PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{AY737635.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                             \PY{l+s}{\PYZsq{}}\PY{l+s}{AY320032.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{OTU\PYZus{}dict}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Geodia\PYZus{}neptuni}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                           \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{EF564339.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                             \PY{l+s}{\PYZsq{}}\PY{l+s}{HM592832.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{OTU\PYZus{}dict}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Pachymatisma\PYZus{}johnstonia}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                           \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{HM592717.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                             \PY{l+s}{\PYZsq{}}\PY{l+s}{HM592765.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{OTU\PYZus{}dict}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Thenea\PYZus{}levis}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                           \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{HM592745.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                             \PY{l+s}{\PYZsq{}}\PY{l+s}{HM592820.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{OTU\PYZus{}dict}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Theonella\PYZus{}swinhoei}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                           \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{KC762708.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                             \PY{l+s}{\PYZsq{}}\PY{l+s}{NC\PYZus{}010198.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{OTU\PYZus{}dict}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Cinachyrella\PYZus{}kuekenthali}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                           \PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{HM592705.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                             \PY{l+s}{\PYZsq{}}\PY{l+s}{HM592826.1\PYZus{}f0}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{OTU\PYZus{}dict}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Calthropella\PYZus{}geodioides}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{]}

\PY{k}{for} \PY{n}{add} \PY{o+ow}{in} \PY{n}{add\PYZus{}to\PYZus{}concatenation\PYZus{}dict}\PY{p}{:}
    \PY{n}{pj}\PY{o}{.}\PY{n}{add\PYZus{}qualifier}\PY{p}{(}\PY{n}{add}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]}\PY{p}{,}\PY{n}{add}\PY{p}{[}\PY{l+m+mi}{1}\PY{p}{]}\PY{p}{,}\PY{n}{add}\PY{p}{[}\PY{l+m+mi}{2}\PY{p}{]}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
The metadata is sorted and we can design the super matrix. The
\texttt{Concatenation} class takes care of this:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}35{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{combined} \PY{o}{=} \PY{n}{Concatenation}\PY{p}{(}\PY{n}{name}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{combined}\PY{l+s}{\PYZsq{}}\PY{p}{,}                          \PY{c}{\PYZsh{} any string}
                         \PY{n}{loci}\PY{o}{=}\PY{n}{pj}\PY{o}{.}\PY{n}{loci}\PY{p}{,}                             \PY{c}{\PYZsh{} In this case we want to include all the loci.}
                         \PY{n}{otu\PYZus{}meta}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{OTU\PYZus{}dict}\PY{l+s}{\PYZsq{}}\PY{p}{,}                      \PY{c}{\PYZsh{} As explained above}
                         \PY{n}{concat\PYZus{}must\PYZus{}have\PYZus{}all\PYZus{}of}\PY{o}{=}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{MT\PYZhy{}CO1}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}       \PY{c}{\PYZsh{} Only species that have the genes in this list}
                         \PY{n}{concat\PYZus{}must\PYZus{}have\PYZus{}one\PYZus{}of} \PY{o}{=}\PY{p}{[}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{18s}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{28s}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{]}\PY{p}{)} \PY{c}{\PYZsh{} Only species that have at least on gene in each sub\PYZhy{}list}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}36{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{add\PYZus{}concatenation}\PY{p}{(}\PY{n}{combined}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}found raxml offensive char   in OTU BIOICE 3659. Replacing with
'\_ro\_'.Backing up original in the qualifier original\_OTU\_dict.
found raxml offensive char   in OTU TAU 25623. Replacing with
'\_ro\_'.Backing up original in the qualifier original\_OTU\_dict.
found raxml offensive char : in OTU TAU:25456. Replacing with
'\_ro\_'.Backing up original in the qualifier original\_OTU\_dict.
found raxml offensive char   in OTU 24-XI-02-3-2 N28. Replacing with
'\_ro\_'.Backing up original in the qualifier original\_OTU\_dict.
found raxml offensive char   in OTU 22-XI-02-1-13 N25. Replacing with
'\_ro\_'.Backing up original in the qualifier original\_OTU\_dict.
found raxml offensive char   in OTU ZMBN 77922. Replacing with
'\_ro\_'.Backing up original in the qualifier original\_OTU\_dict.
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}37{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{make\PYZus{}concatenation\PYZus{}alignments}\PY{p}{(}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}Concatenation combined wil have the following data
OTU                           18s                 28s
MT-CO1
NIWA\_28507                    JX177975.1\_f0  JX177943.1\_f0
JX177896.1\_f0
TAU\_25617                     JX177965.1\_f0  JX177935.1\_f0
JX177913.1\_f0
NIWA\_28910                    JX177982.1\_f0
JX177865.1\_f0
VM\_14754                      JX177986.1\_f0  JX177960.1\_f0
HM032751.1\_f0
ZMBN\_85239                    JX177987.1\_f0  JX177959.1\_f0
HM592669.1\_f0
Thenea\_levis                                 HM592765.1\_f0
HM592717.1\_f0
LB\_113                                       JX177936.1\_f0
JX177890.1\_f0
Calthropella\_geodioides                      HM592826.1\_f0
HM592705.1\_f0
TAU\_25618                                    JX177938.1\_f0
JX177903.1\_f0
TAU\_25619                     JX177961.1\_f0  JX177956.1\_f0
JX177901.1\_f0
NIWA\_28617                    JX177980.1\_f0
JX177912.1\_f0
LB\_817                                       JX177932.1\_f0
JX177881.1\_f0
UFBA\_2021\_POR                                JX177921.1\_f0
JX177907.1\_f0
NIWA\_28586                    JX177978.1\_f0  JX177953.1\_f0
JX177918.1\_f0
LB\_664                                       JX177928.1\_f0
JX177873.1\_f0
QMG\_320270                    JX177963.1\_f0  JX177931.1\_f0
HM032741.1\_f0
QMG\_318785                    JX177985.1\_f0
HM032752.3\_f0
NIWA\_25206                    JX177981.1\_f0
JX177917.1\_f0
QMG\_320216                    JX177966.1\_f0
JX177902.1\_f0
MHNM\_16194                    HM629803.1\_f0  JX177941.1\_f0
JX177905.1\_f0
LB\_45                                        JX177934.1\_f0
JX177884.1\_f0
TAU\_25529                     JX177970.1\_f0  JX177939.1\_f0
JX177906.1\_f0
LB\_647                                       JX177937.1\_f0
JX177879.1\_f0
LB\_1756                                      JX177933.1\_f0
JX177886.1\_f0
MNRJ\_576                                     JX177957.1\_f0
HM032742.1\_f0
NIWA\_28524                    JX177976.1\_f0  JX177945.1\_f0
JX177895.1\_f0
TAU\_25622                     JX177962.1\_f0
HM032739.1\_f0
TAU\_25621                     JX177964.1\_f0
HM032740.1\_f0
TAU\_25620                                    JX177926.1\_f0
JX177900.1\_f0
QMG\_316342                    JX177983.1\_f0  JX177955.1\_f0
HM032747.2\_f0
TAU\_25568                     JX177969.1\_f0  JX177940.1\_f0
JX177904.1\_f0
NIWA\_28929                                   JX177951.1\_f0
JX177863.1\_f0
NIWA\_28957                                   JX177949.1\_f0
JX177867.2\_f0
QMG316372                     HE591469.1\_f0
HM032748.2\_f0
UFBA\_2586\_POR                                JX177958.1\_f0
JX177898.1\_f0
ZMBN\_85240                                   HM592754.1\_f0
HM592668.1\_f0
NIWA\_28496                                   JX177946.1\_f0
JX177897.1\_f0
Cinachyrella\_kuekenthali      KC762708.1\_f0
NC\_010198.1\_f0
NIWA\_36097                                   JX177944.1\_f0
JX177866.1\_f0
LB\_671                        JX177972.1\_f0  JX177923.1\_f0
JX177893.1\_f0
NIWA\_52077                                   JX177948.1\_f0
JX177916.1\_f0
QMG\_320636                    JX177971.1\_f0
HM032745.1\_f0
QMG\_321405                    JX177968.1\_f0  JX177930.1\_f0
HM032743.1\_f0
QMG\_314224                                   JX177924.1\_f0
HM032744.1\_f0
Theonella\_swinhoei                           HM592820.1\_f0
HM592745.1\_f0
RMNH\_POR\_3206                                JX177925.1\_f0
JX177892.1\_f0
NIWA\_28877                    JX177977.1\_f0  JX177950.1\_f0
JX177864.2\_f0
ZMBN\_81789                                   HM592753.1\_f0
HM592667.1\_f0
Pachymatisma\_johnstonia                      HM592832.1\_f0
EF564339.1\_f0
Geodia\_neptuni                AY737635.1\_f0
AY320032.1\_f0
SAM\_S1189                                    JX177929.1\_f0
JX177910.1\_f0
RMNH\_POR\_2877                                JX177920.1\_f0
JX177909.1\_f0

\end{alltt}

            \end{InvisibleVerbatim}
            
        
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}38{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{AlignIO}\PY{o}{.}\PY{n}{write}\PY{p}{(}\PY{n}{pj}\PY{o}{.}\PY{n}{fta}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{combined}\PY{l+s}{\PYZsq{}}\PY{p}{)}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{combined.phy}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{phylip\PYZhy{}sequential}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}returning trimmed alignment object combined
\end{alltt}

            \end{InvisibleVerbatim}
            
                \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-out-prompt}Out\hspace{4pt}{[}38{]}:\hspace{4pt}}\\*
                \vspace{-2.55\baselineskip}\begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}1\end{alltt}

            \end{InvisibleVerbatim}
            
        
    
And now a tree can be built:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}39{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{raxml\PYZus{}method\PYZus{}concat} \PY{o}{=} \PY{n}{RaxmlConf}\PY{p}{(}\PY{n}{pj}\PY{p}{,} \PY{n}{method\PYZus{}name}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{fD\PYZus{}fb\PYZus{}combined}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                                \PY{n}{program\PYZus{}name}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{raxmlHPC\PYZhy{}PTHREADS\PYZhy{}SSE3}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                                \PY{n}{preset} \PY{o}{=} \PY{l+s}{\PYZsq{}}\PY{l+s}{fD\PYZus{}fb}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{n}{alns}\PY{o}{=}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{combined}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
                                \PY{n}{model}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{GAMMA}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{n}{matrix}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{JTT}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{n}{threads}\PY{o}{=}\PY{l+m+mi}{4}\PY{p}{,}
                                \PY{n}{cline\PYZus{}args}\PY{o}{=}\PY{p}{\PYZob{}}\PY{l+s}{\PYZsq{}}\PY{l+s}{\PYZhy{}N}\PY{l+s}{\PYZsq{}}\PY{p}{:} \PY{l+m+mi}{1}\PY{p}{\PYZcb{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}raxmlHPC-PTHREADS-SSE3 -f D -m PROTGAMMAJTT -n
943611417386379.33\_combined0 -q 943611417386379.33\_combined\_partfile
-p 747 -s 943611417386379.33\_combined.fasta -T 4 -N 1
raxmlHPC-PTHREADS-SSE3 -f b -m PROTGAMMAJTT -n
943611417386379.33\_combined1 -q 943611417386379.33\_combined\_partfile
-p 20 -s 943611417386379.33\_combined.fasta -t
RAxML\_bestTree.943611417386379.33\_combined0 -T 4 -z
RAxML\_rellBootstrap.943611417386379.33\_combined0
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}40{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{tree}\PY{p}{(}\PY{p}{[}\PY{n}{raxml\PYZus{}method\PYZus{}concat}\PY{p}{]}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
\subsection{Figures with heatmaps}\label{figures-with-heatmaps}

Here, the heatmap option is used to annotate the OTUs with their
morphological character states, based on the metadata we added earlier

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}41{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{clear\PYZus{}tree\PYZus{}annotations}\PY{p}{(}\PY{p}{)} \PY{c}{\PYZsh{}We first need to clear the previous annotations}

\PY{n}{supports} \PY{o}{=} \PY{p}{\PYZob{}}\PY{l+s}{\PYZsq{}}\PY{l+s}{black}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+m+mi}{100}\PY{p}{,}\PY{l+m+mi}{99}\PY{p}{]}\PY{p}{,}
            \PY{l+s}{\PYZsq{}}\PY{l+s}{dimgray}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+m+mi}{99}\PY{p}{,}\PY{l+m+mi}{75}\PY{p}{]}\PY{p}{,}
            \PY{l+s}{\PYZsq{}}\PY{l+s}{silver}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+m+mi}{75}\PY{p}{,}\PY{l+m+mi}{50}\PY{p}{]}\PY{p}{\PYZcb{}}

\PY{n}{pj}\PY{o}{.}\PY{n}{annotate}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{.}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{genus}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Astrophorida}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{organism}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,} \PY{n}{html}\PY{o}{=}\PY{l+s}{\PYZdq{}}\PY{l+s}{figures}\PY{l+s}{\PYZdq{}}\PY{p}{,}
            \PY{n}{heat\PYZus{}map\PYZus{}meta} \PY{o}{=} \PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{porocalyx}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{cortex}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{calthrops}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,}
            \PY{n}{heat\PYZus{}map\PYZus{}colour\PYZus{}scheme} \PY{o}{=} \PY{l+m+mi}{1}\PY{p}{,}
            \PY{n}{node\PYZus{}support\PYZus{}dict}\PY{o}{=}\PY{n}{supports}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
Below is an example of how this presence absence matrix would look:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}42{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{c}{\PYZsh{}execute if figure is small}
\PY{k+kn}{from} \PY{n+nn}{IPython.display} \PY{k+kn}{import} \PY{n}{Image}
\PY{n}{Image}\PY{p}{(}\PY{n}{filename}\PY{o}{=}\PY{l+s}{\PYZsq{}}\PY{l+s}{example\PYZus{}heatmap.png}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{n}{width}\PY{o}{=}\PY{l+m+mi}{400}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-out-prompt}Out\hspace{4pt}{[}42{]}:\hspace{4pt}}\\*
                \vspace{-2.55\baselineskip}\begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
    \begin{center}
    \includegraphics[max size={\textwidth}{\textheight}]{Tutorial_files/Tutorial_87_0.png}
    \par
    \end{center}
    
            \end{InvisibleVerbatim}
            
        
    
\subsection{Figures with clade background
colours}\label{figures-with-clade-background-colours}

Here we use the \texttt{genus} qualifier we added to color the
background of the tree clades:

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}43{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj}\PY{o}{.}\PY{n}{clear\PYZus{}tree\PYZus{}annotations}\PY{p}{(}\PY{p}{)}

\PY{n}{supports} \PY{o}{=} \PY{p}{\PYZob{}}\PY{l+s}{\PYZsq{}}\PY{l+s}{black}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+m+mi}{100}\PY{p}{,}\PY{l+m+mi}{99}\PY{p}{]}\PY{p}{,}
            \PY{l+s}{\PYZsq{}}\PY{l+s}{dimgray}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+m+mi}{99}\PY{p}{,}\PY{l+m+mi}{75}\PY{p}{]}\PY{p}{,}
            \PY{l+s}{\PYZsq{}}\PY{l+s}{silver}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{p}{[}\PY{l+m+mi}{75}\PY{p}{,}\PY{l+m+mi}{50}\PY{p}{]}\PY{p}{\PYZcb{}}

\PY{n}{genera\PYZus{}colors} \PY{o}{=} \PY{p}{\PYZob{}}\PY{l+s}{\PYZsq{}}\PY{l+s}{Tetilla}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{l+s}{\PYZsq{}}\PY{l+s}{purple}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                 \PY{l+s}{\PYZsq{}}\PY{l+s}{Cinachyra}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{l+s}{\PYZsq{}}\PY{l+s}{steelblue}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                 \PY{l+s}{\PYZsq{}}\PY{l+s}{Cinachyrella}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{l+s}{\PYZsq{}}\PY{l+s}{crimson}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                 \PY{l+s}{\PYZsq{}}\PY{l+s}{Craniella}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{l+s}{\PYZsq{}}\PY{l+s}{royalblue}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                 \PY{l+s}{\PYZsq{}}\PY{l+s}{Paratetilla}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{l+s}{\PYZsq{}}\PY{l+s}{darkred}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                 \PY{l+s}{\PYZsq{}}\PY{l+s}{Fangophilina}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{l+s}{\PYZsq{}}\PY{l+s}{mediumslateblue}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                 \PY{l+s}{\PYZsq{}}\PY{l+s}{Amphitethya}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{l+s}{\PYZsq{}}\PY{l+s}{firebrick}\PY{l+s}{\PYZsq{}}\PY{p}{,}
                 \PY{l+s}{\PYZsq{}}\PY{l+s}{Acanthotetilla}\PY{l+s}{\PYZsq{}}\PY{p}{:}\PY{l+s}{\PYZsq{}}\PY{l+s}{rosybrown}\PY{l+s}{\PYZsq{}}
                 \PY{p}{\PYZcb{}}



\PY{n}{pj}\PY{o}{.}\PY{n}{annotate}\PY{p}{(}\PY{l+s}{\PYZsq{}}\PY{l+s}{.}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{genus}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{l+s}{\PYZsq{}}\PY{l+s}{Astrophorida}\PY{l+s}{\PYZsq{}}\PY{p}{,}\PY{p}{[}\PY{l+s}{\PYZsq{}}\PY{l+s}{organism}\PY{l+s}{\PYZsq{}}\PY{p}{]}\PY{p}{,} \PY{n}{html}\PY{o}{=}\PY{l+s}{\PYZdq{}}\PY{l+s}{figures}\PY{l+s}{\PYZdq{}}\PY{p}{,}
            \PY{n}{node\PYZus{}bg\PYZus{}meta}\PY{o}{=}\PY{l+s}{\PYZdq{}}\PY{l+s}{genus}\PY{l+s}{\PYZdq{}}\PY{p}{,}
            \PY{n}{node\PYZus{}bg\PYZus{}color}\PY{o}{=}\PY{n}{genera\PYZus{}colors}\PY{p}{,}
            \PY{n}{node\PYZus{}support\PYZus{}dict}\PY{o}{=}\PY{n}{supports}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    
To see how this annotation will look open the \texttt{figures.html} file
in the \texttt{Tutorial\_files} direcoty and use the links to the
images.\subsection{Archiving the results}\label{archiving-the-results}As a final step, ReproPhylo conveniently archives essential outputs: 1.
an html report describing the methods in human-readable format 2. a
nexml file containing the sequence alignments and trees 3. a pickled
\texttt{Project} file to easily resusitate a ReproPhylo experiment 4. a
genbank file containing the records that were used in the analysis with
the metadata changes that were conducted 5. all the tree figures.

    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}44{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{publish}\PY{p}{(}\PY{n}{pj}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{Tutorial\PYZus{}results}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{.}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}checking if file exists



[master cff06f0] Records nexml text file from Sun Nov 30 22:26:54 2014
 2 files changed, 12748 insertions(+), 55 deletions(-)
 create mode 100644 Tutorial\_results/tree\_and\_alns.nexml




[master 15dcae2] Records genbank text file from Sun Nov 30 22:26:54
2014
 1 file changed, 14832 insertions(+)
 create mode 100644 Tutorial\_results/sequences\_and\_metadata.gb

reporter was called by publish
now printing species table
now making sequence statistics plots
now reporting concatenations
now reporting methods
now reporting alignment statistics
making RF matrix
reporting trees
pickling

[master 14e6313] A pickled Project from Sun Nov 30 22:26:57 2014
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 Tutorial\_results/Sun\_30\_Nov\_2014\_22:26:57.pkl

archiving
report ready
\end{alltt}

            \end{InvisibleVerbatim}
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}/usr/local/lib/python2.7/dist-packages/pandas/io/excel.py:626:
UserWarning: Installed openpyxl is not supported at this time. Use
>=1.6.1 and <2.0.0.
  .format(openpyxl\_compat.start\_ver, openpyxl\_compat.stop\_ver))
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}45{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{pj} \PY{o}{=} \PY{n}{unpickle\PYZus{}pj}\PY{p}{(}\PY{l+s}{\PYZdq{}}\PY{l+s}{Tutorial\PYZus{}results/Sun\PYZus{}30\PYZus{}Nov\PYZus{}2014\PYZus{}22:26:57.pkl}\PY{l+s}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}46{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{publish}\PY{p}{(}\PY{n}{pj}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{Tutorial\PYZus{}results\PYZus{}test\PYZus{}unpickled}\PY{l+s}{\PYZsq{}}\PY{p}{,} \PY{l+s}{\PYZsq{}}\PY{l+s}{.}\PY{l+s}{\PYZsq{}}\PY{p}{)}
\end{Verbatim}

            
                \vspace{-0.2\baselineskip}
            
        \end{ColorVerbatim}
    

    

        % If the first block is an image, minipage the image.  Else
        % request a certain amount of space for the input text.
        \needspace{4\baselineskip}
        
        

            % Add document contents.
            
                \begin{InvisibleVerbatim}
                \vspace{-0.5\baselineskip}
\begin{alltt}checking if file exists



[master 46f83ec] Records nexml text file from Sun Nov 30 22:27:43 2014
 2 files changed, 12753 insertions(+), 27 deletions(-)
 create mode 100644
Tutorial\_results\_test\_unpickled/tree\_and\_alns.nexml




[master c9817e5] Records genbank text file from Sun Nov 30 22:27:44
2014
 1 file changed, 14832 insertions(+)
 create mode 100644
Tutorial\_results\_test\_unpickled/sequences\_and\_metadata.gb

reporter was called by publish
now printing species table
now making sequence statistics plots
now reporting concatenations
now reporting methods
now reporting alignment statistics
making RF matrix
reporting trees
pickling

[master 4cf8e11] A pickled Project from Sun Nov 30 22:27:47 2014
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644
Tutorial\_results\_test\_unpickled/Sun\_30\_Nov\_2014\_22:27:46.pkl

archiving
report ready
\end{alltt}

            \end{InvisibleVerbatim}
            
        
    


    % Make sure that atleast 4 lines are below the HR
    \needspace{4\baselineskip}

    
        \vspace{6pt}
        \makebox[0.1\linewidth]{\smaller\hfill\tt\color{nbframe-in-prompt}In\hspace{4pt}{[}{]}:\hspace{4pt}}\\*
        \vspace{-2.65\baselineskip}
        \begin{ColorVerbatim}
            \vspace{-0.7\baselineskip}
            \begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}

            
                \vspace{0.3\baselineskip}
            
        \end{ColorVerbatim}
    

        

        \renewcommand{\indexname}{Index}
        \printindex

    % End of document
    \end{document}


