This repository was automatically created and 
maintained by [ReproPhylo V1](http://hulluni-bioinformatics.github.io/ReproPhylo/). 
It contains the analysis done for the paper *ReproPhylo: a python package 
 for reproducible phylogenomics* by Amir Szitenberg, Max John, Mark Blaxter and David Lunt.  

The data files in `data`, the ipython notebook and the pickled project `szitenberg_et_al_reprophylo` were all versioned and commited automatically as the foundation of the experiments reproducibility. The remaining files are manuscript related and where added manually later on.  
All the figures in the repository were created by the workflow, as can be seen in the IPython Notebook, or in its static copy, MethodsS1.html.
  
The commits list in this repository demonstrates the way in which ReproPhylo commits changes to the repo and the automatically generated commit massages. Some are very informative, such as [this one](https://github.com/HullUni-bioinformatics/Szitenberg_et_all_15-ReproPhylo/commit/6c085d736c9e9b9bfe8d3cc3b48e4c663ec8e190), related to a commit of a tree reconstruction step.