{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Introduction\n",
      "This tool will allow most of the functions that do not involve changing the default options in `Conf` objects. It can be used several times in the workflow, and it can combine several tasks in one go.\n",
      "\n",
      "## inputs\n",
      "\n",
      "1. Optional Genbank file\n",
      "2. and/or `Locus` CSV file\n",
      "3. and/or `Project` pickle file\n",
      "4. and/or metadata CSV file\n",
      "5. and/or a fasta file\n",
      "  \n",
      "## options (not mutually exclusive)\n",
      "\n",
      "1. [Read Genbank](#ReadGenbank)\n",
      "   - Add to existing records\n",
      "   - Replace existing records\n",
      "2. [Write a `Locus` CSV file](#WriteLocusCsv)\n",
      "3. [Read a `Locus` CSV file](#LocusCsvFile)\n",
      "   - Add to existing loci\n",
      "   - Replace exisiting loci\n",
      "4. [Read fasta file](#FastaFile)\n",
      "   - Add to exisiting records\n",
      "   - Replace existing methods\n",
      "   - Set a gene to the whole file (optional)\n",
      "   - Set a feature qualifier for the filename (optional)\n",
      "   - Set a feature qualifier for seq.ids (optional)\n",
      "   - Set a feature qualifier for seq.descriptions (optional)\n",
      "5. [Write a metadata CSV](#MetadataCSV)\n",
      "6. [Read a metadata CSV](#MetadataCSV)\n",
      "7. [Set program paths](#DefaultPrograms)\n",
      "8. [Run](#RunWithDefaultConfObjects)\n",
      "   - Replace exisitng analysis\n",
      "   - Add to existing analysis\n",
      "9. [Report](#Report)\n",
      "  \n",
      "  \n",
      "### <a name=\"ReadGenbank\"> Read Genbank: </a>\n",
      "When a Genbank file is read, a `Locus` CSV will be created, a `pj = Project` will be started with this file and the records will be processed and put in `pj.records`. The records will either wipe or be appended to preexisting records.\n",
      "### <a name=\"WriteLocusCsv\"> Write a Locus CSV file: </a>\n",
      "The file will have synonymized lines. The user will be able to change the `char_type`(`\"dna\"` or `\"prot\"`, first value in each line), They will also be able to delete lines in order to ignore these loci, and also to further synonymize lines by adding a comma followed by shared integers at the end of lines that are the same locus.\n",
      "### <a name=\"LocusCsvFile\"> Read a Locus CSV file: </a>  \n",
      "Reads an edited or manually written `Locus` CSV file into the `Project`. The user can choose if to add them to the loci in the `Project` or to wipe. If they choose to add, the records, if any, will be reassessed and amended. This can be run together with 'Read Genbank' and then the records from the Genbank file will wipe/ be added to existing records, if any.\n",
      "### <a name=\"FastaFile\"> Read a fasta file: </a>\n",
      "By default, the records will be put in `pj.records` with the `seq.id` as the value of the 'original_id' feature qualifier, `seq.description` of 'original_desc' and the file name as 'original_filename'. The user can choose to a. set a gene qualifier for the whole file, b. give qualifier names to the file name, the `seq.id` and the `seq.description`. This allows basic batch metadata composition for whole files. Additional/ alternative annotation can be done by reading the files, printing a metadata CSV, editing it and reading it back.\n",
      "### <a name=\"MetadataCSV\"> Write and Read a metadata CSV: </a>\n",
      "The user can choose to print a metadata CSV file, edit it and read it back to edit the records metadata as contained in feature qualifiers and in the taxonomy line. Either writing or reading a file can be coupled with reading sequences.\n",
      "### <a name=\"DefaultPrograms\"> Set program paths: </a>  \n",
      "Reprophylo assumes that all the programs are executable and are in the path. This option allows specifying an alternative path to each of the programs. Prefixes such as 'perl' can be included here as well. \n",
      "### <a name=\"RunWithDefaultConfObjects\"> Run: </a> \n",
      "A set of default `Conf` objects are set in the `Project`. They will produce all the gene trees for loci that pass the filters. A different set of galaxy tools will be used to configure `Concatenation` and `Conf` objects and add them to the `Project`. A `Project` can be ran several times, in each the user will choose whether to add the new analysis or to wipe the old ones.\n",
      "### <a name=\"Report\"> Report: </a>  \n",
      "Prints a report of whatever components that exist in the `Project` at that given moment."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%file base_reprophylo.py\n",
      "#######################################################################\n",
      "#                        COMMAND LINE PARSING\n",
      "#######################################################################\n",
      "import argparse as agp\n",
      "\n",
      "desc = (\"Base reprophylo analysis. Deals with data and \"+\n",
      "        \"metadata and runs default analyses. If the Project was \"+\n",
      "        \"reconfigured with other tools, \"+\n",
      "        \"it will run the reconfigured \"+\n",
      "        \"analysis\")\n",
      "parser = agp.ArgumentParser(description=desc)\n",
      "\n",
      "#----------------------------------------------------------------------\n",
      "parser.add_argument(\"projfile\",\n",
      "                    help=\"A pickle file to read and write\",\n",
      "                    type=str,\n",
      "                    default=None)\n",
      "parser.add_argument(\"-a\", \"--altprojfile\",\n",
      "                    help=\"An alternative pickle file to write\",\n",
      "                    type=str, \n",
      "                    default=None)\n",
      "parser.add_argument(\"-g\", \"--genbank\",\n",
      "                    help=\"An input genbank file name\",\n",
      "                    type=str,\n",
      "                    default=None)\n",
      "parser.add_argument(\"-m\", \"--dump_alignments\",\n",
      "                    help=\"print sequene alignment in the specified format\",\n",
      "                    type=str,\n",
      "                    default=None)\n",
      "parser.add_argument(\"--dump_alignments_zip\",\n",
      "                    type=str,\n",
      "                    default=\"alignments.zip\")\n",
      "parser.add_argument(\"-p\", \"--dump_records\",\n",
      "                    help=\"print a genbank file of all records and metadata\",\n",
      "                    type=str,\n",
      "                    default=None)\n",
      "parser.add_argument(\"-T\", \"--threads\",\n",
      "                    help=(\"Number of threads to run in parrallel \"+\n",
      "                    \"components of the pipeline\"),\n",
      "                    type=int,\n",
      "                    default=4)\n",
      "parser.add_argument(\"-j\", \"--figs\",\n",
      "                    help=\"A directory to keep temp and final fig files\",\n",
      "                    type=str, \n",
      "                    default='.')\n",
      "parser.add_argument(\"-_\", \"--__\",\n",
      "                    help=(\"Turn off warnings for Galaxy\"),\n",
      "                    action=\"store_true\")\n",
      "parser.add_argument(\"--noninteractive\",\n",
      "                    help=(\"Overwrite reports without asking\"),\n",
      "                    action=\"store_true\")\n",
      "#----------------------------------------------------------------------\n",
      "parser.add_argument(\"-f\", \"--fasta\",\n",
      "                    help=\"An input fasta file name\",\n",
      "                    type=str, nargs='+',\n",
      "                    default=None)\n",
      "parser.add_argument(\"-t\", \"--fasta_chartype\",\n",
      "                    help=(\"Specify the type of sequence in all the\"+\n",
      "                    \" fasta files [dna|prot]. Required with --fasta\"),\n",
      "                    type=str,\n",
      "                    choices=['dna', 'prot'])# Only with --fasta\n",
      "parser.add_argument(\"-n\", \"--feature\",\n",
      "                    help=(\"Set the gene name and feature type\"+\n",
      "                          \" for all the sequences in\"+\n",
      "                          \" fasta file\"),\n",
      "                    type=str, nargs='+',\n",
      "                    default=None)#Only with -f one_filename. \n",
      "parser.add_argument(\"-l\", \"--featureis_fasname\",\n",
      "                    help=(\"Set the file name as the gene name and \"+\n",
      "                          \"feature type (eg cox1@CDS)\"+\n",
      "                          \"for all the sequence in that\"+\n",
      "                          \" fasta file\"),\n",
      "                    action=\"store_true\")# only with -f, will apply to all fas\n",
      "parser.add_argument(\"-q\", \"--qualis_fasname\",\n",
      "                    help=(\"Set a feature qualifier name contain\"+\n",
      "                          \"ing the fasta file name for all \"+\n",
      "                          \"the sequences in that file\"),\n",
      "                    type=str,\n",
      "                    default=None)#Only with -f, will apply to all fas\n",
      "parser.add_argument(\"-i\", \"--qualis_seqid\",\n",
      "                    help=(\"Set a feature qualifier name contain\"+\n",
      "                          \"ing the sequence ID for all \"+\n",
      "                          \"the sequences in the file\"),\n",
      "                    type=str,\n",
      "                    default=None)#Only with -f, will apply to all fas\n",
      "parser.add_argument(\"-d\", \"--qualis_seqdesc\",\n",
      "                    help=(\"Set a feature qualifier name contain\"+\n",
      "                          \"ing the sequence description for all \"+\n",
      "                          \"the sequences in the file\"),\n",
      "                    type=str,\n",
      "                    default=None)#Only with -f, will apply to all fas\n",
      "parser.add_argument(\"-e\", \"--expng_seqandmeta\",\n",
      "                    help=(\"Replace existing sequence records with\"+\n",
      "                          \"new ones from gb and/or fasta file(s)\"),\n",
      "                    action=\"store_true\")# Only with -g and or -f\n",
      "#----------------------------------------------------------------------\n",
      "loci = parser.add_mutually_exclusive_group()\n",
      "loci.add_argument(\"-w\", \"--write_loci_csv\",\n",
      "                   help=(\"Write a loci CSV file. It can be \"+\n",
      "                         \"edited and read back\"),\n",
      "                   type=str,\n",
      "                   default=None)\n",
      "loci.add_argument(\"-r\", \"--read_loci_csv\",\n",
      "                   help=(\"Read loci from an edited CSV file. \"),\n",
      "                   type=str,\n",
      "                   default=None)\n",
      "loci.add_argument(\"-u\", \"--add_locus\",\n",
      "                    help=(\"Add a locus to the analysis. \"+\n",
      "                          \"Takes three space delimited values: \"+\n",
      "                          \"[dna|prot] [CDS|rRNA|gene|...] locus_name\"),\n",
      "                    type=str, nargs=3,\n",
      "                    default=None)\n",
      "parser.add_argument(\"-b\", \"--banish_loci\",\n",
      "                    help=(\"Replace existing loci with\"+\n",
      "                          \"new ones from a loci CSV file\"),\n",
      "                    action=\"store_true\")# Only with -r\n",
      "#----------------------------------------------------------------------\n",
      "meta = parser.add_mutually_exclusive_group()\n",
      "meta.add_argument(\"-y\", \"--write_meta_csv\",\n",
      "                   help=(\"Write a metadata CSV file. It can be \"+\n",
      "                         \"edited and read back\"),\n",
      "                   type=str,\n",
      "                   default=None)\n",
      "meta.add_argument(\"-z\", \"--read_meta_csv\",\n",
      "                   help=(\"Read metadata from an edited CSV file\"),\n",
      "                   type=str,\n",
      "                   default=None)\n",
      "#----------------------------------------------------------------------\n",
      "parser.add_argument(\"-s\", \"--start\",\n",
      "                   help=(\"Start the analysis\"),\n",
      "                   action=\"store_true\")\n",
      "parser.add_argument(\"-x\", \"--overwrite\",\n",
      "                    help=(\"Replace existing results with the ones\"+\n",
      "                          \" of current configuration\"),\n",
      "                    action=\"store_true\")# Only with -s\n",
      "parser.add_argument(\"-v\", \"--verbose_report\",\n",
      "                   help=(\"Write full html report to filename\"),\n",
      "                   type=str,\n",
      "                   default=None)\n",
      "parser.add_argument(\"-c\", \"--version_ctrl\",\n",
      "                    help=(\"perform version control in CWD\"),\n",
      "                    action=\"store_true\")\n",
      "#----------------------------------------------------------------------\n",
      "cl = parser.parse_args()\n",
      "#----------------------------------------------------------------------\n",
      "# At least one of these is required\n",
      "if len([x for x in (cl.dump_alignments,\n",
      "                    cl.dump_records,\n",
      "                    cl.genbank,\n",
      "                    cl.fasta,\n",
      "                    cl.add_locus,\n",
      "                    cl.write_loci_csv,\n",
      "                    cl.read_loci_csv,\n",
      "                    cl.write_meta_csv,\n",
      "                    cl.read_meta_csv,\n",
      "                    cl.start,\n",
      "                    cl.verbose_report) if ((x is not None) and  \n",
      "                                  (x is not False))])== 0:\n",
      "    parser.error('One of the following is required:\\n'+\n",
      "                 '--genbank\\n--fasta\\n--write_loci_csv\\n'+\n",
      "                 '--read_loci_csv\\n--write_meta_csv\\n'+\n",
      "                 '--read_meta_csv\\n--start\\nverbose_report\\n'+\n",
      "                 '--dump_alignments\\n--dump_alignments\\n'+\n",
      "                 '--dump_records')\n",
      "#----------------------------------------------------------------------\n",
      "# The following go only with --fasta:\n",
      "if not cl.fasta and any([cl.feature,\n",
      "                         cl.featureis_fasname,\n",
      "                         cl.qualis_fasname,\n",
      "                         cl.qualis_seqid,\n",
      "                         cl.qualis_seqdesc,\n",
      "                         cl.fasta_chartype]):\n",
      "    parser.error('The following only work with --fasta:\\n'+\n",
      "                 '--name_gene\\n--gennameis_fasname\\n'+\n",
      "                 '--qualis_fasname\\n--fasta_chartype\\n'+\n",
      "                 '--qualis_seqid\\n--qualis_seqdesc')\n",
      "# Gene name qualifier can be passed through cline only for a single\n",
      "# fasta file\n",
      "if cl.fasta and len(cl.fasta) > 1 and cl.feature:\n",
      "    parser.error('--name_gene only works with one --fasta file')\n",
      "# When reading fasta(s) the sequence type needs to be passed as well:\n",
      "if cl.fasta and not cl.fasta_chartype:\n",
      "    parser.error('Must specify the type of sequence in the fasta '+\n",
      "                 'file(s) with --fasta_chartype [dna|prot]')\n",
      "if not (cl.feature == None or len(cl.feature)==2):\n",
      "    parser.error('--feature takes two space delimited strings to '+\n",
      "                 'represent the gene name and feature type of the '+\n",
      "                 'fasta sequences')\n",
      "#----------------------------------------------------------------------\n",
      "# Overwrite only when analysis is run:\n",
      "if cl.overwrite and not cl.start:\n",
      "    parser.error('--overwrite requires --start')\n",
      "#----------------------------------------------------------------------\n",
      "# Banish Locus objects only when reading new ones\n",
      "if cl.banish_loci and not (cl.read_loci_csv or\n",
      "                           cl.genbank or cl.add_locus):\n",
      "    parser.error('--banish_loci only when reading new ones with '+\n",
      "                 '--read_loci_csv or --genbank or --add_locus')\n",
      "# Expunge SeqRecord objects only when reading new ones\n",
      "if cl.expng_seqandmeta and not cl.genbank and not cl.fasta:\n",
      "    parser.error('--expng_seqandmeta only when reading new ones with '+\n",
      "                 '--genbank or --fasta')\n",
      "#----------------------------------------------------------------------\n",
      "#if False:\"\"\"    \n",
      "#######################################################################\n",
      "#                       Functions\n",
      "#######################################################################\n",
      "from reprophylo import *\n",
      "\n",
      "def random_filename():\n",
      "    return str(random.randint(10000,99999))+str(time.time())\n",
      "\n",
      "def add_locus_from_cline():\n",
      "    return Locus(cl.add_locus[0],cl.add_locus[1],\n",
      "                 cl.add_locus[2],[cl.add_locus[2]])\n",
      "\n",
      "# Functions to start a project and deal with a genbenk file and loci\n",
      "#----------------------------------------------------------------------\n",
      "def make_project(loci_csv=None, genbank=None):\n",
      "    # If the project exists, modify the loci if requested:\n",
      "    if os.path.isfile(cl.projfile) and os.path.getsize(cl.projfile)>10:\n",
      "        pj = unpickle_pj(cl.projfile)\n",
      "        # remove the old loci if requested\n",
      "        if cl.banish_loci:\n",
      "            pj.loci = []\n",
      "        if cl.expng_seqandmeta: # We trust the cline parser to break\n",
      "                                # if no new records are \n",
      "                                # supplied instead\n",
      "            pj.records = []\n",
      "            pj.records_by_locus = {}\n",
      "            pj.concatenations = []\n",
      "            pj.alignments = {}\n",
      "            pj.trimmed_alignments = {}\n",
      "            pj.trees = {}\n",
      "        # Add a locus form the cline\n",
      "        if cl.add_locus:\n",
      "            new_locus = add_locus_from_cline()\n",
      "            if ((not new_locus.name in [x.name for x in pj.loci])and\n",
      "                not any(new_locus.name in x.aliases for x in pj.loci)):   \n",
      "                pj.loci += [add_locus_from_cline()]\n",
      "            else:\n",
      "                raise IOError(\"%s already in the Project loci\"%\n",
      "                              cl.add_locus[2])\n",
      "        # or add loci from a file\n",
      "        elif loci_csv:\n",
      "            new_loci_list = loci_list_from_csv(loci_csv)\n",
      "            for l in new_loci_list:\n",
      "                name_and_aliases = [l.name]+l.aliases\n",
      "                for old in pj.loci:\n",
      "                    if any(n in [old.name] + old.aliases \n",
      "                           for n in name_and_aliases):\n",
      "                        raise IOError(\"%s already in the Project loci\"%\n",
      "                                      l.name)\n",
      "            pj.loci += new_loci_list\n",
      "            if genbank:\n",
      "                pj.read_embl_genbank([genbank])\n",
      "        # if genbank is read and no loci from file or cline, take loci\n",
      "        # from genbank.\n",
      "        elif genbank and not loci_csv and not cl.add_locus and pj.loci == []:\n",
      "            loci_csv_out = random_filename()\n",
      "            loci_readable_report = random_filename()\n",
      "            # write the file if requested by cline\n",
      "            if cl.write_loci_csv:\n",
      "                loci_csv_out = cl.write_loci_csv\n",
      "            list_loci_in_genbank(genbank, loci_csv_out,\n",
      "                                 loci_report = loci_readable_report)\n",
      "            print open(loci_readable_report,'r').readlines()[1]\n",
      "            print ('No loci in the Project, reading'+\n",
      "                   ' from genbank file')\n",
      "            print '-------------------------------'\n",
      "            print 'Gene and count sorted by counts'\n",
      "            print '-------------------------------'\n",
      "            handle = open(loci_readable_report,'r')\n",
      "            print handle.read().rpartition('----')[-1]\n",
      "            os.remove(loci_readable_report)\n",
      "            # make aproject using loci in the gb file\n",
      "            new_loci_list = loci_list_from_csv(loci_csv_out)\n",
      "            for l in new_loci_list:\n",
      "                name_and_aliases = [l.name]+l.aliases\n",
      "                for old in pj.loci:\n",
      "                    if any(n in [old.name] + old.aliases for n in name_and_aliases):\n",
      "                        raise IOError(\"%s already in the Project loci\"%l.name)\n",
      "            pj.loci += loci_list_from_csv(loci_csv_out)\n",
      "            pj.read_embl_genbank([genbank])\n",
      "            # delete loci file if not requested to keep by cline\n",
      "            if not cl.write_loci_csv:\n",
      "                os.remove(loci_csv_out)\n",
      "        elif genbank and not loci_csv and not cl.add_locus and len(pj.loci)>0:\n",
      "            pj.read_embl_genbank([genbank])\n",
      "        elif cl.write_loci_csv:\n",
      "            handle = open(cl.write_loci_csv,'wt')\n",
      "            for l in pj.loci:\n",
      "                handle.write(\"%s,%s,%s\"%(l.char_type,l.feature_type,l.name))\n",
      "                for a in l.aliases:\n",
      "                    handle.write(\",%s\"%a)\n",
      "                handle.write('\\n')\n",
      "            handle.close()\n",
      "    # if the project does not exist:\n",
      "    else:\n",
      "        # if a locus is passed through the cline:\n",
      "        if cl.add_locus:\n",
      "            pj = Project([add_locus_from_cline()])\n",
      "        # if loci passed through a CSV file\n",
      "        elif cl.read_loci_csv:\n",
      "            pj = Project(loci_csv)\n",
      "            if genbank:\n",
      "                pj.read_embl_genbank([genbank])\n",
      "        elif genbank and not loci_csv and not cl.add_locus:\n",
      "            loci_csv_out = random_filename()\n",
      "            loci_readable_report = random_filename()\n",
      "            # write the file if requested by cline\n",
      "            if cl.write_loci_csv:\n",
      "                loci_csv_out = cl.write_loci_csv\n",
      "            list_loci_in_genbank(genbank, loci_csv_out,\n",
      "                                 loci_report = loci_readable_report)\n",
      "            print open(loci_readable_report,'r').readlines()[1]\n",
      "            print '-------------------------------'\n",
      "            print 'Gene and count sorted by counts'\n",
      "            print '-------------------------------'\n",
      "            handle = open(loci_readable_report,'r')\n",
      "            print handle.read().rpartition('----')[-1]\n",
      "            os.remove(loci_readable_report)\n",
      "            # make aproject using loci in the gb file\n",
      "            pj = Project(loci_csv_out)\n",
      "            pj.read_embl_genbank([genbank])\n",
      "            # delete loci file if not requested to keep by cline\n",
      "            if not cl.write_loci_csv:\n",
      "                os.remove(loci_csv_out)\n",
      "                    \n",
      "    return pj    \n",
      "\n",
      "# Functions to read fasta files and deal with their metadata\n",
      "#----------------------------------------------------------------------\n",
      "def deal_with_fasta(pj):\n",
      "    # Read the records and make a feature based on cline instructions\n",
      "    if cl.fasta:\n",
      "        fst_dnov_in_file = 0\n",
      "        fst_dnov_in_next_file = 0\n",
      "        cline_quals = [cl.qualis_fasname,\n",
      "                       cl.qualis_seqid,\n",
      "                       cl.qualis_seqdesc]\n",
      "        for fasta in cl.fasta:\n",
      "            fst_dnov_in_next_file = pj.read_denovo([fasta],\n",
      "                                                   cl.fasta_chartype)\n",
      "            for dnov_index in range(fst_dnov_in_file,\n",
      "                                    fst_dnov_in_next_file):\n",
      "                current_record = [r for r in pj.records if\n",
      "                                  r.id == 'denovo%i'%dnov_index][0]\n",
      "                new_quals = {}\n",
      "                feature_type = None\n",
      "                if cl.feature:\n",
      "                    new_quals['gene'] = cl.feature[0]\n",
      "                    feature_type = cl.feature[1]\n",
      "                elif cl.featureis_fasname:\n",
      "                    new_quals['gene'] = fasta.split('/')[-1].split('@')[0]\n",
      "                    feature_type = fasta.split('@')[1]\n",
      "                elif ('feature_type' in cline_quals\n",
      "                      and \n",
      "                      'gene' in cline_quals):\n",
      "                    if cl.qualis_fasname == 'feature_type':\n",
      "                        feature_type = fasta\n",
      "                    elif cl.qualis_seqid  == 'feature_type':\n",
      "                        feature_type = current_record.id\n",
      "                    elif cl.qualis_seqdesc ==  'feature_type':\n",
      "                        feature_type = current_record.description\n",
      "                    if cl.qualis_fasname == 'gene':\n",
      "                        new_quals['gene'] = fasta\n",
      "                    elif cl.qualis_seqid  == 'gene':\n",
      "                        new_quals['gene'] = current_record.id\n",
      "                    elif cl.qualis_seqdesc ==  'gene':\n",
      "                        new_quals['gene'] = current_record.description\n",
      "                feat_id = None\n",
      "                if (feature_type and not new_quals == {}):\n",
      "                    rec_id = 'denovo%i'%dnov_index\n",
      "                    feat_id = pj.add_feature_to_record(rec_id,\n",
      "                                                       feature_type,\n",
      "                                                       qualifiers=new_quals)\n",
      "                if (not 'feature_type' in cline_quals\n",
      "                    or \n",
      "                    not 'gene' in cline_quals):\n",
      "                    if feat_id:\n",
      "                        for feat in current_record.features:\n",
      "                            if feat.qualifiers['feature_id'] == feat_id:\n",
      "                                requiered = ['gene','feature_type']\n",
      "                                if not cl.qualis_fasname in requiered:\n",
      "                                    feat.qualifiers[cl.qualis_fasname] == fasta\n",
      "                                if not cl.qualis_seqid in requiered:\n",
      "                                    feat.qualifiers[cl.qualis_seqid] == current_record.id\n",
      "                                if not cl.qualis_seqdesc in requiered:\n",
      "                                    feat.qualifiers[cl.qualis_seqdesc] == current_record.description\n",
      "            fst_dnov_in_file = fst_dnov_in_next_file\n",
      "\n",
      "\n",
      "def removetype(path):\n",
      "    import shutil\n",
      "    if os.path.isdir(path):\n",
      "        shutil.rmtree(path)\n",
      "    else:\n",
      "        os.remove(path)\n",
      "\n",
      "def extract_and_log_by_locus(pj):\n",
      "    print\n",
      "    print \"---------------------------------------------\"\n",
      "    print \"The project now contains the following loci:\".title()\n",
      "    print \"---------------------------------------------\"\n",
      "    if len(pj.records)>0:\n",
      "        pj.extract_by_locus()\n",
      "        print (\"Locus\".ljust(30,' ')+\"Records\".ljust(10,' ')+\n",
      "               \"Sequence length (max min mean)\")\n",
      "        for l in pj.loci:\n",
      "            seq_lengths = [len(r) for r in pj.records_by_locus[l.name]] \n",
      "            if len(seq_lengths) > 0:\n",
      "                print(str(l.name).ljust(30,' ')+\n",
      "                       (\"%i\"%len(pj.records_by_locus[l.name])).ljust(10,' ')+\n",
      "                       (\"%i\"%max(seq_lengths)).ljust(9,' ')+\n",
      "                       ('%i'%min(seq_lengths)).ljust(9,' ')+\n",
      "                       '%.1f'%np.mean(seq_lengths))\n",
      "            else:\n",
      "                print (str(l.name).ljust(30,' ')+'No records')\n",
      "        if cl.verbose_report:\n",
      "            if cl.noninteractive:\n",
      "                if os.path.exists(cl.verbose_report):\n",
      "                    removetype(cl.verbose_report)\n",
      "                if os.path.exists(cl.verbose_report.rpartition('.')[0]):\n",
      "                    removetype(cl.verbose_report.rpartition('.')[0])\n",
      "                if os.path.exists(cl.verbose_report+'.zip'):\n",
      "                    removetype(cl.verbose_report+'.zip')\n",
      "            publish(pj, cl.verbose_report, cl.figs)\n",
      "    else:\n",
      "        for l in pj.loci:\n",
      "            print l.name.ljust(30, ' ')+'No records'\n",
      "               \n",
      "              \n",
      "\n",
      "def run_default_analysis(pj):\n",
      "    if cl.start:\n",
      "        if cl.overwrite:\n",
      "            pj.concatenations = []\n",
      "            pj.alignments = {}\n",
      "            pj.trimmed_alignments = {}\n",
      "            pj.trees = {}\n",
      "            \n",
      "        # make default labels:\n",
      "        pj.add_qualifier_from_source('organism')\n",
      "        pj.copy_paste_within_feature('organism','tree_label')\n",
      "        for record in pj.records:\n",
      "            for feature in record.features:\n",
      "                if not feature.type =='source':\n",
      "                    if not 'tree_label' in feature.qualifiers.keys():\n",
      "                        feature.qualifiers['tree_label'] = feature.qualifiers['feature_id']\n",
      "        \n",
      "        print '------------------'\n",
      "        print 'Sequence alignment'.title()\n",
      "        print '------------------'\n",
      "        mafft = AlnConf(pj, method_name='MafftDefaults',\n",
      "                        CDSAlign=False, cline_args=dict())\n",
      "        pj.align([mafft])\n",
      "        print '------------------'\n",
      "        print 'Trimming alignment'.title()\n",
      "        print '------------------'        \n",
      "        gappyout = TrimalConf(pj)\n",
      "        pj.trim([gappyout])\n",
      "        print '-------------------'\n",
      "        print 'Tree reconstructios'.title()\n",
      "        print '-------------------'   \n",
      "        \n",
      "        raxml = RaxmlConf(pj, method_name='fD_fb',\n",
      "                          program_name=pj.defaults['raxmlHPC'],\n",
      "                          preset = 'fD_fb',\n",
      "                          model='GAMMA', matrix='JTT',\n",
      "                          threads=int(cl.threads))\n",
      "        \n",
      "        pj.tree([raxml])\n",
      "        \n",
      "        pj.clear_tree_annotations()\n",
      "    \n",
      "        supports = {'black':[100,99],\n",
      "                    'dimgray':[99,75],\n",
      "                    'silver':[75,50]}\n",
      "        print '---------------'\n",
      "        print 'Tree annotation'.title()\n",
      "        print '---------------' \n",
      "        \n",
      "        pj.annotate(cl.figs,\n",
      "                    root_meta='mid',\n",
      "                    root_value='mid',\n",
      "                    leaf_labels_txt_meta=['tree_label'],\n",
      "                    node_support_dict=supports,\n",
      "                    scale = 1000)\n",
      "        \n",
      "        \n",
      "def ref():\n",
      "    print '------------------'\n",
      "    print 'Program References'\n",
      "    print '------------------'\n",
      "    if cl.start:\n",
      "        print '-------------------------------'\n",
      "        print 'Exploratory Pipeline References'\n",
      "        print '-------------------------------'\n",
      "        print ('RAxML: A. Stamatakis: \"RAxML Version 8: A tool for '+\n",
      "               'Phylogenetic Analysis and Post-Analysis of Large'+\n",
      "               ' Phylogenies\". In Bioinformatics, 2014')\n",
      "        print ('http://bioinformatics.oxfordjournals.org/content/'+\n",
      "               'early/2014/01/21/bioinformatics.btu033.abstract?'+\n",
      "               'keytype=ref&ijkey=VTEqgUJYCDcf0kP')\n",
      "        print \n",
      "        print ('MAFFT: Katoh, Standley 2013 (Molecular Biology and Evolution '+\n",
      "               '30:772-780) MAFFT multiple sequence alignment software'+\n",
      "               ' version 7: improvements in performance and usability.')\n",
      "        print 'mbe.oxfordjournals.org/content/30/4/772'\n",
      "        print \n",
      "        print ('trimAl: Salvador Capella-Gutierrez; Jose M. Silla-Martinez; Toni'+\n",
      "               ' Gabaldon. trimAl: a tool for automated alignment trimming'+\n",
      "               ' in large-scale phylogenetic analyses. Bioinformatics 2009 25: '+\n",
      "               '1972-1973.')\n",
      "        print 'http://trimal.cgenomics.org/_media/trimal.2009.pdf'\n",
      "        print \n",
      "        print ('ETE: Jaime Huerta-Cepas, Joaquin Dopazo and Toni Gabaldon. '+\n",
      "               'ETE: a python Environment for Tree Exploration. BMC '+\n",
      "               'Bioinformatics 2010, 11:24.')\n",
      "        print 'http://www.biomedcentral.com/1471-2105/11/24'\n",
      "        print \n",
      "        print ('NumPy: Stefan van der Walt, S. Chris Colbert and Gael Varoquaux. '+\n",
      "               'The NumPy Array: A Structure for Efficient Numerical Computation, '+\n",
      "               'Computing in Science & Engineering, 13, 22-30 (2011)')\n",
      "        print 'DOI:10.1109/MCSE.2011.37'\n",
      "        print \n",
      "        print ('Matplotlib: John D. Hunter. Matplotlib: A 2D Graphics Environment ,'+\n",
      "               'Computing in Science & Engineering, 9, 90-95 (2007)')\n",
      "        print 'DOI:10.1109/MCSE.2007.55'\n",
      "        print\n",
      "        print ('Pandas: Wes McKinney. Data Structures for Statistical Computing '+\n",
      "               'in Python, Proceedings of the 9th Python in Science Conference, '+\n",
      "               '51-56 (2010)')\n",
      "        print 'http://conference.scipy.org/proceedings/scipy2010/mckinney.html'\n",
      "        print \n",
      "        print 'HTML.py: http://www.decalage.info/python/html'\n",
      "        print '----------------'\n",
      "        print 'Other References'\n",
      "        print '----------------'\n",
      "    print ('Biopython: Cock PJ, Antao T, Chang JT, Chapman BA, Cox CJ, Dalke A, Friedberg'+\n",
      "           ' I, Hamelryck T, Kauff F, Wilczynski B, and de Hoon MJ. Biopython: freely '+\n",
      "           ' available Python tools for computational molecular biology and'+\n",
      "           ' bioinformatics. Bioinformatics 2009 Jun 1; 25(11) 1422-3.'+\n",
      "           'doi:10.1093/bioinformatics/btp163 pmid:19304878')\n",
      "    print 'http://www.hubmed.org/display.cgi?uids=19304878'\n",
      "    print \n",
      "    print ('Cython: Stefan Behnel, Robert Bradshaw, Craig Citro, '+\n",
      "           'Lisandro Dalcin, Dag Sverre Seljebotn and Kurt Smith. Cython: The '+\n",
      "           'Best of Both Worlds, Computing in Science and Engineering, 13,'+\n",
      "           ' 31-39 (2011)')\n",
      "    print 'DOI:10.1109/MCSE.2010.118'\n",
      "    print\n",
      "    print 'Cloud: https://pypi.python.org/pypi/cloud/2.8.5'\n",
      "    \n",
      "    if cl.verbose_report:  \n",
      "        print ('DendroPy: Sukumaran, J. and Mark T. Holder. 2010. DendroPy: A Python '+\n",
      "               'library for phylogenetic computing. Bioinformatics 26: 1569-1571.')\n",
      "        print 'http://bioinformatics.oxfordjournals.org/content/26/12/1569.abstract'  \n",
      "        print\n",
      "\n",
      "        \n",
      "\n",
      "\n",
      "def chekpoint_project(pj):\n",
      "    if cl.altprojfile:\n",
      "        pickle_pj(pj, cl.altprojfile)\n",
      "    else:\n",
      "        pickle_pj(pj, cl.projfile)\n",
      "\n",
      "\n",
      "\n",
      "#######################################################################\n",
      "#                       Execution\n",
      "#######################################################################\n",
      "cline = ''\n",
      "\n",
      "for i in sys.argv:\n",
      "    cline += i+' '\n",
      "print 'ReproPhylo was called with:'\n",
      "print cline\n",
      "print\n",
      "\n",
      "if cl.__:\n",
      "    warnings.simplefilter('ignore')\n",
      "\n",
      "if cl.version_ctrl:\n",
      "    start_git()\n",
      "pj = make_project(loci_csv=cl.read_loci_csv, genbank=cl.genbank)    \n",
      "deal_with_fasta(pj)\n",
      "\n",
      "\n",
      "\n",
      "if cl.write_meta_csv:\n",
      "    pj.write(cl.write_meta_csv, format='csv')\n",
      "if cl.read_meta_csv:\n",
      "    pj.correct_metadata_from_file(cl.read_meta_csv)\n",
      "\n",
      "extract_and_log_by_locus(pj)\n",
      "\n",
      "run_default_analysis(pj)\n",
      "\n",
      "if cl.verbose_report:\n",
      "    print '---------------------------------'\n",
      "    print ('Writing report to %s.zip'%cl.verbose_report).title()\n",
      "    print '---------------------------------' \n",
      "    if cl.noninteractive:\n",
      "        if os.path.exists(cl.verbose_report):\n",
      "            removetype(cl.verbose_report)\n",
      "            if os.path.exists(cl.verbose_report.rpartition('.')[0]):\n",
      "                removetype(cl.verbose_report.rpartition('.')[0])\n",
      "        if os.path.exists(cl.verbose_report+'.zip'):\n",
      "            removetype(cl.verbose_report+'.zip')\n",
      "    publish(pj, cl.verbose_report, cl.figs)\n",
      "\n",
      "\n",
      "\n",
      "if cl.dump_alignments:\n",
      "    if len(pj.alignments.keys()+pj.trimmed_alignments.keys()) == 0:\n",
      "        print \"No alignments to print\"\n",
      "    else:\n",
      "        import zipfile, shutil\n",
      "        print '------------------------------------'\n",
      "        print ('Writing alignments in %s format'%cl.dump_alignments).title()\n",
      "        print '------------------------------------' \n",
      "        aln_files = pj.write_alns(format=cl.dump_alignments)\n",
      "        trim_aln_files = pj.write_trimmed_alns(format=cl.dump_alignments)\n",
      "        os.makedirs(cl.dump_alignments_zip.rpartition('.')[0])\n",
      "        for filename in aln_files + trim_aln_files:\n",
      "            shutil.move(filename, cl.dump_alignments_zip.rpartition('.')[0]+'/'+filename.split('/')[-1])\n",
      "        zf = zipfile.ZipFile(cl.dump_alignments_zip, \"w\")\n",
      "        for dirname, subdirs, files in os.walk(cl.dump_alignments_zip.rpartition('.')[0]):\n",
      "            zf.write(dirname)\n",
      "            for filename in files:\n",
      "                zf.write(os.path.join(dirname, filename))\n",
      "        zf.close()\n",
      "        shutil.rmtree(cl.dump_alignments_zip.rpartition('.')[0])\n",
      "    \n",
      "if cl.dump_records:\n",
      "    print '------------------------------------'\n",
      "    print ('Writing records to %s'%cl.dump_records).title()\n",
      "    print '------------------------------------' \n",
      "    pj.write(cl.dump_records, format='genbank')\n",
      "\n",
      "chekpoint_project(pj)\n",
      "ref()\n",
      "print '----'\n",
      "print 'DONE'\n",
      "print '----'\n",
      "#\"\"\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%file start_a_project.xml\n",
      "<tool id=\"Start a Project\" name=\"Start a Project\"  version=\"0.1\" hidden=\"false\">\n",
      "    <description>\n",
      "    </description>\n",
      "    <command interpreter=\"python\">\n",
      "        base_reprophylo.py\n",
      "        #if str($howinit.inputs)==\"gb\"\n",
      "        $projout -_ -g $genbank -w $lociout -y $metaout\n",
      "        #else if str($howinit.inputs)==\"loci\"\n",
      "        $projout -_ -r $loci_csv_in\n",
      "        #else if str($howinit.inputs)==\"locus\"\n",
      "        $projout -_ -u $howinit.chartype $howinit.featuretype $howinit.name\n",
      "        #else if str($howinit.inputs)==\"both\"\n",
      "        $projout -_ -r $loci_csv_in  -g $genbanki -y $metaout\n",
      "        #end if\n",
      "        > $log\n",
      "    </command>\n",
      "    <inputs>\n",
      "        <conditional name=\"howinit\">\n",
      "           <param name=\"inputs\" type=\"select\" label=\"Initiate the Project with\">\n",
      "                <option value=\"gb\">a GenBank file</option>\n",
      "                <option value=\"loci\">a loci CSV file</option> \n",
      "                <option value=\"locus\">feed in a locus</option> \n",
      "                <option value=\"both\">GenBank and loci CSV files</option> \n",
      "           </param>\n",
      "           <when value=\"gb\">\n",
      "                <param name=\"genbank\" type=\"data\" label=\"genbank or embl file\"/>\n",
      "           </when>\n",
      "           <when value=\"loci\">\n",
      "                <param name=\"loci_csv_in\" type=\"data\" label=\"Input loci CSV file\"/>\n",
      "           </when>\n",
      "           <when value=\"locus\">\n",
      "                <param name=\"chartype\" type=\"select\" label=\"DNA or Protein\">\n",
      "                    <option value=\"dna\">DNA</option>\n",
      "                    <option value=\"prot\">Protein</option>\n",
      "                </param>\n",
      "                <param name=\"featuretype\" type=\"text\" label=\"Feature type (eg, CDS, rRNA, Protein ...)\"/>\n",
      "                <param name=\"name\" type=\"text\" label=\"Gene name (eg cox1. To control aliases use a CSV)\"/>\n",
      "           </when>\n",
      "           <when value=\"both\">\n",
      "                <param name=\"loci_csv_in\" type=\"data\" label=\"Input loci CSV file\"/>\n",
      "                <param name=\"genbanki\" type=\"data\" label=\"genbank or embl file\"/>\n",
      "           </when>\n",
      "        </conditional>\n",
      "    </inputs>\n",
      "    <outputs>\n",
      "        <data name=\"log\" format='txt' label=\"${tool.name} on ${on_string}: Log\" />\n",
      "        <data name=\"projout\" format='data' label=\"${tool.name} on ${on_string}: ReproPhylo Project\" />\n",
      "        <data name=\"lociout\" format='txt' label=\"${tool.name} on ${on_string}: Loci CSV\" >\n",
      "            <filter>(howinit[\"inputs\"]==\"gb\") </filter>\n",
      "        </data>\n",
      "        <data name=\"metaout\" format='csv' label=\"${tool.name} on ${on_string}: Metadata CSV\" >\n",
      "            <filter>((howinit[\"inputs\"]==\"gb\") or (howinit[\"inputs\"]==\"both\"))</filter>\n",
      "        </data>\n",
      "    </outputs>\n",
      "    <help>\n",
      "    \n",
      "\n",
      "ReproPhylo\n",
      "==========\n",
      "\n",
      "\"Start a Project\"\n",
      "------------------\n",
      "\n",
      "This tool starts a ReproPhylo Project. For the very least, a project requires that the analysed loci are specified.\n",
      "Loci can be specified in three ways:  \n",
      "\n",
      "*A genbank or embl file*\n",
      "\n",
      "Genbank files contain information about the gene or product identity of the \n",
      "entries they contain. If ReproPhylo is started only with such a file, all\n",
      "the loci in the file will be used to initiate ReproPhylo. A loci CSV file (see below)\n",
      "will be printed. It can be edited and used to re-start a Project.\n",
      "                   \n",
      "\n",
      "*A loci CSV file*\n",
      "\n",
      "A loci CSV file lists the loci to be analysed. It consists of one line\n",
      "for each locus. The lines have the following form:\n",
      "char_type,feature_type,gene_name,alias1,alias2,...,aliasn\n",
      "\n",
      "**char_type** can be dna or prot\n",
      "\n",
      "**feature_type** could be CDS, rRNA, Protein, tRNA etc.\n",
      "\n",
      "**gene_name** is any name you want to call you locus.\n",
      "\n",
      "**aliases** are any name the locus is given in genbank under the 'gene' or 'product' qualifiers.\n",
      "\n",
      "If you first ran the tool using the 'genbank or embl file' option, you will get a loci CSV file \n",
      "containing all the loci in the genbank file, as output. If a certain gene apears in more than one\n",
      "name in the genbenk file, the loci CSV file is likely to account for this bay adding additional\n",
      "names as aliases of the first name. However, you may find that some loci have not been collapsed\n",
      "and apear in more than one line. In this case, you may add a coma and a shared integer at the\n",
      "end of lines you wish to synonymize. You may also edit the first value to switch between DNA and\n",
      "protein sequences.\n",
      "\n",
      "Example:\n",
      "\n",
      "\n",
      "The output loci CSV may be:\n",
      "\n",
      "\n",
      "``dna,rRNA,18S,18s,SSU,rRNA`` \n",
      "\n",
      "``dna,CDS,cox1,cox1``\n",
      "\n",
      "``dna,CDS,COI,COI``\n",
      "\n",
      "\n",
      "This file can be edited as follows:\n",
      "\n",
      "            \n",
      "``dna,rRNA,18S,18s,SSU,rRNA``\n",
      "\n",
      "``prot,CDS,cox1,cox1,1``\n",
      "\n",
      "``dna,CDS,COI,COI,1``\n",
      "            \n",
      "\n",
      "It will read by ReproPhylo as follows:\n",
      "\n",
      "            \n",
      "``dna,rRNA,18S,18s,SSU,rRNA`` \n",
      "\n",
      "``prot,CDS,cox1,cox1,COI``\n",
      "                            \n",
      "                           \n",
      "*Feed in a locus*\n",
      "\n",
      "If you intend to read a single fasta file with a single locus, you can feed \n",
      "the locus manually. You will need to specify a char_type (dna or prot), a feature\n",
      "type (CDS, Protein, rRNA etc...) and a gene name.\n",
      " \n",
      "    \n",
      "*GenBank and loci CSV files*\n",
      "\n",
      "\n",
      "Provide both a loci CSV and a genbank file. The Project will include \n",
      "the loci in the CSV file and all the genbank records that fit those loci.\n",
      "\n",
      "\n",
      "</help>\n",
      "</tool>    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%file read_data.xml\n",
      "<tool id=\"Read Data\" name=\"Read Data\"  version=\"0.1\" hidden=\"false\">\n",
      "    <description>\n",
      "    </description>\n",
      "    <command interpreter=\"python\">\n",
      "        base_reprophylo.py\n",
      "        $projin -_ -a $altprojout\n",
      "        #if str($readData.newData)==\"gb\"\n",
      "        -g $genbankii $readData.replacedata -y $metaout\n",
      "        #else if str($readData.newData)==\"fasta\"\n",
      "        -f $fastafile -t $readData.fastaCharType -n $readData.GenName $readData.FeatureType $readData.replacedata -y $metaout\n",
      "        #else if str($readData.newData)==\"both\"\n",
      "        -g $genbankii -f $readData.fastafile -t $readData.fastaCharType -n $readData.GenName $readData.FeatureType $readData.replacedata -y $metaout\n",
      "        #end if\n",
      "        > $log\n",
      "    </command>\n",
      "    <inputs>\n",
      "        <param name=\"projin\" type=\"data\" label=\"Input Project file\" />\n",
      "        <conditional name=\"readData\">\n",
      "            <param name=\"newData\" type=\"select\" label=\"Read new data?\">\n",
      "                <option value=\"gb\">Read a genbank file</option>\n",
      "                <option value=\"fasta\">Read a fasta file</option>\n",
      "                <option value=\"both\">Read a fasta and a genbank file</option>\n",
      "            </param>\n",
      "            <when value=\"gb\">\n",
      "                <param name=\"genbankii\" type=\"data\" label=\"Input genbank or embl file\"/>\n",
      "                <param name=\"replacedata\" type=\"boolean\" truevalue=\"-e\" falsevalue=\"\" label=\"Replace data with new one?\"/>\n",
      "            </when>\n",
      "            <when value=\"fasta\">\n",
      "                <param name=\"fastafile\" type=\"data\" label=\"Input fasta file\"/>\n",
      "                <param name=\"fastaCharType\" type=\"select\" label=\"DNA or Protein sequences?\">\n",
      "                    <option value='dna'>DNA</option>\n",
      "                    <option value='prot'>Protein</option>\n",
      "                </param>\n",
      "                <param name='GenName' type=\"text\" label=\"The gene name of the fasta sequences (eg cox1)\"/>\n",
      "                <param name='FeatureType' type=\"text\" lable=\"The fasta sequence feature type (eg CDS, rRNA...)\"/>\n",
      "                <param name=\"replacedata\" type=\"boolean\" truevalue=\"-e\" falsevalue=\"\" label=\"Replace data with new one?\"/>\n",
      "            </when>\n",
      "            <when value=\"both\">\n",
      "                <param name=\"genbankii\" type=\"data\" label=\"Input genbank or embl file\"/>\n",
      "                <param name=\"fastafile\" type=\"data\" lable=\"Input fasta file\"/>\n",
      "                <param name=\"fastaCharType\" type=\"select\" label=\"DNA or Protein sequences?\">\n",
      "                    <option value='dna'>DNA</option>\n",
      "                    <option value='prot'>Protein</option>\n",
      "                </param>\n",
      "                <param name='GenName' type=\"text\" label=\"The gene name of the fasta sequences (eg cox1)\"/>\n",
      "                <param name='FeatureType' type=\"text\" lable=\"The fasta sequence feature type (eg CDS, rRNA...)\"/>\n",
      "                <param name=\"replacedata\" type=\"boolean\" truevalue=\" -e \" falsevalue=\" \" label=\"Replace data with new one?\"/>\n",
      "            </when>\n",
      "        </conditional>\n",
      "    </inputs>\n",
      "    <outputs>\n",
      "        <data name=\"log\" format='txt' label=\"${tool.name} on ${on_string}: Log\" />\n",
      "        <data name=\"altprojout\" format='data' label=\"${tool.name} on ${on_string}: ReproPhylo Project\" />\n",
      "        <data name=\"metaout\" format='csv' label=\"${tool.name} on ${on_string}: Metadata CSV\" />\n",
      "    </outputs>\n",
      "    <help>\n",
      "    \n",
      "\n",
      "ReproPhylo\n",
      "==========\n",
      "\n",
      "\"Read Data\"\n",
      "-----------\n",
      "This tool allows to add or replace the data in the project. You can choose whether the new data will replace the existing one \n",
      "by checking ``Replace data with new one?``. If you read a fasta file you will need to feed in a char_type (dna or prot), \n",
      "a feature_type (CDS, rRNA etc...) and a gene name.\n",
      "\n",
      "</help>\n",
      "</tool>    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%file read_revised_metadata.xml\n",
      "<tool id=\"Read Revised Metadata\" name=\"Read Revised Metadata\"  version=\"0.1\" hidden=\"false\">\n",
      "    <description>\n",
      "    </description>\n",
      "    <command interpreter=\"python\">\n",
      "        base_reprophylo.py\n",
      "        $projin -_ -a $altprojout -z $metain \n",
      "        > $log\n",
      "    </command>\n",
      "    <inputs>\n",
      "        <param name=\"projin\" type=\"data\" label=\"Input Project file\" />\n",
      "        <param name=\"metain\" type=\"data\" label=\"Input metadata CSV\"/>\n",
      "    </inputs>\n",
      "    <outputs>\n",
      "        <data name=\"log\" format='txt' label=\"${tool.name} on ${on_string}: Log\" />\n",
      "        <data name=\"altprojout\" format='data' label=\"${tool.name} on ${on_string}: ReproPhylo Project\" />\n",
      "    </outputs>\n",
      "    <help>\n",
      "    \n",
      "\n",
      "ReproPhylo\n",
      "==========\n",
      "\n",
      "\"Read Revised Metadata\"\n",
      "-----------------------\n",
      "When you read data with `Read Data`, a CSV file containing the metadata of all the sequence features is printed.\n",
      "This is a tab delimited text file. It contains a line for each feature. For example, a single genbank entry that \n",
      "contains ten genes will produce 10 lines in the metadata CSV. In this example, each line starts with the \n",
      "source feature qualifiers. Since there is only one source feature in each genbank record, the first few (source)\n",
      "columns will be identical accross the ten lines. Then there's a taxonomy column and finaly columns representing\n",
      "each of the ten features' metadata. In order to extend or correct the metadata, you can manually edit this file in\n",
      "excell and read it back using this tool. You can modify existing columns or add new ones. This file should contain\n",
      "any data relevant to your experiment, eg, morphological, geographic or ecological data. This way it will from now\n",
      "on be associated with the tree's leaves and you will be able to use it for tree annotations and graphics.\n",
      "\n",
      "</help>\n",
      "</tool>    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%file run_exploratory_pipeline.xml\n",
      "<tool id=\"Run Exploratory Pipeline\" name=\"Run Exploratory Pipeline\"  version=\"0.1\" hidden=\"false\">\n",
      "    <description>\n",
      "    </description>\n",
      "    <command interpreter=\"python\">\n",
      "        base_reprophylo.py\n",
      "        $projin -_ -a $altprojout\n",
      "        $overwrite -s  --threads $threads --noninteractive -j $figs\n",
      "        > $log\n",
      "    </command>\n",
      "    <inputs>\n",
      "        <param name=\"projin\" type=\"data\" label=\"Input Project file\" />\n",
      "        <param name=\"threads\" type=\"text\" value=\"4\" label=\"How many threads?\"/>\n",
      "        <param name=\"overwrite\" type=\"boolean\" truevalue=\"-x\" falsevalue=\"\" label=\"Overwrite Project's analyses?\"/>\n",
      "        <param name=\"figs\" type=\"text\" labe=\"A local directory to write the figures to\"/>\n",
      "    </inputs>\n",
      "    <outputs>\n",
      "        <data name=\"log\" format='txt' label=\"${tool.name} on ${on_string}: Log\" />\n",
      "        <data name=\"altprojout\" format='data' label=\"${tool.name} on ${on_string}: ReproPhylo Project\" />\n",
      "    </outputs>\n",
      "    <help>\n",
      "    \n",
      "\n",
      "ReproPhylo\n",
      "==========\n",
      "\n",
      "\"Run Exploratory Pipeline\"\n",
      "--------------------------\n",
      "Runs a quick exploratory gene tree reconstructions (specify pipeline here)\n",
      "\n",
      "</help>\n",
      "</tool>    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%file report.xml\n",
      "<tool id=\"Report\" name=\"Report\"  version=\"0.1\" hidden=\"false\">\n",
      "    <description>\n",
      "    </description>\n",
      "    <command interpreter=\"python\">\n",
      "        base_reprophylo.py \n",
      "        $projin -_ -a $altprojout\n",
      "        #if (str($moreOptions.action)==\"report\")\n",
      "        -v report_arcive -j $moreOptions.figsdir  --noninteractive \n",
      "        #end if\n",
      "        #if (str($moreOptions.action)==\"dump\")\n",
      "        -m $moreOptions.format -p gbout \n",
      "        #end if\n",
      "        > $log\n",
      "    </command>\n",
      "    <inputs>\n",
      "        <param name=\"projin\" type=\"data\" label=\"Input Project file\" />\n",
      "        <conditional name=\"moreOptions\">\n",
      "            <param name=\"action\" type=\"select\" label=\"Misc Utils\">\n",
      "                <option value=\"report\">Dump a full report to zip file</option>\n",
      "                <option value=\"dump\">Dump sequence alignments and reocrds</option>\n",
      "            </param>\n",
      "            <when value=\"report\">\n",
      "                <param name=\"figsdir\" type=\"text\" label=\"Local path to figures\"/>\n",
      "            </when>\n",
      "            <when value=\"dump\">\n",
      "                <param name=\"format\" type=\"select\" label=\"Choose alignement format\" >\n",
      "                    <option value=\"fasta\">Fasta</option>\n",
      "                    <option value=\"phylip-relaxed\">Phylip</option>\n",
      "                    <option value=\"nexus\">Nexus</option>\n",
      "                    <option value=\"stockholm\">Stockholm</option>\n",
      "                    <option value=\"clustal\">Clustal</option>\n",
      "                </param>\n",
      "            </when>\n",
      "        </conditional>\n",
      "    </inputs>\n",
      "    <outputs>\n",
      "        <data name=\"log\" format='txt' label=\"${tool.name} on ${on_string}: Log\" />\n",
      "        <data name=\"altprojout\" format='data' label=\"${tool.name} on ${on_string}: ReproPhylo Project\" />\n",
      "        <data name=\"reportName\" from_work_dir=\"report_arcive.zip\" format=\"data\" label=\"${tool.name} on ${on_string}: Report archive\">\n",
      "            <filter>(moreOptions[\"action\"]==\"report\")</filter>\n",
      "        </data>\n",
      "        <data name=\"dumpzip\" from_work_dir=\"alignments.zip\" format=\"data\" label=\"${tool.name} on ${on_string}: Sequence alignments archive\">\n",
      "            <filter>(moreOptions[\"action\"]==\"dump\")</filter>\n",
      "        </data>\n",
      "        <data name=\"gbout\" format=\"txt\" from_work_dir=\"gbout\" label=\"${tool.name} on ${on_string}: Sequence records genbank file\">\n",
      "            <filter>(moreOptions[\"action\"]==\"dump\")</filter>\n",
      "        </data>\n",
      "    </outputs>\n",
      "    <help>\n",
      "    \n",
      "\n",
      "ReproPhylo\n",
      "==========\n",
      "\n",
      "\"Report\"\n",
      "--------\n",
      "This tool can either dump all the seuqnece alignments (raw and trimmed) in the Project or print a verbose html report that includes\n",
      "all the methods and results, as well as statistics concerning the quality of sequence data, sequence alignments and trees.\n",
      "</help>\n",
      "</tool>    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%file manage_project.xml\n",
      "<tool id=\"Manage a Project\" name=\"Manage a Project\"  version=\"0.1\" hidden=\"false\">\n",
      "    <description>\n",
      "    </description>\n",
      "    <command interpreter=\"python\">\n",
      "        base_reprophylo.py\n",
      "        #if str($theProjectIsNew.isnew)==\"yes\" and str($theProjectIsNew.howinit.inputs)==\"gb\"\n",
      "        $projout -_ -g $genbank -w $lociout -y $metaout\n",
      "        #else if str($theProjectIsNew.isnew)==\"yes\" and str($theProjectIsNew.howinit.inputs)==\"loci\"\n",
      "        $projout -_ -r $loci_csv_in\n",
      "        #else if str($theProjectIsNew.isnew)==\"yes\" and str($theProjectIsNew.howinit.inputs)==\"locus\"\n",
      "        $projout -_ -u $theProjectIsNew.howinit.chartype $theProjectIsNew.howinit.featuretype $theProjectIsNew.howinit.name\n",
      "        #else if str($theProjectIsNew.isnew)==\"yes\" and str($theProjectIsNew.howinit.inputs)==\"both\"\n",
      "        $projout -_ -r $loci_csv_in  -g $genbanki -y $metaout\n",
      "        #else if str($theProjectIsNew.isnew)==\"no\" \n",
      "        $projin -_ -a $altprojout\n",
      "        #end if\n",
      "        #if (str($theProjectIsNew.isnew)==\"no\") and (str($theProjectIsNew.moreOptions.action)==\"writemeta\")\n",
      "        -y $metaout \n",
      "        #end if\n",
      "        #if (str($theProjectIsNew.isnew)==\"no\") and (str($theProjectIsNew.moreOptions.action)==\"writeloci\")\n",
      "        -w $lociout \n",
      "        #end if\n",
      "        #if (str($theProjectIsNew.isnew)==\"no\") and (str($theProjectIsNew.moreOptions.action)==\"readmeta\")\n",
      "        -z $theProjectIsNew.moreOptions.metain \n",
      "        #end if\n",
      "        #if (str($theProjectIsNew.isnew)==\"no\") and (str($theProjectIsNew.moreOptions.action)==\"report\")\n",
      "        -v report_arcive -j $theProjectIsNew.moreOptions.figsdir  --noninteractive \n",
      "        #end if\n",
      "        #if (str($theProjectIsNew.isnew)==\"no\") and (str($theProjectIsNew.moreOptions.action)==\"dump\")\n",
      "        -m $theProjectIsNew.moreOptions.format -p gbout \n",
      "        #end if\n",
      "        #if str($readData.newData)==\"gb\"\n",
      "        -g $genbankii $readData.replacedata -y $metaout\n",
      "        #else if str($readData.newData)==\"fasta\"\n",
      "        -f $fastafile -t $readData.fastaCharType -n $readData.GenName $readData.FeatureType $readData.replacedata -y $metaout\n",
      "        #else if str($readData.newData)==\"both\"\n",
      "        -g $genbankii -f $readData.fastafile -t $readData.fastaCharType -n $readData.GenName $readData.FeatureType $readData.replacedata -y $metaout\n",
      "        #end if\n",
      "        #if str($readLoci.newLoci)==\"csv\"\n",
      "         -r $readLoci.loci_csv_in $readLoci.replaceloci\n",
      "        #else if str($readLoci.newLoci)==\"cline\"\n",
      "        -u $readLoci.CharType $readLoci.FeatureType $readLoci.GenName $readLoci.replaceloci\n",
      "        #end if \n",
      "        #if str($runAnalysis.execute)==\"yes\"\n",
      "        $runAnalysis.overwrite -s  --threads $runAnalysis.threads --noninteractive -j $runAnalysis.figs\n",
      "        #end if\n",
      "        > $log\n",
      "    </command>\n",
      "    <inputs>\n",
      "        <conditional name=\"theProjectIsNew\">\n",
      "            <param name=\"isnew\" type=\"select\" label=\"This is a new project\">\n",
      "                <option value=\"yes\">Yes</option>\n",
      "                <option value=\"no\">No</option>\n",
      "            </param>\n",
      "            <when value=\"yes\">\n",
      "                <conditional name=\"howinit\">\n",
      "                   <param name=\"inputs\" type=\"select\" label=\"Initiate the Project with\">\n",
      "                        <option value=\"gb\">a GenBank file</option>\n",
      "                        <option value=\"loci\">a loci CSV file</option> \n",
      "                        <option value=\"locus\">feed in a locus</option> \n",
      "                        <option value=\"both\">GenBank and loci CSV files</option> \n",
      "                   </param>\n",
      "                   <when value=\"gb\">\n",
      "                        <param name=\"genbank\" type=\"data\" label=\"genbank or embl file\"/>\n",
      "                   </when>\n",
      "                   <when value=\"loci\">\n",
      "                        <param name=\"loci_csv_in\" type=\"data\" label=\"Input loci CSV file\"/>\n",
      "                   </when>\n",
      "                   <when value=\"locus\">\n",
      "                        <param name=\"chartype\" type=\"select\" label=\"DNA or Protein\">\n",
      "                            <option value=\"dna\">DNA</option>\n",
      "                            <option value=\"prot\">Protein</option>\n",
      "                        </param>\n",
      "                        <param name=\"featuretype\" type=\"text\" label=\"Feature type (eg, CDS, rRNA, Protein ...)\"/>\n",
      "                        <param name=\"name\" type=\"text\" label=\"Gene name (eg cox1. To control aliases use a CSV)\"/>\n",
      "                   </when>\n",
      "                   <when value=\"both\">\n",
      "                        <param name=\"loci_csv_in\" type=\"data\" label=\"Input loci CSV file\"/>\n",
      "                        <param name=\"genbanki\" type=\"data\" label=\"genbank or embl file\"/>\n",
      "                   </when>\n",
      "                </conditional>\n",
      "            </when>\n",
      "            <when value=\"no\">\n",
      "                <param name=\"projin\" type=\"data\" label=\"Input Project file\" />\n",
      "                <conditional name=\"moreOptions\">\n",
      "                    <param name=\"action\" type=\"select\" label=\"Misc Utils\">\n",
      "                        <option value=\"none\">None</option>\n",
      "                        <option value=\"writemeta\">Write a metadta CSV file</option>\n",
      "                        <option value=\"writeloci\">Write a loci CSV (not when reading)</option>\n",
      "                        <option value=\"readmeta\">Read a metadata CSV</option>\n",
      "                        <option value=\"report\">Dump a full report to zip file</option>\n",
      "                        <option value=\"dump\">Dump sequence alignments and reocrds</option>\n",
      "                    </param>\n",
      "                    <when value=\"none\"/>\n",
      "                    <when value=\"writemeta\"/>\n",
      "                    <whne value=\"writeloci\"/>\n",
      "                    <when value=\"readmeta\">\n",
      "                        <param name=\"metain\" type=\"data\" label=\"Input metadata CSV\"/>\n",
      "                    </when>\n",
      "                    <when value=\"report\">\n",
      "                        <param name=\"figsdir\" type=\"text\" label=\"Local path to figures\"/>\n",
      "                    </when>\n",
      "                    <when value=\"dump\">\n",
      "                        <param name=\"format\" type=\"select\" label=\"Choose alignement format\" >\n",
      "                            <option value=\"fasta\">Fasta</option>\n",
      "                            <option value=\"phylip-relaxed\">Phylip</option>\n",
      "                            <option value=\"nexus\">Nexus</option>\n",
      "                            <option value=\"stockholm\">Stockholm</option>\n",
      "                            <option value=\"clustal\">Clustal</option>\n",
      "                        </param>\n",
      "                    </when>\n",
      "                </conditional>\n",
      "            </when>\n",
      "        </conditional>\n",
      "        <conditional name=\"readData\">\n",
      "            <param name=\"newData\" type=\"select\" label=\"Read new data?\">\n",
      "                <option value=\"no\">No</option>\n",
      "                <option value=\"gb\">Read a genbank file</option>\n",
      "                <option value=\"fasta\">Read a fasta file</option>\n",
      "                <option value=\"both\">Read a fasta and a genbank file</option>\n",
      "            </param>\n",
      "            <when value=\"no\"/>\n",
      "            <when value=\"gb\">\n",
      "                <param name=\"genbankii\" type=\"data\" label=\"Input genbank or embl file\"/>\n",
      "                <param name=\"replacedata\" type=\"boolean\" truevalue=\"-e\" falsevalue=\"\" label=\"Replace data with new one?\"/>\n",
      "            </when>\n",
      "            <when value=\"fasta\">\n",
      "                <param name=\"fastafile\" type=\"data\" label=\"Input fasta file\"/>\n",
      "                <param name=\"fastaCharType\" type=\"select\" label=\"DNA or Protein sequences?\">\n",
      "                    <option value='dna'>DNA</option>\n",
      "                    <option value='prot'>Protein</option>\n",
      "                </param>\n",
      "                <param name='GenName' type=\"text\" label=\"The gene name of the fasta sequences (eg cox1)\"/>\n",
      "                <param name='FeatureType' type=\"text\" lable=\"The fasta sequence feature type (eg CDS, rRNA...)\"/>\n",
      "                <param name=\"replacedata\" type=\"boolean\" truevalue=\"-e\" falsevalue=\"\" label=\"Replace data with new one?\"/>\n",
      "            </when>\n",
      "            <when value=\"both\">\n",
      "                <param name=\"genbankii\" type=\"data\" label=\"Input genbank or embl file\"/>\n",
      "                <param name=\"fastafile\" type=\"data\" lable=\"Input fasta file\"/>\n",
      "                <param name=\"fastaCharType\" type=\"select\" label=\"DNA or Protein sequences?\">\n",
      "                    <option value='dna'>DNA</option>\n",
      "                    <option value='prot'>Protein</option>\n",
      "                </param>\n",
      "                <param name='GenName' type=\"text\" label=\"The gene name of the fasta sequences (eg cox1)\"/>\n",
      "                <param name='FeatureType' type=\"text\" lable=\"The fasta sequence feature type (eg CDS, rRNA...)\"/>\n",
      "                <param name=\"replacedata\" type=\"boolean\" truevalue=\" -e \" falsevalue=\" \" label=\"Replace data with new one?\"/>\n",
      "            </when>\n",
      "        </conditional>\n",
      "        <conditional name=\"readLoci\">\n",
      "            <param name=\"newLoci\" type=\"select\" label=\"Read new loci?\">\n",
      "                <option value=\"no\">No</option>\n",
      "                <option value=\"csv\">from a loci CSV</option>\n",
      "                <option value=\"cline\">feed a locus manually</option>\n",
      "            </param>\n",
      "            <when value=\"no\"/>\n",
      "            <when value=\"csv\">\n",
      "                <param name=\"loci_csv_in\" type=\"data\" label=\"Input loci CSV file\"/>\n",
      "                <param name=\"replaceloci\" type=\"boolean\" truevalue=\"-b\" falsevalue=\"\" label=\"Replace loci with new ones?\"/>\n",
      "            </when>\n",
      "            <when value=\"cline\">\n",
      "                <param name=\"CharType\" type=\"select\" label=\"DNA or Protein locus?\">\n",
      "                    <option value='dna'>DNA</option>\n",
      "                    <option value='prot'>Protein</option>\n",
      "                </param>\n",
      "                <param name='FeatureType' type=\"text\" lable=\"The feature type (eg CDS, rRNA...)\"/>\n",
      "                <param name='GenName' type=\"text\" label=\"The gene name (eg cox1)\"/>\n",
      "                <param name=\"replaceloci\" type=\"boolean\" truevalue=\"-b\" falsevalue=\"\" label=\"Replace loci with new ones?\"/>\n",
      "            </when>\n",
      "        </conditional>\n",
      "        <conditional name=\"runAnalysis\">\n",
      "            <param name=\"execute\" type=\"select\" label=\"Run exploratory analysis\" >\n",
      "                <option value=\"no\">No</option>\n",
      "                <option value=\"yes\">Yes</option>\n",
      "            </param>\n",
      "            <when value=\"no\"/>\n",
      "            <when value=\"yes\">\n",
      "                <param name=\"threads\" type=\"text\" value=\"4\" label=\"How many threads?\"/>\n",
      "                <param name=\"overwrite\" type=\"boolean\" truevalue=\"-x\" falsevalue=\"\" label=\"Overwrite Project's analyses?\"/>\n",
      "                <param name=\"figs\" type=\"text\" labe=\"A local directory to write the figures to\"/>\n",
      "            </when>\n",
      "        </conditional>\n",
      "    </inputs>\n",
      "    <outputs>\n",
      "        <data name=\"log\" format='txt' label=\"${tool.name} on ${on_string}: Log\" />\n",
      "        <data name=\"projout\" format='data' label=\"${tool.name} on ${on_string}: ReproPhylo Project\" >\n",
      "            <filter>(theProjectIsNew[\"isnew\"]==\"yes\")</filter>\n",
      "        </data>\n",
      "        <data name=\"altprojout\" format='data' label=\"${tool.name} on ${on_string}: ReproPhylo Project\" >\n",
      "            <filter>(theProjectIsNew[\"isnew\"]==\"no\")</filter>\n",
      "        </data>\n",
      "        <data name=\"lociout\" format='txt' label=\"${tool.name} on ${on_string}: Loci CSV\" >\n",
      "            <filter>(((theProjectIsNew[\"isnew\"]==\"yes\") and (theProjectIsNew[\"howinit\"][\"inputs\"]==\"gb\")) or\n",
      "                     ((theProjectIsNew[\"isnew\"]==\"no\") and (theProjectIsNew[\"moreOptions\"][\"action\"]==\"writeloci\")))</filter>\n",
      "        </data>\n",
      "        <data name=\"metaout\" format='csv' label=\"${tool.name} on ${on_string}: Metadata CSV\" >\n",
      "            <filter>(((theProjectIsNew[\"isnew\"]==\"yes\") and (theProjectIsNew[\"howinit\"][\"inputs\"]==\"gb\")) or \n",
      "                     ((theProjectIsNew[\"isnew\"]==\"yes\") and (theProjectIsNew[\"howinit\"][\"inputs\"]==\"both\")) or\n",
      "                     (readData[\"newData\"]!=\"no\") or\n",
      "                     (theProjectIsNew[\"moreOptions\"][\"action\"]==\"writemeta\"))\n",
      "            </filter>\n",
      "        </data>\n",
      "        <data name=\"reportName\" from_work_dir=\"report_arcive.zip\" format=\"data\" label=\"${tool.name} on ${on_string}: Report archive\">\n",
      "            <filter>((theProjectIsNew[\"isnew\"]==\"no\") and (theProjectIsNew[\"moreOptions\"][\"action\"]==\"report\"))</filter>\n",
      "        </data>\n",
      "        <data name=\"dumpzip\" from_work_dir=\"alignments.zip\" format=\"data\" label=\"${tool.name} on ${on_string}: Sequence alignments archive\">\n",
      "            <filter>((theProjectIsNew[\"isnew\"]==\"no\") and (theProjectIsNew[\"moreOptions\"][\"action\"]==\"dump\"))</filter>\n",
      "        </data>\n",
      "        <data name=\"gbout\" format=\"txt\" from_work_dir=\"gbout\" label=\"${tool.name} on ${on_string}: Sequence records genbank file\">\n",
      "            <filter>((theProjectIsNew[\"isnew\"]==\"no\") and (theProjectIsNew[\"moreOptions\"][\"action\"]==\"dump\"))</filter>\n",
      "        </data>\n",
      "    </outputs>\n",
      "    <help>\n",
      "    \n",
      "\n",
      "ReproPhylo\n",
      "==========\n",
      "\n",
      "\"Manage a Project\"\n",
      "----------------\n",
      "This tool is the starting point when working with ReproPhylo in galaxy. It allows to:\n",
      "    \n",
      "1. Explore a genbank file\n",
      "\n",
      "2. Pick loci to analyze\n",
      "\n",
      "3. Add your own data\n",
      "\n",
      "4. Manipulate the metadata\n",
      "\n",
      "5. Run a quick exploratory phylogenetic analysis to make gene trees\n",
      "\n",
      "6. Print a detailed report of the analysis, specifically the taxonomic content, sequence\n",
      "and alignment statistics, the resulting trees and their pairwise comparison.\n",
      "\n",
      "Other ReproPhylo tools may then be used to create configure through phylogenetic analyses. \n",
      "This help section provides use cases to illustrate the usage of this tool.\n",
      "\n",
      "Tetillidae Tutorial\n",
      "===================\n",
      "Geting reference data\n",
      "---------------------\n",
      "\n",
      "This tutorial will make use of the Tetillidae sequence in genbank. To get it, use the search\n",
      "phrase ``Tetillidae[orgn]`` in the NCBI Nucleotide database and download the matches as a genbank\n",
      "file\n",
      "\n",
      "1. Go to the NCBI Nucleotide page at http://www.ncbi.nlm.nih.gov/nuccore/\n",
      "        \n",
      "2. Paste the phrase ``Tetillidae[orgn]`` in the search box\n",
      "\n",
      "3. On the top right hand side click the drop down menu \"Send to:\"\n",
      "\n",
      "4. Choose \"File\"\n",
      "\n",
      "5. Choose \"GenBank\" in the \"Format\" drop down menu\n",
      "\n",
      "6. Click \"Create File\" and save the file\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "</help>\n",
      "</tool>    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}